
<!DOCTYPE html>


<html lang="en" data-content_root="../../" >

  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="viewport" content="width=device-width, initial-scale=1" />

    <title>4. An Introduction to Principal Component Analysis (PCA) &#8212; Introduction to Machine Learning for Science</title>
  
  
  
  <script data-cfasync="false">
    document.documentElement.dataset.mode = localStorage.getItem("mode") || "";
    document.documentElement.dataset.theme = localStorage.getItem("theme") || "";
  </script>
  
  <!-- Loaded before other Sphinx assets -->
  <link href="../../_static/styles/theme.css?digest=dfe6caa3a7d634c4db9b" rel="stylesheet" />
<link href="../../_static/styles/bootstrap.css?digest=dfe6caa3a7d634c4db9b" rel="stylesheet" />
<link href="../../_static/styles/pydata-sphinx-theme.css?digest=dfe6caa3a7d634c4db9b" rel="stylesheet" />

  
  <link href="../../_static/vendor/fontawesome/6.5.2/css/all.min.css?digest=dfe6caa3a7d634c4db9b" rel="stylesheet" />
  <link rel="preload" as="font" type="font/woff2" crossorigin href="../../_static/vendor/fontawesome/6.5.2/webfonts/fa-solid-900.woff2" />
<link rel="preload" as="font" type="font/woff2" crossorigin href="../../_static/vendor/fontawesome/6.5.2/webfonts/fa-brands-400.woff2" />
<link rel="preload" as="font" type="font/woff2" crossorigin href="../../_static/vendor/fontawesome/6.5.2/webfonts/fa-regular-400.woff2" />

    <link rel="stylesheet" type="text/css" href="../../_static/pygments.css?v=03e43079" />
    <link rel="stylesheet" type="text/css" href="../../_static/styles/sphinx-book-theme.css?v=eba8b062" />
    <link rel="stylesheet" type="text/css" href="../../_static/togglebutton.css?v=13237357" />
    <link rel="stylesheet" type="text/css" href="../../_static/copybutton.css?v=76b2166b" />
    <link rel="stylesheet" type="text/css" href="../../_static/mystnb.8ecb98da25f57f5357bf6f572d296f466b2cfe2517ffebfabe82451661e28f02.css" />
    <link rel="stylesheet" type="text/css" href="../../_static/sphinx-thebe.css?v=4fa983c6" />
    <link rel="stylesheet" type="text/css" href="../../_static/exercise.css" />
    <link rel="stylesheet" type="text/css" href="../../_static/tabs.css?v=4c969af8" />
    <link rel="stylesheet" type="text/css" href="../../_static/proof.css" />
    <link rel="stylesheet" type="text/css" href="../../_static/styles/sphinx-examples.css?v=e236af4b" />
    <link rel="stylesheet" type="text/css" href="../../_static/css/tooltipster.custom.css" />
    <link rel="stylesheet" type="text/css" href="../../_static/css/tooltipster.bundle.min.css" />
    <link rel="stylesheet" type="text/css" href="../../_static/css/tooltipster-sideTip-shadow.min.css" />
    <link rel="stylesheet" type="text/css" href="../../_static/css/tooltipster-sideTip-punk.min.css" />
    <link rel="stylesheet" type="text/css" href="../../_static/css/tooltipster-sideTip-noir.min.css" />
    <link rel="stylesheet" type="text/css" href="../../_static/css/tooltipster-sideTip-light.min.css" />
    <link rel="stylesheet" type="text/css" href="../../_static/css/tooltipster-sideTip-borderless.min.css" />
    <link rel="stylesheet" type="text/css" href="../../_static/css/micromodal.css" />
    <link rel="stylesheet" type="text/css" href="../../_static/sphinx-design.min.css?v=95c83b7e" />
    <link rel="stylesheet" type="text/css" href="../../_static/styles.css?v=76e7d31e" />
  
  <!-- Pre-loaded scripts that we'll load fully later -->
  <link rel="preload" as="script" href="../../_static/scripts/bootstrap.js?digest=dfe6caa3a7d634c4db9b" />
<link rel="preload" as="script" href="../../_static/scripts/pydata-sphinx-theme.js?digest=dfe6caa3a7d634c4db9b" />
  <script src="../../_static/vendor/fontawesome/6.5.2/js/all.min.js?digest=dfe6caa3a7d634c4db9b"></script>

    <script src="../../_static/jquery.js?v=5d32c60e"></script>
    <script src="../../_static/_sphinx_javascript_frameworks_compat.js?v=2cd50e6c"></script>
    <script src="../../_static/documentation_options.js?v=9eb32ce0"></script>
    <script src="../../_static/doctools.js?v=9a2dae69"></script>
    <script src="../../_static/sphinx_highlight.js?v=dc90522c"></script>
    <script src="../../_static/clipboard.min.js?v=a7894cd8"></script>
    <script src="../../_static/copybutton.js?v=f281be69"></script>
    <script src="../../_static/scripts/sphinx-book-theme.js?v=887ef09a"></script>
    <script src="../../_static/tabs.js?v=3ee01567"></script>
    <script src="../../_static/js/hoverxref.js"></script>
    <script src="../../_static/js/tooltipster.bundle.min.js"></script>
    <script src="../../_static/js/micromodal.min.js"></script>
    <script>let toggleHintShow = 'Click to show';</script>
    <script>let toggleHintHide = 'Click to hide';</script>
    <script>let toggleOpenOnPrint = 'true';</script>
    <script src="../../_static/togglebutton.js?v=4a39c7ea"></script>
    <script>var togglebuttonSelector = '.toggle, .admonition.dropdown';</script>
    <script src="../../_static/design-tabs.js?v=f930bc37"></script>
    <script async="async" src="https://www.googletagmanager.com/gtag/js?id=G-JMZ8EV2JDJ"></script>
    <script>
                window.dataLayer = window.dataLayer || [];
                function gtag(){ dataLayer.push(arguments); }
                gtag('js', new Date());
                gtag('config', 'G-JMZ8EV2JDJ');
            </script>
    <script>const THEBE_JS_URL = "https://unpkg.com/thebe@0.8.2/lib/index.js"; const thebe_selector = ".thebe,.cell"; const thebe_selector_input = "pre"; const thebe_selector_output = ".output, .cell_output"</script>
    <script async="async" src="../../_static/sphinx-thebe.js?v=c100c467"></script>
    <script>var togglebuttonSelector = '.toggle, .admonition.dropdown';</script>
    <script>
                window.dataLayer = window.dataLayer || [];
                function gtag(){ dataLayer.push(arguments); }
                gtag('js', new Date());
                gtag('config', 'G-JMZ8EV2JDJ');
            </script>
    <script>const THEBE_JS_URL = "https://unpkg.com/thebe@0.8.2/lib/index.js"; const thebe_selector = ".thebe,.cell"; const thebe_selector_input = "pre"; const thebe_selector_output = ".output, .cell_output"</script>
    <script>window.MathJax = {"options": {"processHtmlClass": "tex2jax_process|mathjax_process|math|output_area"}}</script>
    <script defer="defer" src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-mml-chtml.js"></script>
    <script>DOCUMENTATION_OPTIONS.pagename = 'lectures/03-PCA/Intro-PCA';</script>
    <script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
    <link rel="index" title="Index" href="../../genindex.html" />
    <link rel="search" title="Search" href="../../search.html" />
    <link rel="next" title="5. Python Review:" href="../01-reviewPython/PythonReview.html" />
    <link rel="prev" title="3. Pandas for data cleaning and analysis" href="../02-LimpiezayPrepDatos/DataCleaning.html" />
  <meta name="viewport" content="width=device-width, initial-scale=1"/>
  <meta name="docsearch:language" content="en"/>
  </head>
  
  
  <body data-bs-spy="scroll" data-bs-target=".bd-toc-nav" data-offset="180" data-bs-root-margin="0px 0px -60%" data-default-mode="">

  
  
  <div id="pst-skip-link" class="skip-link d-print-none"><a href="#main-content">Skip to main content</a></div>
  
  <div id="pst-scroll-pixel-helper"></div>
  
  <button type="button" class="btn rounded-pill" id="pst-back-to-top">
    <i class="fa-solid fa-arrow-up"></i>Back to top</button>

  
  <input type="checkbox"
          class="sidebar-toggle"
          id="pst-primary-sidebar-checkbox"/>
  <label class="overlay overlay-primary" for="pst-primary-sidebar-checkbox"></label>
  
  <input type="checkbox"
          class="sidebar-toggle"
          id="pst-secondary-sidebar-checkbox"/>
  <label class="overlay overlay-secondary" for="pst-secondary-sidebar-checkbox"></label>
  
  <div class="search-button__wrapper">
    <div class="search-button__overlay"></div>
    <div class="search-button__search-container">
<form class="bd-search d-flex align-items-center"
      action="../../search.html"
      method="get">
  <i class="fa-solid fa-magnifying-glass"></i>
  <input type="search"
         class="form-control"
         name="q"
         id="search-input"
         placeholder="Search this book..."
         aria-label="Search this book..."
         autocomplete="off"
         autocorrect="off"
         autocapitalize="off"
         spellcheck="false"/>
  <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd>K</kbd></span>
</form></div>
  </div>

  <div class="pst-async-banner-revealer d-none">
  <aside id="bd-header-version-warning" class="d-none d-print-none" aria-label="Version warning"></aside>
</div>

  
    <header class="bd-header navbar navbar-expand-lg bd-navbar d-print-none">
    </header>
  

  <div class="bd-container">
    <div class="bd-container__inner bd-page-width">
      
      
      
      <div class="bd-sidebar-primary bd-sidebar">
        

  
  <div class="sidebar-header-items sidebar-primary__section">
    
    
    
    
  </div>
  
    <div class="sidebar-primary-items__start sidebar-primary__section">
        <div class="sidebar-primary-item">

  
    
  

<a class="navbar-brand logo" href="../../README.html">
  
  
  
  
  
  
    <p class="title logo__title">Introduction to Machine Learning for Science</p>
  
</a></div>
        <div class="sidebar-primary-item">

 <script>
 document.write(`
   <button class="btn search-button-field search-button__button" title="Search" aria-label="Search" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <i class="fa-solid fa-magnifying-glass"></i>
    <span class="search-button__default-text">Search</span>
    <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd class="kbd-shortcut__modifier">K</kbd></span>
   </button>
 `);
 </script></div>
        <div class="sidebar-primary-item"><nav class="bd-links bd-docs-nav" aria-label="Main">
    <div class="bd-toc-item navbar-nav active">
        
        <ul class="nav bd-sidenav bd-sidenav__home-link">
            <li class="toctree-l1">
                <a class="reference internal" href="../../README.html">
                    Introduction to Machine Learning for Science
                </a>
            </li>
        </ul>
        <p aria-level="2" class="caption" role="heading"><span class="caption-text">Introduction</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../00-IntroductionCourseML/IntroCourse-MachineLearning.html">1. Introduction to the course and to Machine Learning</a></li>
<li class="toctree-l1"><a class="reference internal" href="../00-TALK-IntroML/IntroMachineLearning.html">2. A fast introduction to Machine Learning</a></li>
<li class="toctree-l1"><a class="reference internal" href="../02-LimpiezayPrepDatos/DataCleaning.html">3. Pandas for data cleaning and analysis</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Unsupervised learning I</span></p>
<ul class="current nav bd-sidenav">
<li class="toctree-l1 current active"><a class="current reference internal" href="#">4. An Introduction to Principal Component Analysis (PCA)</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Appendix</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../01-reviewPython/PythonReview.html">5. Python Review:</a></li>
<li class="toctree-l1"><a class="reference internal" href="../01-reviewPython/01-IntroProgrammingPython.html">6. Python Programming (very fast) Introduction</a></li>
<li class="toctree-l1"><a class="reference internal" href="../01-reviewPython/02-IntroProgrammingPython-DataStructs.html">7. Intro Python II: python data structures</a></li>
<li class="toctree-l1"><a class="reference internal" href="../02-LimpiezayPrepDatos/TutorialPandas.html">8. Pandas for data cleaning and analysis</a></li>
</ul>

    </div>
</nav></div>
    </div>
  
  
  <div class="sidebar-primary-items__end sidebar-primary__section">
  </div>
  
  <div id="rtd-footer-container"></div>


      </div>
      
      <main id="main-content" class="bd-main" role="main">
        
        

<div class="sbt-scroll-pixel-helper"></div>

          <div class="bd-content">
            <div class="bd-article-container">
              
              <div class="bd-header-article d-print-none">
<div class="header-article-items header-article__inner">
  
    <div class="header-article-items__start">
      
        <div class="header-article-item"><button class="sidebar-toggle primary-toggle btn btn-sm" title="Toggle primary sidebar" data-bs-placement="bottom" data-bs-toggle="tooltip">
  <span class="fa-solid fa-bars"></span>
</button></div>
      
    </div>
  
  
    <div class="header-article-items__end">
      
        <div class="header-article-item">

<div class="article-header-buttons">





<div class="dropdown dropdown-launch-buttons">
  <button class="btn dropdown-toggle" type="button" data-bs-toggle="dropdown" aria-expanded="false" aria-label="Launch interactive content">
    <i class="fas fa-rocket"></i>
  </button>
  <ul class="dropdown-menu">
      
      
      
      <li><a href="https://mybinder.org/v2/gh/iluvatar1/ML4Sci-lectures/master?urlpath=lab/tree/lectures/03-PCA/Intro-PCA.ipynb" target="_blank"
   class="btn btn-sm dropdown-item"
   title="Launch on Binder"
   data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  
    <img alt="Binder logo" src="../../_static/images/logo_binder.svg">
  </span>
<span class="btn__text-container">Binder</span>
</a>
</li>
      
      
      
      
      <li><a href="https://colab.research.google.com/github/iluvatar1/ML4Sci-lectures/blob/master/lectures/03-PCA/Intro-PCA.ipynb" target="_blank"
   class="btn btn-sm dropdown-item"
   title="Launch on Colab"
   data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  
    <img alt="Colab logo" src="../../_static/images/logo_colab.png">
  </span>
<span class="btn__text-container">Colab</span>
</a>
</li>
      
      
      
      
      <li>
<button onclick="initThebeSBT()"
  class="btn btn-sm btn-launch-thebe dropdown-item"
  title="Launch Thebe"
  data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-play"></i>
  </span>
<span class="btn__text-container">Live Code</span>
</button>
</li>
      
  </ul>
</div>






<div class="dropdown dropdown-source-buttons">
  <button class="btn dropdown-toggle" type="button" data-bs-toggle="dropdown" aria-expanded="false" aria-label="Source repositories">
    <i class="fab fa-github"></i>
  </button>
  <ul class="dropdown-menu">
      
      
      
      <li><a href="https://github.com/iluvatar1/ML4Sci-lectures" target="_blank"
   class="btn btn-sm btn-source-repository-button dropdown-item"
   title="Source repository"
   data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fab fa-github"></i>
  </span>
<span class="btn__text-container">Repository</span>
</a>
</li>
      
      
      
      
      <li><a href="https://github.com/iluvatar1/ML4Sci-lectures/issues/new?title=Issue%20on%20page%20%2Flectures/03-PCA/Intro-PCA.html&body=Your%20issue%20content%20here." target="_blank"
   class="btn btn-sm btn-source-issues-button dropdown-item"
   title="Open an issue"
   data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-lightbulb"></i>
  </span>
<span class="btn__text-container">Open issue</span>
</a>
</li>
      
  </ul>
</div>






<div class="dropdown dropdown-download-buttons">
  <button class="btn dropdown-toggle" type="button" data-bs-toggle="dropdown" aria-expanded="false" aria-label="Download this page">
    <i class="fas fa-download"></i>
  </button>
  <ul class="dropdown-menu">
      
      
      
      <li><a href="../../_sources/lectures/03-PCA/Intro-PCA.ipynb" target="_blank"
   class="btn btn-sm btn-download-source-button dropdown-item"
   title="Download source file"
   data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-file"></i>
  </span>
<span class="btn__text-container">.ipynb</span>
</a>
</li>
      
      
      
      
      <li>
<button onclick="window.print()"
  class="btn btn-sm btn-download-pdf-button dropdown-item"
  title="Print to PDF"
  data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-file-pdf"></i>
  </span>
<span class="btn__text-container">.pdf</span>
</button>
</li>
      
  </ul>
</div>




<button onclick="toggleFullScreen()"
  class="btn btn-sm btn-fullscreen-button"
  title="Fullscreen mode"
  data-bs-placement="bottom" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-expand"></i>
  </span>

</button>



<script>
document.write(`
  <button class="btn btn-sm nav-link pst-navbar-icon theme-switch-button" title="light/dark" aria-label="light/dark" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <i class="theme-switch fa-solid fa-sun fa-lg" data-mode="light"></i>
    <i class="theme-switch fa-solid fa-moon fa-lg" data-mode="dark"></i>
    <i class="theme-switch fa-solid fa-circle-half-stroke fa-lg" data-mode="auto"></i>
  </button>
`);
</script>


<script>
document.write(`
  <button class="btn btn-sm pst-navbar-icon search-button search-button__button" title="Search" aria-label="Search" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <i class="fa-solid fa-magnifying-glass fa-lg"></i>
  </button>
`);
</script>
<button class="sidebar-toggle secondary-toggle btn btn-sm" title="Toggle secondary sidebar" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <span class="fa-solid fa-list"></span>
</button>
</div></div>
      
    </div>
  
</div>
</div>
              
              

<div id="jb-print-docs-body" class="onlyprint">
    <h1>An Introduction to Principal Component Analysis (PCA)</h1>
    <!-- Table of contents -->
    <div id="print-main-content">
        <div id="jb-print-toc">
            
            <div>
                <h2> Contents </h2>
            </div>
            <nav aria-label="Page">
                <ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#dimensionality-reduction-methods">4.1. Dimensionality reduction methods</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#introduction-to-pca">4.2. Introduction to PCA</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#example-applications-to-basic-sciences">4.3. Example applications to basic sciences</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#biology-molecular-biology">4.3.1. Biology &amp; Molecular Biology</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#environmental-science-earth-science">4.3.2. Environmental Science &amp; Earth Science</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#chemistry">4.3.3. Chemistry</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#physics">4.3.4. Physics</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#mathematical-foundation">4.4. Mathematical Foundation</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#general-algorithm">4.5. General algorithm</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#step-1-data-centering">4.5.1. Step 1: Data Centering</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#step-2-covariance-matrix">4.5.2. Step 2: Covariance Matrix</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#step-3-eigendecomposition">4.5.3. Step 3: Eigendecomposition</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#step-4-variance-explained">4.5.4. Step 4: Variance Explained</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#pca-applied-to-the-previous-data-using-scikit-learn">4.6. PCA applied to the previous data using scikit learn</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#pca-applied-to-real-data">4.7. PCA applied to real data</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#interpretation-guidelines">4.8. Interpretation Guidelines</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#choosing-number-of-components">4.8.1. Choosing Number of Components</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#loading-interpretation">4.8.2. Loading Interpretation</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#common-pitfalls">4.9. Common Pitfalls</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#advanced-details">4.10. Advanced Details</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#pca-assumptions-and-limitations">4.10.1. PCA Assumptions and Limitations</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#robust-pca">4.10.2. Robust PCA</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#a-detailed-worked-example-the-iris-dataset">4.11. A detailed Worked Example - The Iris Dataset</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#setup-importing-libraries">4.11.1. Setup - Importing Libraries</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#loading-data">4.11.2. Loading data</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#step-1-standardize-the-data">4.11.3. Step 1: Standardize the Data</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#step-2-perform-pca">4.11.4. Step 2: Perform PCA</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#step-3-analyze-and-visualize-the-results">4.11.5. Step 3: Analyze and Visualize the Results</a><ul class="nav section-nav flex-column">
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#scree-plot">4.11.5.1. Scree Plot</a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#scores-plot">4.11.5.2. Scores Plot</a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#biplot">4.11.5.3. Biplot</a></li>
</ul>
</li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#application-exercises">4.12. Application Exercises</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#genomics-cancer-subtype-identification">4.12.1. Genomics - Cancer Subtype Identification</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#chemistry-classifying-olive-oils">4.12.2. Chemistry - Classifying Olive Oils</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#environmental-science-air-pollution-sources">4.12.3. Environmental Science - Air Pollution Sources</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#agriculture-crop-yield-analysis">4.12.4. Agriculture - Crop Yield Analysis</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#pharmaceutical-analysis">4.12.5. Pharmaceutical Analysis</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#protein-structure-analysis">4.12.6. Protein Structure Analysis</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#astronomical-data">4.12.7. Astronomical Data</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#summary">4.13. Summary</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#further-reading">4.14. Further Reading</a></li>
</ul>
            </nav>
        </div>
    </div>
</div>

              
                
<div id="searchbox"></div>
                <article class="bd-article">
                  
  <section class="tex2jax_ignore mathjax_ignore" id="an-introduction-to-principal-component-analysis-pca">
<h1><span class="section-number">4. </span>An Introduction to Principal Component Analysis (PCA)<a class="headerlink" href="#an-introduction-to-principal-component-analysis-pca" title="Link to this heading">#</a></h1>
<section id="dimensionality-reduction-methods">
<h2><span class="section-number">4.1. </span>Dimensionality reduction methods<a class="headerlink" href="#dimensionality-reduction-methods" title="Link to this heading">#</a></h2>
<p>The common goal of dimensionality reduction is to transform high-dimensional data into a lower-dimensional space while preserving as much of the meaningful structure and properties of the original data as possible. This is done to simplify datasets, reduce computational and storage costs, and make data easier to visualize and analyze. The following list show some of the more relevant methods used:</p>
<ul class="simple">
<li><p><strong>Principal Component Analysis (PCA)</strong> A linear technique that transforms the data into a new coordinate system of orthogonal axes called principal components. These components are ordered by the amount of variance they capture from the original data. Reference: <a class="reference external" href="https://scikit-learn.org/stable/modules/generated/sklearn.decomposition.PCA.html">https://scikit-learn.org/stable/modules/generated/sklearn.decomposition.PCA.html</a></p></li>
<li><p><strong>Linear Discriminant Analysis (LDA)</strong> A supervised linear technique that finds a linear combination of features that best separates two or more classes of objects. It aims to maximize the between-class variance while minimizing the within-class variance. Reference: <a class="reference external" href="https://scikit-learn.org/stable/modules/generated/sklearn.discriminant_analysis.LinearDiscriminantAnalysis.html">https://scikit-learn.org/stable/modules/generated/sklearn.discriminant_analysis.LinearDiscriminantAnalysis.html</a></p></li>
<li><p><strong>t-Distributed Stochastic Neighbor Embedding (t-SNE)</strong> A non-linear method primarily used for visualizing high-dimensional data. It models data points by their similarities in both high and low dimensions, preserving the local structure of the data. Reference: <a class="reference external" href="https://scikit-learn.org/stable/modules/generated/sklearn.manifold.TSNE.html">https://scikit-learn.org/stable/modules/generated/sklearn.manifold.TSNE.html</a></p></li>
<li><p><strong>Uniform Manifold Approximation and Projection (UMAP)</strong> A modern non-linear technique that constructs a graph representation of the data in high dimensions and then optimizes a similar graph in a lower dimension. It is known for its speed and ability to preserve both local and global data structure. Reference: <a class="reference external" href="https://umap-learn.readthedocs.io/en/latest/">https://umap-learn.readthedocs.io/en/latest/</a></p></li>
<li><p><strong>Kernel PCA</strong> An extension of PCA that uses kernel functions to perform non-linear dimensionality reduction. It implicitly maps the data to a higher-dimensional space where linear separation is possible, then applies PCA. Reference: <a class="reference external" href="https://scikit-learn.org/stable/modules/generated/sklearn.decomposition.KernelPCA.html">https://scikit-learn.org/stable/modules/generated/sklearn.decomposition.KernelPCA.html</a></p></li>
<li><p><strong>Autoencoders</strong> A type of neural network used for unsupervised learning. It consists of an “encoder” that compresses the input into a low-dimensional code and a “decoder” that reconstructs the input from this code. The compressed code is the lower-dimensional representation. Reference: <a class="reference external" href="https://blog.keras.io/building-autoencoders-in-keras.html">https://blog.keras.io/building-autoencoders-in-keras.html</a></p></li>
<li><p><strong>Feature Selection</strong> This is a category of methods that select a subset of the original features rather than creating new ones. Techniques include filter methods (e.g., chi-squared test), wrapper methods (e.g., recursive feature elimination), and embedded methods (e.g., LASSO regularization). Reference: <a class="reference external" href="https://scikit-learn.org/stable/modules/feature_selection.html">https://scikit-learn.org/stable/modules/feature_selection.html</a></p></li>
<li><p><strong>ISO MAP</strong> Preserves geodesic distances on data manifold using neighborhood graph + MDS.</p></li>
</ul>
<p>A nice visualization to compare: <a class="reference external" href="https://projector.tensorflow.org/">https://projector.tensorflow.org/</a></p>
<div class="pst-scrollable-table-container"><table class="table">
<thead>
<tr class="row-odd"><th class="head text-left"><p>Method</p></th>
<th class="head text-left"><p>Advantages</p></th>
<th class="head text-left"><p>Disadvantages</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td class="text-left"><p><strong>PCA</strong></p></td>
<td class="text-left"><p>Simple, fast, and removes correlated features. Effective for linear data.</p></td>
<td class="text-left"><p>Can be sensitive to data scaling. Performs poorly with non-linear data.</p></td>
</tr>
<tr class="row-odd"><td class="text-left"><p><strong>LDA</strong></p></td>
<td class="text-left"><p>Improves classification performance by focusing on class separability.</p></td>
<td class="text-left"><p>Supervised (requires labels). Assumes data is normally distributed.</p></td>
</tr>
<tr class="row-even"><td class="text-left"><p><strong>t-SNE</strong></p></td>
<td class="text-left"><p>Excellent at visualizing local clusters in high-dimensional data.</p></td>
<td class="text-left"><p>Computationally expensive. Can lose global structure information.</p></td>
</tr>
<tr class="row-odd"><td class="text-left"><p><strong>UMAP</strong></p></td>
<td class="text-left"><p>Very fast and scalable. Preserves both local and global data structure.</p></td>
<td class="text-left"><p>Can be sensitive to hyperparameters. Interpretation can be nuanced.</p></td>
</tr>
<tr class="row-even"><td class="text-left"><p><strong>Kernel PCA</strong></p></td>
<td class="text-left"><p>Can capture complex, non-linear relationships in the data.</p></td>
<td class="text-left"><p>More computationally intensive than PCA. Requires careful kernel selection.</p></td>
</tr>
<tr class="row-odd"><td class="text-left"><p><strong>Autoencoders</strong></p></td>
<td class="text-left"><p>Can learn highly complex, non-linear representations. Flexible for various data types.</p></td>
<td class="text-left"><p>Can be prone to overfitting. Requires a large amount of training data.</p></td>
</tr>
<tr class="row-even"><td class="text-left"><p><strong>Feature Selection</strong></p></td>
<td class="text-left"><p>Improves model interpretability by using original features. Reduces overfitting.</p></td>
<td class="text-left"><p>May discard features that are useful only in combination with others.</p></td>
</tr>
<tr class="row-odd"><td class="text-left"><p><strong>ISO Map</strong></p></td>
<td class="text-left"><p>aptures global nonlinear structure; preserves manifold geometry well if assumptions hold.</p></td>
<td class="text-left"><p>(Classic references like “Isomap: Learning Nonlinear Structure from High‑Dimensional Data”, etc.) — not explicitly in the small set I pulled but widely cited.</p></td>
</tr>
</tbody>
</table>
</div>
</section>
<section id="introduction-to-pca">
<h2><span class="section-number">4.2. </span>Introduction to PCA<a class="headerlink" href="#introduction-to-pca" title="Link to this heading">#</a></h2>
<p>What is Principal Component Analysis?</p>
<ul class="simple">
<li><p>Many scientific datasets have many correlated measurements: spectra (many wavelengths), climate grids (many locations), images (many pixels), recordings from many sensors or neurons.</p></li>
<li><p>PCA finds orthogonal directions (principal components) that explain the most variance, which helps visualization, noise reduction, compression, and discovering latent degrees of freedom.</p></li>
</ul>
<p>Principal Component Analysis (PCA) is a statistical technique used to reduce the dimensionality of data while preserving as much variance as possible. Think of it as finding the best camera angle to capture the most information about a 3D object in a 2D photograph.</p>
<p>Key Concepts:</p>
<ul class="simple">
<li><p>Dimensionality Reduction: Transform high-dimensional data to lower dimensions</p></li>
<li><p>Variance Preservation: Keep the most important patterns in the data</p></li>
<li><p>Data Visualization: Make complex datasets interpretable</p></li>
<li><p>Noise Reduction: Filter out less important variations</p></li>
</ul>
<div class="pst-scrollable-table-container"><table class="table">
<thead>
<tr class="row-odd"><th class="head"><p>Field</p></th>
<th class="head"><p>Application</p></th>
<th class="head"><p>Benefits</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p><strong>Spectroscopy</strong></p></td>
<td><p>Identify key wavelengths in complex spectra</p></td>
<td><p>Reduce noise, find characteristic peaks</p></td>
</tr>
<tr class="row-odd"><td><p><strong>Genomics</strong></p></td>
<td><p>Find patterns in gene expression data</p></td>
<td><p>Identify gene clusters, reduce computational complexity</p></td>
</tr>
<tr class="row-even"><td><p><strong>Climate Science</strong></p></td>
<td><p>Understand weather patterns from multiple variables</p></td>
<td><p>Visualize climate patterns, identify trends</p></td>
</tr>
<tr class="row-odd"><td><p><strong>Chemistry</strong></p></td>
<td><p>Analyze molecular properties and reactions</p></td>
<td><p>Optimize reaction conditions, understand structure-property relationships</p></td>
</tr>
<tr class="row-even"><td><p><strong>Physics</strong></p></td>
<td><p>Process sensor data from experiments</p></td>
<td><p>Extract signals from noise, identify fundamental modes</p></td>
</tr>
</tbody>
</table>
</div>
<div class="exercise admonition" id="lectures/03-PCA/Intro-PCA-exercise-0">

<p class="admonition-title"><span class="caption-number">Exercise 4.1 </span> (Understanding Dimensionality)</p>
<section id="exercise-content">
<p><strong>Scenario</strong>: A researcher has measurements of temperature, humidity, pressure, and wind speed from 1000 weather stations. They want to create a 2D map showing weather patterns.</p>
<p><strong>Question</strong>: How could PCA help in this situation?</p>
<p><strong>Tasks</strong>:</p>
<ol class="arabic simple">
<li><p>Identify the original dimensionality of the data</p></li>
<li><p>Explain what the principal components might represent</p></li>
<li><p>Discuss potential limitations of reducing to 2D</p></li>
</ol>
</section>
</div>
<p>Now let’s see an example.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span><span class="w"> </span><span class="nn">numpy</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">np</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">matplotlib.pyplot</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">plt</span>

<span class="c1"># Data generation</span>
<span class="n">np</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">seed</span><span class="p">(</span><span class="mi">0</span><span class="p">)</span>
<span class="n">n</span> <span class="o">=</span> <span class="mi">200</span>
<span class="n">x</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">normal</span><span class="p">(</span><span class="n">size</span><span class="o">=</span><span class="n">n</span><span class="p">)</span>
<span class="n">y</span> <span class="o">=</span> <span class="mf">2.0</span> <span class="o">*</span> <span class="n">x</span> <span class="o">+</span> <span class="mf">0.8</span> <span class="o">*</span> <span class="n">np</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">normal</span><span class="p">(</span><span class="n">size</span><span class="o">=</span><span class="n">n</span><span class="p">)</span>
<span class="n">X</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">vstack</span><span class="p">([</span><span class="n">x</span><span class="p">,</span><span class="n">y</span><span class="p">])</span><span class="o">.</span><span class="n">T</span>
<span class="n">Xc</span> <span class="o">=</span> <span class="n">X</span> <span class="o">-</span> <span class="n">X</span><span class="o">.</span><span class="n">mean</span><span class="p">(</span><span class="n">axis</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">plt</span><span class="o">.</span><span class="n">figure</span><span class="p">(</span><span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">6</span><span class="p">,</span><span class="mi">6</span><span class="p">))</span>
<span class="n">plt</span><span class="o">.</span><span class="n">scatter</span><span class="p">(</span><span class="n">Xc</span><span class="p">[:,</span><span class="mi">0</span><span class="p">],</span> <span class="n">Xc</span><span class="p">[:,</span><span class="mi">1</span><span class="p">],</span> <span class="n">s</span><span class="o">=</span><span class="mi">15</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">xlabel</span><span class="p">(</span><span class="s1">&#39;x (centered)&#39;</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">ylabel</span><span class="p">(</span><span class="s1">&#39;y (centered)&#39;</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">title</span><span class="p">(</span><span class="s1">&#39;2D data&#39;</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">axis</span><span class="p">(</span><span class="s1">&#39;equal&#39;</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">grid</span><span class="p">(</span><span class="kc">True</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="../../_images/917a65aaa9999697a40ae1fe21214abb02fdf4563d8350757663cc40a68308f2.png" src="../../_images/917a65aaa9999697a40ae1fe21214abb02fdf4563d8350757663cc40a68308f2.png" />
</div>
</div>
<p>This data is clearly positively correlated, as can be also be seen from a seaborn pair plot</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span><span class="w"> </span><span class="nn">seaborn</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">sns</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">pandas</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">pd</span>

<span class="c1"># Create a DataFrame with the centered data</span>
<span class="n">df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">Xc</span><span class="p">,</span> <span class="n">columns</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;x&#39;</span><span class="p">,</span> <span class="s1">&#39;y&#39;</span><span class="p">])</span>

<span class="c1"># Pairplot to visualize pairwise relationships and distributions</span>
<span class="n">sns</span><span class="o">.</span><span class="n">pairplot</span><span class="p">(</span><span class="n">df</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="../../_images/a673db959636894c86de7abf818cdff1f1914748c6b7b1781d143a875dcc25d2.png" src="../../_images/a673db959636894c86de7abf818cdff1f1914748c6b7b1781d143a875dcc25d2.png" />
</div>
</div>
<p>Some info can be drawn from the so-called covariance matrix:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Compute the covariance matrix, and its eigen values and vectors, ordered</span>
<span class="c1"># rowvar = False means vars are in columns in this case</span>
<span class="n">C</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">cov</span><span class="p">(</span><span class="n">Xc</span><span class="p">,</span> <span class="n">rowvar</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Plot the covariance matrix as a heatmap</span>
<span class="n">plt</span><span class="o">.</span><span class="n">figure</span><span class="p">(</span><span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">6</span><span class="p">,</span> <span class="mi">6</span><span class="p">))</span>
<span class="n">sns</span><span class="o">.</span><span class="n">heatmap</span><span class="p">(</span><span class="n">C</span><span class="p">,</span> <span class="n">annot</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">cmap</span><span class="o">=</span><span class="s2">&quot;coolwarm&quot;</span><span class="p">,</span> <span class="n">cbar</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">square</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">fmt</span><span class="o">=</span><span class="s2">&quot;.2f&quot;</span><span class="p">,</span> 
            <span class="n">xticklabels</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;x&quot;</span><span class="p">,</span> <span class="s2">&quot;y&quot;</span><span class="p">],</span> <span class="n">yticklabels</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;x&quot;</span><span class="p">,</span> <span class="s2">&quot;y&quot;</span><span class="p">])</span>
<span class="n">plt</span><span class="o">.</span><span class="n">title</span><span class="p">(</span><span class="s2">&quot;Covariance Matrix Heatmap&quot;</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="../../_images/fd97c57cb4065e323ef11df1fa4a019920de3d3755011750ef3cf64805509182.png" src="../../_images/fd97c57cb4065e323ef11df1fa4a019920de3d3755011750ef3cf64805509182.png" />
</div>
</div>
<p>Now, there is more info embedded in the correlation matrix. You can compute the correlation coefficient, defined as</p>
<div class="amsmath math notranslate nohighlight" id="equation-b37226e4-1c47-44f4-ae33-b71b1f0e65b0">
<span class="eqno">(4.1)<a class="headerlink" href="#equation-b37226e4-1c47-44f4-ae33-b71b1f0e65b0" title="Permalink to this equation">#</a></span>\[\begin{equation}
Corr(X, Y) = \frac{Cov(X, Y)}{\sqrt{Var(X)Var(Y)}}.
\end{equation}\]</div>
<p>Also, the condition number</p>
<div class="amsmath math notranslate nohighlight" id="equation-3fe8e930-71ce-4861-a08f-f3cf6047a6b4">
<span class="eqno">(4.2)<a class="headerlink" href="#equation-3fe8e930-71ce-4861-a08f-f3cf6047a6b4" title="Permalink to this equation">#</a></span>\[\begin{equation}
\kappa = \frac{\lambda_{\max}}{\lambda_{\min}}
\end{equation}\]</div>
<p>also helps to interpret the covariance matrix. For instance, if <span class="math notranslate nohighlight">\(\kappa\)</span> is very large, it suggest that the problem is ill-represented and there might be multi-linearities (high correlation among features).</p>
<p>Its determinant speaks about the volume in feature space, and small value might indicate that the data is redundant.</p>
<p>Furthermore, the Covariance matrix  eigen-values (explained variance) and eigen-vectors (principal components) are also meaningfull</p>
<div class="cell tag_hide-input tag_hide-output docutils container">
<details class="admonition hide above-input">
<summary aria-label="Toggle hidden content">
<p class="collapsed admonition-title">Show code cell source</p>
<p class="expanded admonition-title">Hide code cell source</p>
</summary>
<div class="cell_input above-output-prompt docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Visual demo: 2D correlated data and PC1</span>

<span class="n">eigvals</span><span class="p">,</span> <span class="n">eigvecs</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">linalg</span><span class="o">.</span><span class="n">eigh</span><span class="p">(</span><span class="n">C</span><span class="p">)</span>
<span class="n">order</span> <span class="o">=</span> <span class="n">eigvals</span><span class="o">.</span><span class="n">argsort</span><span class="p">()[::</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span>
<span class="n">eigvals</span> <span class="o">=</span> <span class="n">eigvals</span><span class="p">[</span><span class="n">order</span><span class="p">]</span>
<span class="n">eigvecs</span> <span class="o">=</span> <span class="n">eigvecs</span><span class="p">[:,</span> <span class="n">order</span><span class="p">]</span>

<span class="c1"># Visualize data and eigen directions</span>
<span class="n">pc1</span> <span class="o">=</span> <span class="n">eigvecs</span><span class="p">[:,</span><span class="mi">0</span><span class="p">]</span>
<span class="n">plt</span><span class="o">.</span><span class="n">figure</span><span class="p">(</span><span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">6</span><span class="p">,</span><span class="mi">6</span><span class="p">))</span>
<span class="n">plt</span><span class="o">.</span><span class="n">scatter</span><span class="p">(</span><span class="n">Xc</span><span class="p">[:,</span><span class="mi">0</span><span class="p">],</span> <span class="n">Xc</span><span class="p">[:,</span><span class="mi">1</span><span class="p">],</span> <span class="n">s</span><span class="o">=</span><span class="mi">15</span><span class="p">)</span>
<span class="n">origin</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">zeros</span><span class="p">(</span><span class="mi">2</span><span class="p">)</span>
<span class="n">c</span> <span class="o">=</span> <span class="p">{</span><span class="o">-</span><span class="mi">3</span><span class="p">:</span> <span class="s1">&#39;r&#39;</span><span class="p">,</span> <span class="mi">3</span> <span class="p">:</span> <span class="s1">&#39;k&#39;</span><span class="p">}</span>
<span class="k">for</span> <span class="n">s</span> <span class="ow">in</span> <span class="p">[</span><span class="o">-</span><span class="mi">3</span><span class="p">,</span><span class="mi">3</span><span class="p">]:</span>
    <span class="n">p</span> <span class="o">=</span> <span class="n">s</span> <span class="o">*</span> <span class="n">np</span><span class="o">.</span><span class="n">sqrt</span><span class="p">(</span><span class="n">eigvals</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span> <span class="o">*</span> <span class="n">pc1</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">plot</span><span class="p">([</span><span class="n">origin</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span> <span class="n">p</span><span class="p">[</span><span class="mi">0</span><span class="p">]],</span> <span class="p">[</span><span class="n">origin</span><span class="p">[</span><span class="mi">1</span><span class="p">],</span> <span class="n">p</span><span class="p">[</span><span class="mi">1</span><span class="p">]],</span> <span class="n">linewidth</span><span class="o">=</span><span class="mi">3</span><span class="p">,</span> <span class="n">c</span><span class="o">=</span><span class="n">c</span><span class="p">[</span><span class="n">s</span><span class="p">])</span>
<span class="n">plt</span><span class="o">.</span><span class="n">xlabel</span><span class="p">(</span><span class="s1">&#39;x (centered)&#39;</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">ylabel</span><span class="p">(</span><span class="s1">&#39;y (centered)&#39;</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">title</span><span class="p">(</span><span class="s1">&#39;2D correlated data and first principal component&#39;</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">axis</span><span class="p">(</span><span class="s1">&#39;equal&#39;</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">grid</span><span class="p">(</span><span class="kc">True</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>
</pre></div>
</div>
</div>
</details>
<details class="admonition hide below-input">
<summary aria-label="Toggle hidden content">
<p class="collapsed admonition-title">Show code cell output</p>
<p class="expanded admonition-title">Hide code cell output</p>
</summary>
<div class="cell_output docutils container">
<img alt="../../_images/c99efdaad782f40b6ad1d9f17e0301bb1e15006ab5b1d889c6338bb5e605917e.png" src="../../_images/c99efdaad782f40b6ad1d9f17e0301bb1e15006ab5b1d889c6338bb5e605917e.png" />
</div>
</details>
</div>
<div class="cell tag_hide-input docutils container">
<details class="admonition hide above-input">
<summary aria-label="Toggle hidden content">
<p class="collapsed admonition-title">Show code cell source</p>
<p class="expanded admonition-title">Hide code cell source</p>
</summary>
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span><span class="w"> </span><span class="nn">bokeh.plotting</span><span class="w"> </span><span class="kn">import</span> <span class="n">figure</span><span class="p">,</span> <span class="n">show</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">bokeh.io</span><span class="w"> </span><span class="kn">import</span> <span class="n">output_notebook</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">numpy</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">np</span>

<span class="c1"># Data generation</span>
<span class="n">np</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">seed</span><span class="p">(</span><span class="mi">0</span><span class="p">)</span>
<span class="n">n</span> <span class="o">=</span> <span class="mi">200</span>
<span class="n">x</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">normal</span><span class="p">(</span><span class="n">size</span><span class="o">=</span><span class="n">n</span><span class="p">)</span>
<span class="n">y</span> <span class="o">=</span> <span class="mf">2.0</span> <span class="o">*</span> <span class="n">x</span> <span class="o">+</span> <span class="mf">0.8</span> <span class="o">*</span> <span class="n">np</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">normal</span><span class="p">(</span><span class="n">size</span><span class="o">=</span><span class="n">n</span><span class="p">)</span>
<span class="n">X</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">vstack</span><span class="p">([</span><span class="n">x</span><span class="p">,</span> <span class="n">y</span><span class="p">])</span><span class="o">.</span><span class="n">T</span>
<span class="n">Xc</span> <span class="o">=</span> <span class="n">X</span> <span class="o">-</span> <span class="n">X</span><span class="o">.</span><span class="n">mean</span><span class="p">(</span><span class="n">axis</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>

<span class="c1"># Compute the covariance matrix, and its eigen values and vectors</span>
<span class="n">C</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">cov</span><span class="p">(</span><span class="n">Xc</span><span class="p">,</span> <span class="n">rowvar</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
<span class="n">eigvals</span><span class="p">,</span> <span class="n">eigvecs</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">linalg</span><span class="o">.</span><span class="n">eigh</span><span class="p">(</span><span class="n">C</span><span class="p">)</span>
<span class="n">order</span> <span class="o">=</span> <span class="n">eigvals</span><span class="o">.</span><span class="n">argsort</span><span class="p">()[::</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span>
<span class="n">eigvals</span> <span class="o">=</span> <span class="n">eigvals</span><span class="p">[</span><span class="n">order</span><span class="p">]</span>
<span class="n">eigvecs</span> <span class="o">=</span> <span class="n">eigvecs</span><span class="p">[:,</span> <span class="n">order</span><span class="p">]</span>


<span class="c1"># Check orthogonality: the dot product should be 0 for orthogonal vectors</span>
<span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">eigvecs</span><span class="p">[:,</span><span class="w"> </span><span class="mi">0</span><span class="p">]</span><span class="si">=}</span><span class="s2">   </span><span class="si">{</span><span class="n">eigvecs</span><span class="p">[:,</span><span class="w"> </span><span class="mi">1</span><span class="p">]</span><span class="si">=}</span><span class="s2">&quot;</span><span class="p">)</span>
<span class="n">dot_product</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">dot</span><span class="p">(</span><span class="n">eigvecs</span><span class="p">[:,</span> <span class="mi">0</span><span class="p">],</span> <span class="n">eigvecs</span><span class="p">[:,</span> <span class="mi">1</span><span class="p">])</span>
<span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Dot product of PC1 and PC2: </span><span class="si">{</span><span class="n">dot_product</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>

<span class="c1"># print the eigevals normalized by the total sum</span>
<span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Eigenvals proportion: </span><span class="si">{</span><span class="n">eigvals</span><span class="o">/</span><span class="n">np</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="n">eigvals</span><span class="p">)</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>

<span class="c1"># First principal component</span>
<span class="n">pc1</span> <span class="o">=</span> <span class="n">eigvecs</span><span class="p">[:,</span> <span class="mi">0</span><span class="p">]</span>

<span class="c1"># Bokeh plot setup</span>
<span class="n">output_notebook</span><span class="p">()</span>  <span class="c1"># Ensure the plot is rendered inline</span>

<span class="c1"># Create figure</span>
<span class="n">p</span> <span class="o">=</span> <span class="n">figure</span><span class="p">(</span><span class="n">width</span><span class="o">=</span><span class="mi">600</span><span class="p">,</span> <span class="n">height</span><span class="o">=</span><span class="mi">600</span><span class="p">,</span> <span class="n">title</span><span class="o">=</span><span class="s2">&quot;2D Correlated Data and PC1 and PC2&quot;</span><span class="p">,</span> <span class="n">match_aspect</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="n">p</span><span class="o">.</span><span class="n">scatter</span><span class="p">(</span><span class="n">Xc</span><span class="p">[:,</span> <span class="mi">0</span><span class="p">],</span> <span class="n">Xc</span><span class="p">[:,</span> <span class="mi">1</span><span class="p">],</span> <span class="n">size</span><span class="o">=</span><span class="mi">8</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="s2">&quot;blue&quot;</span><span class="p">,</span> <span class="n">alpha</span><span class="o">=</span><span class="mf">0.5</span><span class="p">)</span>

<span class="c1"># Plot the principal component lines</span>
<span class="n">origin</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">([</span><span class="mi">0</span><span class="p">,</span> <span class="mi">0</span><span class="p">])</span>
<span class="n">color</span> <span class="o">=</span> <span class="p">{</span><span class="o">-</span><span class="mi">3</span><span class="p">:</span><span class="s2">&quot;red&quot;</span><span class="p">,</span> <span class="mi">3</span><span class="p">:</span><span class="s2">&quot;black&quot;</span><span class="p">}</span>
<span class="k">for</span> <span class="n">s</span> <span class="ow">in</span> <span class="p">[</span><span class="o">-</span><span class="mi">3</span><span class="p">,</span> <span class="mi">3</span><span class="p">]:</span>
    <span class="n">p</span><span class="o">.</span><span class="n">line</span><span class="p">([</span><span class="n">origin</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span> <span class="n">s</span> <span class="o">*</span> <span class="n">np</span><span class="o">.</span><span class="n">sqrt</span><span class="p">(</span><span class="n">eigvals</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span> <span class="o">*</span> <span class="n">pc1</span><span class="p">[</span><span class="mi">0</span><span class="p">]],</span> 
           <span class="p">[</span><span class="n">origin</span><span class="p">[</span><span class="mi">1</span><span class="p">],</span> <span class="n">s</span> <span class="o">*</span> <span class="n">np</span><span class="o">.</span><span class="n">sqrt</span><span class="p">(</span><span class="n">eigvals</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span> <span class="o">*</span> <span class="n">pc1</span><span class="p">[</span><span class="mi">1</span><span class="p">]],</span> 
           <span class="n">line_width</span><span class="o">=</span><span class="mi">3</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="n">color</span><span class="p">[</span><span class="n">s</span><span class="p">],</span> <span class="n">alpha</span><span class="o">=</span><span class="mf">0.6</span><span class="p">)</span>

<span class="c1"># Plot the second principal component lines</span>
<span class="c1"># First principal component</span>
<span class="n">pc2</span> <span class="o">=</span> <span class="n">eigvecs</span><span class="p">[:,</span> <span class="mi">1</span><span class="p">]</span>
<span class="n">color</span> <span class="o">=</span> <span class="p">{</span><span class="o">-</span><span class="mi">3</span><span class="p">:</span><span class="s2">&quot;orange&quot;</span><span class="p">,</span> <span class="mi">3</span><span class="p">:</span><span class="s2">&quot;cyan&quot;</span><span class="p">}</span>
<span class="k">for</span> <span class="n">s</span> <span class="ow">in</span> <span class="p">[</span><span class="o">-</span><span class="mi">3</span><span class="p">,</span> <span class="mi">3</span><span class="p">]:</span>
    <span class="n">p</span><span class="o">.</span><span class="n">line</span><span class="p">([</span><span class="n">origin</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span> <span class="n">s</span> <span class="o">*</span> <span class="n">np</span><span class="o">.</span><span class="n">sqrt</span><span class="p">(</span><span class="n">eigvals</span><span class="p">[</span><span class="mi">1</span><span class="p">])</span> <span class="o">*</span> <span class="n">pc2</span><span class="p">[</span><span class="mi">0</span><span class="p">]],</span> 
           <span class="p">[</span><span class="n">origin</span><span class="p">[</span><span class="mi">1</span><span class="p">],</span> <span class="n">s</span> <span class="o">*</span> <span class="n">np</span><span class="o">.</span><span class="n">sqrt</span><span class="p">(</span><span class="n">eigvals</span><span class="p">[</span><span class="mi">1</span><span class="p">])</span> <span class="o">*</span> <span class="n">pc2</span><span class="p">[</span><span class="mi">1</span><span class="p">]],</span> 
           <span class="n">line_width</span><span class="o">=</span><span class="mi">3</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="n">color</span><span class="p">[</span><span class="n">s</span><span class="p">],</span> <span class="n">alpha</span><span class="o">=</span><span class="mf">0.9</span><span class="p">)</span>


<span class="c1"># Labels and formatting</span>
<span class="n">p</span><span class="o">.</span><span class="n">xaxis</span><span class="o">.</span><span class="n">axis_label</span> <span class="o">=</span> <span class="s2">&quot;x (centered)&quot;</span>
<span class="n">p</span><span class="o">.</span><span class="n">yaxis</span><span class="o">.</span><span class="n">axis_label</span> <span class="o">=</span> <span class="s2">&quot;y (centered)&quot;</span>
<span class="n">p</span><span class="o">.</span><span class="n">grid</span><span class="o">.</span><span class="n">grid_line_color</span> <span class="o">=</span> <span class="s2">&quot;white&quot;</span>
<span class="n">p</span><span class="o">.</span><span class="n">title</span><span class="o">.</span><span class="n">text_font_size</span> <span class="o">=</span> <span class="s1">&#39;32px&#39;</span>
<span class="n">p</span><span class="o">.</span><span class="n">grid</span><span class="o">.</span><span class="n">grid_line_color</span> <span class="o">=</span> <span class="s2">&quot;gray&quot;</span>  <span class="c1"># Grid line color</span>
<span class="n">p</span><span class="o">.</span><span class="n">grid</span><span class="o">.</span><span class="n">grid_line_alpha</span> <span class="o">=</span> <span class="mf">0.3</span>  <span class="c1"># Transparency of grid lines</span>
<span class="n">p</span><span class="o">.</span><span class="n">grid</span><span class="o">.</span><span class="n">grid_line_width</span> <span class="o">=</span> <span class="mi">1</span>  <span class="c1"># Width of grid lines</span>

<span class="c1"># Show plot</span>
<span class="n">show</span><span class="p">(</span><span class="n">p</span><span class="p">)</span>
</pre></div>
</div>
</div>
</details>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>eigvecs[:, 0]=array([0.40265359, 0.91535244])   eigvecs[:, 1]=array([-0.91535244,  0.40265359])
Dot product of PC1 and PC2: 0.0
Eigenvals proportion: [0.98342079 0.01657921]
</pre></div>
</div>
<div class="output text_html">    <style>
        .bk-notebook-logo {
            display: block;
            width: 20px;
            height: 20px;
            background-image: url(data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAABQAAAAUCAYAAACNiR0NAAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAALEgAACxIB0t1+/AAAABx0RVh0U29mdHdhcmUAQWRvYmUgRmlyZXdvcmtzIENTNui8sowAAAOkSURBVDiNjZRtaJVlGMd/1/08zzln5zjP1LWcU9N0NkN8m2CYjpgQYQXqSs0I84OLIC0hkEKoPtiH3gmKoiJDU7QpLgoLjLIQCpEsNJ1vqUOdO7ppbuec5+V+rj4ctwzd8IIbbi6u+8f1539dt3A78eXC7QizUF7gyV1fD1Yqg4JWz84yffhm0qkFqBogB9rM8tZdtwVsPUhWhGcFJngGeWrPzHm5oaMmkfEg1usvLFyc8jLRqDOMru7AyC8saQr7GG7f5fvDeH7Ej8CM66nIF+8yngt6HWaKh7k49Soy9nXurCi1o3qUbS3zWfrYeQDTB/Qj6kX6Ybhw4B+bOYoLKCC9H3Nu/leUTZ1JdRWkkn2ldcCamzrcf47KKXdAJllSlxAOkRgyHsGC/zRday5Qld9DyoM4/q/rUoy/CXh3jzOu3bHUVZeU+DEn8FInkPBFlu3+nW3Nw0mk6vCDiWg8CeJaxEwuHS3+z5RgY+YBR6V1Z1nxSOfoaPa4LASWxxdNp+VWTk7+4vzaou8v8PN+xo+KY2xsw6une2frhw05CTYOmQvsEhjhWjn0bmXPjpE1+kplmmkP3suftwTubK9Vq22qKmrBhpY4jvd5afdRA3wGjFAgcnTK2s4hY0/GPNIb0nErGMCRxWOOX64Z8RAC4oCXdklmEvcL8o0BfkNK4lUg9HTl+oPlQxdNo3Mg4Nv175e/1LDGzZen30MEjRUtmXSfiTVu1kK8W4txyV6BMKlbgk3lMwYCiusNy9fVfvvwMxv8Ynl6vxoByANLTWplvuj/nF9m2+PDtt1eiHPBr1oIfhCChQMBw6Aw0UulqTKZdfVvfG7VcfIqLG9bcldL/+pdWTLxLUy8Qq38heUIjh4XlzZxzQm19lLFlr8vdQ97rjZVOLf8nclzckbcD4wxXMidpX30sFd37Fv/GtwwhzhxGVAprjbg0gCAEeIgwCZyTV2Z1REEW8O4py0wsjeloKoMr6iCY6dP92H6Vw/oTyICIthibxjm/DfN9lVz8IqtqKYLUXfoKVMVQVVJOElGjrnnUt9T9wbgp8AyYKaGlqingHZU/uG2NTZSVqwHQTWkx9hxjkpWDaCg6Ckj5qebgBVbT3V3NNXMSiWSDdGV3hrtzla7J+duwPOToIg42ChPQOQjspnSlp1V+Gjdged7+8UN5CRAV7a5EdFNwCjEaBR27b3W890TE7g24NAP/mMDXRWrGoFPQI9ls/MWO2dWFAar/xcOIImbbpA3zgAAAABJRU5ErkJggg==);
        }
    </style>
    <div>
        <a href="https://bokeh.org" target="_blank" class="bk-notebook-logo"></a>
        <span id="e9d2d2d6-e1a2-4884-ac3a-a891180fd139">Loading BokehJS ...</span>
    </div>
</div><script type="application/javascript">'use strict';
(function(root) {
  function now() {
    return new Date();
  }

  const force = true;

  if (typeof root._bokeh_onload_callbacks === "undefined" || force === true) {
    root._bokeh_onload_callbacks = [];
    root._bokeh_is_loading = undefined;
  }

const JS_MIME_TYPE = 'application/javascript';
  const HTML_MIME_TYPE = 'text/html';
  const EXEC_MIME_TYPE = 'application/vnd.bokehjs_exec.v0+json';
  const CLASS_NAME = 'output_bokeh rendered_html';

  /**
   * Render data to the DOM node
   */
  function render(props, node) {
    const script = document.createElement("script");
    node.appendChild(script);
  }

  /**
   * Handle when an output is cleared or removed
   */
  function handleClearOutput(event, handle) {
    function drop(id) {
      const view = Bokeh.index.get_by_id(id)
      if (view != null) {
        view.model.document.clear()
        Bokeh.index.delete(view)
      }
    }

    const cell = handle.cell;

    const id = cell.output_area._bokeh_element_id;
    const server_id = cell.output_area._bokeh_server_id;

    // Clean up Bokeh references
    if (id != null) {
      drop(id)
    }

    if (server_id !== undefined) {
      // Clean up Bokeh references
      const cmd_clean = "from bokeh.io.state import curstate; print(curstate().uuid_to_server['" + server_id + "'].get_sessions()[0].document.roots[0]._id)";
      cell.notebook.kernel.execute(cmd_clean, {
        iopub: {
          output: function(msg) {
            const id = msg.content.text.trim()
            drop(id)
          }
        }
      });
      // Destroy server and session
      const cmd_destroy = "import bokeh.io.notebook as ion; ion.destroy_server('" + server_id + "')";
      cell.notebook.kernel.execute(cmd_destroy);
    }
  }

  /**
   * Handle when a new output is added
   */
  function handleAddOutput(event, handle) {
    const output_area = handle.output_area;
    const output = handle.output;

    // limit handleAddOutput to display_data with EXEC_MIME_TYPE content only
    if ((output.output_type != "display_data") || (!Object.prototype.hasOwnProperty.call(output.data, EXEC_MIME_TYPE))) {
      return
    }

    const toinsert = output_area.element.find("." + CLASS_NAME.split(' ')[0]);

    if (output.metadata[EXEC_MIME_TYPE]["id"] !== undefined) {
      toinsert[toinsert.length - 1].firstChild.textContent = output.data[JS_MIME_TYPE];
      // store reference to embed id on output_area
      output_area._bokeh_element_id = output.metadata[EXEC_MIME_TYPE]["id"];
    }
    if (output.metadata[EXEC_MIME_TYPE]["server_id"] !== undefined) {
      const bk_div = document.createElement("div");
      bk_div.innerHTML = output.data[HTML_MIME_TYPE];
      const script_attrs = bk_div.children[0].attributes;
      for (let i = 0; i < script_attrs.length; i++) {
        toinsert[toinsert.length - 1].firstChild.setAttribute(script_attrs[i].name, script_attrs[i].value);
        toinsert[toinsert.length - 1].firstChild.textContent = bk_div.children[0].textContent
      }
      // store reference to server id on output_area
      output_area._bokeh_server_id = output.metadata[EXEC_MIME_TYPE]["server_id"];
    }
  }

  function register_renderer(events, OutputArea) {

    function append_mime(data, metadata, element) {
      // create a DOM node to render to
      const toinsert = this.create_output_subarea(
        metadata,
        CLASS_NAME,
        EXEC_MIME_TYPE
      );
      this.keyboard_manager.register_events(toinsert);
      // Render to node
      const props = {data: data, metadata: metadata[EXEC_MIME_TYPE]};
      render(props, toinsert[toinsert.length - 1]);
      element.append(toinsert);
      return toinsert
    }

    /* Handle when an output is cleared or removed */
    events.on('clear_output.CodeCell', handleClearOutput);
    events.on('delete.Cell', handleClearOutput);

    /* Handle when a new output is added */
    events.on('output_added.OutputArea', handleAddOutput);

    /**
     * Register the mime type and append_mime function with output_area
     */
    OutputArea.prototype.register_mime_type(EXEC_MIME_TYPE, append_mime, {
      /* Is output safe? */
      safe: true,
      /* Index of renderer in `output_area.display_order` */
      index: 0
    });
  }

  // register the mime type if in Jupyter Notebook environment and previously unregistered
  if (root.Jupyter !== undefined) {
    const events = require('base/js/events');
    const OutputArea = require('notebook/js/outputarea').OutputArea;

    if (OutputArea.prototype.mime_types().indexOf(EXEC_MIME_TYPE) == -1) {
      register_renderer(events, OutputArea);
    }
  }
  if (typeof (root._bokeh_timeout) === "undefined" || force === true) {
    root._bokeh_timeout = Date.now() + 5000;
    root._bokeh_failed_load = false;
  }

  const NB_LOAD_WARNING = {'data': {'text/html':
     "<div style='background-color: #fdd'>\n"+
     "<p>\n"+
     "BokehJS does not appear to have successfully loaded. If loading BokehJS from CDN, this \n"+
     "may be due to a slow or bad network connection. Possible fixes:\n"+
     "</p>\n"+
     "<ul>\n"+
     "<li>re-rerun `output_notebook()` to attempt to load from CDN again, or</li>\n"+
     "<li>use INLINE resources instead, as so:</li>\n"+
     "</ul>\n"+
     "<code>\n"+
     "from bokeh.resources import INLINE\n"+
     "output_notebook(resources=INLINE)\n"+
     "</code>\n"+
     "</div>"}};

  function display_loaded(error = null) {
    const el = document.getElementById("e9d2d2d6-e1a2-4884-ac3a-a891180fd139");
    if (el != null) {
      const html = (() => {
        if (typeof root.Bokeh === "undefined") {
          if (error == null) {
            return "BokehJS is loading ...";
          } else {
            return "BokehJS failed to load.";
          }
        } else {
          const prefix = `BokehJS ${root.Bokeh.version}`;
          if (error == null) {
            return `${prefix} successfully loaded.`;
          } else {
            return `${prefix} <b>encountered errors</b> while loading and may not function as expected.`;
          }
        }
      })();
      el.innerHTML = html;

      if (error != null) {
        const wrapper = document.createElement("div");
        wrapper.style.overflow = "auto";
        wrapper.style.height = "5em";
        wrapper.style.resize = "vertical";
        const content = document.createElement("div");
        content.style.fontFamily = "monospace";
        content.style.whiteSpace = "pre-wrap";
        content.style.backgroundColor = "rgb(255, 221, 221)";
        content.textContent = error.stack ?? error.toString();
        wrapper.append(content);
        el.append(wrapper);
      }
    } else if (Date.now() < root._bokeh_timeout) {
      setTimeout(() => display_loaded(error), 100);
    }
  }

  function run_callbacks() {
    try {
      root._bokeh_onload_callbacks.forEach(function(callback) {
        if (callback != null)
          callback();
      });
    } finally {
      delete root._bokeh_onload_callbacks
    }
    console.debug("Bokeh: all callbacks have finished");
  }

  function load_libs(css_urls, js_urls, callback) {
    if (css_urls == null) css_urls = [];
    if (js_urls == null) js_urls = [];

    root._bokeh_onload_callbacks.push(callback);
    if (root._bokeh_is_loading > 0) {
      console.debug("Bokeh: BokehJS is being loaded, scheduling callback at", now());
      return null;
    }
    if (js_urls == null || js_urls.length === 0) {
      run_callbacks();
      return null;
    }
    console.debug("Bokeh: BokehJS not loaded, scheduling load and callback at", now());
    root._bokeh_is_loading = css_urls.length + js_urls.length;

    function on_load() {
      root._bokeh_is_loading--;
      if (root._bokeh_is_loading === 0) {
        console.debug("Bokeh: all BokehJS libraries/stylesheets loaded");
        run_callbacks()
      }
    }

    function on_error(url) {
      console.error("failed to load " + url);
    }

    for (let i = 0; i < css_urls.length; i++) {
      const url = css_urls[i];
      const element = document.createElement("link");
      element.onload = on_load;
      element.onerror = on_error.bind(null, url);
      element.rel = "stylesheet";
      element.type = "text/css";
      element.href = url;
      console.debug("Bokeh: injecting link tag for BokehJS stylesheet: ", url);
      document.body.appendChild(element);
    }

    for (let i = 0; i < js_urls.length; i++) {
      const url = js_urls[i];
      const element = document.createElement('script');
      element.onload = on_load;
      element.onerror = on_error.bind(null, url);
      element.async = false;
      element.src = url;
      console.debug("Bokeh: injecting script tag for BokehJS library: ", url);
      document.head.appendChild(element);
    }
  };

  function inject_raw_css(css) {
    const element = document.createElement("style");
    element.appendChild(document.createTextNode(css));
    document.body.appendChild(element);
  }

  const js_urls = ["https://cdn.bokeh.org/bokeh/release/bokeh-3.8.0.min.js", "https://cdn.bokeh.org/bokeh/release/bokeh-gl-3.8.0.min.js", "https://cdn.bokeh.org/bokeh/release/bokeh-widgets-3.8.0.min.js", "https://cdn.bokeh.org/bokeh/release/bokeh-tables-3.8.0.min.js", "https://cdn.bokeh.org/bokeh/release/bokeh-mathjax-3.8.0.min.js"];
  const css_urls = [];

  const inline_js = [    function(Bokeh) {
      Bokeh.set_log_level("info");
    },
function(Bokeh) {
    }
  ];

  function run_inline_js() {
    if (root.Bokeh !== undefined || force === true) {
      try {
            for (let i = 0; i < inline_js.length; i++) {
      inline_js[i].call(root, root.Bokeh);
    }

      } catch (error) {display_loaded(error);throw error;
      }if (force === true) {
        display_loaded();
      }} else if (Date.now() < root._bokeh_timeout) {
      setTimeout(run_inline_js, 100);
    } else if (!root._bokeh_failed_load) {
      console.log("Bokeh: BokehJS failed to load within specified timeout.");
      root._bokeh_failed_load = true;
    } else if (force !== true) {
      const cell = $(document.getElementById("e9d2d2d6-e1a2-4884-ac3a-a891180fd139")).parents('.cell').data().cell;
      cell.output_area.append_execute_result(NB_LOAD_WARNING)
    }
  }

  if (root._bokeh_is_loading === 0) {
    console.debug("Bokeh: BokehJS loaded, going straight to plotting");
    run_inline_js();
  } else {
    load_libs(css_urls, js_urls, function() {
      console.debug("Bokeh: BokehJS plotting callback run at", now());
      run_inline_js();
    });
  }
}(window));</script><div class="output text_html">
  <div id="e09fafd1-825f-4475-9fd7-641d44443cf7" data-root-id="p1008" style="display: contents;"></div>
</div><script type="application/javascript">(function(root) {
  function embed_document(root) {
  const docs_json = {"f73174ac-ad4c-40f5-8009-d07230c291a2":{"version":"3.8.0","title":"Bokeh Application","config":{"type":"object","name":"DocumentConfig","id":"p1088","attributes":{"notifications":{"type":"object","name":"Notifications","id":"p1089"}}},"roots":[{"type":"object","name":"Figure","id":"p1008","attributes":{"x_range":{"type":"object","name":"DataRange1d","id":"p1009"},"y_range":{"type":"object","name":"DataRange1d","id":"p1010"},"x_scale":{"type":"object","name":"LinearScale","id":"p1018"},"y_scale":{"type":"object","name":"LinearScale","id":"p1019"},"title":{"type":"object","name":"Title","id":"p1011","attributes":{"text":"2D Correlated Data and PC1 and PC2","text_font_size":"32px"}},"renderers":[{"type":"object","name":"GlyphRenderer","id":"p1049","attributes":{"data_source":{"type":"object","name":"ColumnDataSource","id":"p1043","attributes":{"selected":{"type":"object","name":"Selection","id":"p1044","attributes":{"indices":[],"line_indices":[]}},"selection_policy":{"type":"object","name":"UnionRenderers","id":"p1045"},"data":{"type":"map","entries":[["x",{"type":"ndarray","array":{"type":"bytes","data":"H4sIAAEAAAAC/w3TiSMUiBoAcEnylKxyrCNquqwjOVIK31RWVLYVsqUc6xUtw6KUo9pdV3moeGW3hg4hJGUpg+kbcmwtGcfsmCw5hjFjGOYyxgy7v//hp6V9kmtssAiS9Gx68HoGJHusdOGvnQZV7wdsszMriAevbo7VQQWw0kctQ1pnsV53w1snAh+EN7wL3eI6kBs4Sh/ybEONHeZtsVQGhNSfimXVVUNQK+nAAYIU2OExNWl6E7DT1wY/KsuAFDdmcWmICel5JefNuulQ1E/94VirFCh78o+YVHTjMdO08RXLH8FsdWFFyW/TmG4ki0GlGm3Ps+Px/yseBRdtn1W3/8sFjVdJs6/VeBgzfjl/f/4KYrGcvJgRJkOL/cXpPzb9hol9fxIKU+lo0/7hIj10HuJh8NiaYClse7HYEpNSCwyOd5haeS8sbN1ILaqcQY3jwyfHL6jQiizXJSs7P2Ho/3Vbjn6uhUoNP5eeLCFkPYwiXzcSwsnQ0rvNPoPok7u2kzLKxKmtIyknuHMoiVzXyvw0j7JGl5V+0Qp00HTcO71hCc5ZvbWS5Y3i93xR4n3PYVR/fnzXhLUECQ88grzIE0A+sjt/QSzHl5zmXL8dPehNuJ57iz+DzqKUDEenPogysDH8oWAU3X3cC7zVxfhWffXSlkAKsjtecZWf/wKiSUlR5VIMvv5qlNVZ1AnmNtozh+5MYL88bkxIG8DerSsGXr3noFrQW9KbrAG8578j2vYIH9fFGmV5NSlwW1icFtmtAcYEN7PziENYud2eQ92ziJZU3jGSLgsGbC+XJXsJkL4qbsu9iDyUJ/OLu7TH4clBvqK8oRIMa5otb345B2U59c1H2WL84kik0sKRAWvCbQV/75zEsvjWL64QZnDNdzl0xgc2TnictXNP7Ef6NymlNXY5qBcRYs9pE6HXySRFRQ4PXG+aVxQcYoG0ZqUX5/UCftKjXI2qlMJiyky/ib0S2JHbpvsoczB26snDKi067n7Wnkl0E2JYqdVlywQBlOhU9cf6D6FHlH9aVaQQ7m3e021+twm21OmT6hf5wI/XYU669YCdrS7xlD8bwPGdvtnMCxS96R1P+HERTCWegy7PnwMzQj/q3C4G1CpS7wqWFCAqe9M/VC1F+8dXTAPiJWgheP7VeRofqkVN+rc5IrRa/3vyQ6kSzG/pczNUP+OdGxHRa3fxsPSpdffUYyVkZ+Q7U7Ol8C70Xvz6fz+Rm4c9Iig80H38JVG7bxp9FSb7rElKeKpNbnv3koH7yqdFPeEcKCfXuis1+XBC/YXzPK8DF+4nVBeFjMAf114ldbtMwQU/a2FOdi+0M419fqUI8bRRb+d6w06wyY8NJumLYe/XkZc6rk2iecEqRve+DtSlHXWqtB3G7x2fpREuKsB12cEx03UMcn5WLV6XyYANmVamWpM8rMtQGT64MAihrk8tNVgczM89q+ap+hQDrTZKsGACjW8xT8cEj8G5wpifbuwcBi3f5cu3h7qx2I4SFPyxD6oXDIwfPxRilk6CtGhJhue2pznY3WVC6yZaLj+lDnxass4e1hqFkZTxqovXVYm/74uUuounYPuND/+hJgrwyR8GUazGWVjYFCAL2yXFEPVhc3vNEeR5aIY3djWhGWm3WthhOQTolG6bMeRhar2et94bPpaxOqRVla1o/wt93OQnDhJDBd8R2bPQ4HDJi3VBiOrhqZuTromQ5PrXobTDw2hK28xodxjF6TFLKbdRCVzumbImEz48ootrv9a8AwFXBr5lUMWY/mhDpBltEpYro1m+W+aQ7eA/28daQNXEq8l1n+dAI1W2LL/aBb6hkcw2uylo9AhUTajrglqrKeGB7VywsCDUGhpx8DzhwqHaF3P4PijNiZI5BpYsZ7co+RSW+Fu1v/SbxFvhDYuF/cOYk7RX53RABSZ42rs+sR3AP/mKeSDMY6C1Rk3E4ASy2B0mTadVaZQ4+2+WG0cgWjQCiS1yPLF/hKPyUYivxQejlm3zsGJrouF5ARcVD2Srqdnz4HqlvEXZLUHTzML6TcbvoUF+W/FodyOeIR5mSXpEOHv/pdMd0d9g+qD8QK9GF7qMyFeSSROQuVHLC/QmQbdMskmPqkJ0zlExkIEY/gGK6OjjQAYAAA=="},"shape":[200],"dtype":"float64","order":"little"}],["y",{"type":"ndarray","array":{"type":"bytes","data":"H4sIAAEAAAAC/wFABr/5E0bW/fqLCUCzT63ciTjiP8o5CvPhYQZAHA+IPcndE0Cuv/CdzNQQQFTWnIbcTArAjNCm1/11/T9LvfoKcNftv4Bm6lLmbpa/bNbSaCqB5j+X3P3WHULvP78mb2K5+whAqw/C3tHkAEDmgPuKK6jFv6BUupUfv7c/trjyrI0x0z901zm0JnsHQGIne0hivcK/E/eQQkQnA0Co59I0pID8vxauIw/9oxfAiU98fZ627z9w4trc7Rj1P8qOS6yEN/K/1+7FPcMjCkBDvTPQJC4HwO4PUR8wssY/i8suVCkwzb83aMIa6WIEQJtdKtz4qwVAGdXio+jM67+02ZdJ55/UP0GfMU+B/gHAU9aweWVYDcCUgqiHGJD6v7zQ5ohwvuw/ccfppx/vDEDZCTGFGLPmP0J1luGMRt6/UPHjPa88ur9TotLQ9SsFwLy6zXr+kwnAdQwcAE53DMA1GITWw/0MQIC3iLUQ5fS/iaBfBQcNAsCOOV1EXPj5v6O+xbW8CANAANoGH9pXD8AUk91WXjX6v+ZbT1eonfa/MU4hj16F0T8XGvCkkVHuvz/VeX7EPwXAel3+VPpG3T/uQmgcB/j1P4IVLOOYKN+/1ji4/ydR4b/H6rlIYZgEwC1XkRyfs9G/auPJee+uAsBGmzLBnZ/yv/b6ZslbJAHAMOwNtLBFDMCy4Gdjv7zzv9btkfVwJOa/H5JX6qgLB8BEhiec3Z7uPw5Hu0Zc0gDAaDG/kjY6wj++oSRR4s77P7ap4FsG/v+/BUA2a9duDkBQeqOYbZMBwMADz7v+Jc8/zME2EHaM+79CQHAcGS32v17JmIbOo/S/OFGGLudLAsCitIOknZf7PyvlkCVGqgPAnzOTm4agBEAKK6+1TqbVP3JW5sjtH/6/+sS151BUCUCi8y/QMPUQQLTSHxwOtvc/veBhp6E64j8y9bNVoRD6v3enmshE4whAxma2H12S9b/cmUpKTyoAQFvpY/gewwFAyMaCCb8N8T+NnkPB5xDiP0VXFZcYRAJA9gx0E0Garj/fDhOPU/0PQJi2LPC+6rq/XHrbbEb48D/Fn4mrD3QFQNg1WYjmiPa/C/qyizdmBcAQuJupEa/1P5wJTjMLpP2/ElG4xITVC0B+E3zA4vH+vxfdT1C/5ATAlvdcRCtyEUA9IaVzCFoGQKN1cCHBtQxAC/7yUjkUBUCft3Z1e04FwPzuTu7FkQlACSnqRyhE7L9u1HYieUH6P4RnhKFHJP0/50c1y6uG4r+k/2gOLjvyP7TNFqCtgfs/3M/cIg+iwT87ewsi3RsHwOjFzRaQ2ug/0POoWPto/j/zFegK39UCwCKCBbUX1eK/JrEvkteS8L9VaksQpdsVQJfonDa2uec/ZCzJGWJ7+D8oa2KQX2jvv7iQwkDHgrY/09NOdISf6L+o1reVAojtv+U/93zsgQvAHBlCXrXG/D9q82nIw57Sv+Qw4orqYri/VMaluGRKyj+NVMeKh7vsv5nr03hBVgHA+KsXba+B3j97eG9rwGfQv36oCAbfmdA//vKTFx9/EUDcakgPFgH6P0xai+kkQQHASETw9FSPAkCLmAdN0aoBwPVG9NHy2eW/QxdBTs4w6b9UIMG6duUBQKplKiae/Nu/cHKIcGXzAcAMqkhhwkLov7xjXtl3iPy/jdeMdTip5z9s4tf2OKcEwEl13zGcvgXAxMOnsgLN2r9Io9u1ZILev4OTz8DUlA5AS8d6iNvWBEBM1z08LxfaP78YQ7TJBQTAbigVFu5y/D+NpuyN9YkBwDD+kqkDjwvA7DZfEc76AEDCWZyfLbXTv9+oCV2aNgBAJM4VD18lyb+1DtTgy8cCQJ5vS2YMqPi/4Fy3ZeCLAMCIF7jeZ4D7PwSwzKXwGvK/c6lwbWpmBcBYgNgIINvxvymME5TgY+Y/RZEImdjYAMCtfBOZ/dUDwMHrfeR8DAnA+r22wYU3FcDsvA1LD9/zP+yJeuGMRhLA1QYU5+xKCMBaSopX/szxv9S5atZ/msm/L2hzT4xsCUAEI2TQzV4BwFwGzmdkQtQ/tjuvUmel2b9TCiYZp1kNwEaI1wrwweg/FxNl5fi7779N222MKoYBQB37/6DxKQRAftF3j1/XFUBKUvZ0h4YKQPZy0O1ABgAA"},"shape":[200],"dtype":"float64","order":"little"}]]}}},"view":{"type":"object","name":"CDSView","id":"p1050","attributes":{"filter":{"type":"object","name":"AllIndices","id":"p1051"}}},"glyph":{"type":"object","name":"Scatter","id":"p1046","attributes":{"x":{"type":"field","field":"x"},"y":{"type":"field","field":"y"},"size":{"type":"value","value":8},"line_color":{"type":"value","value":"blue"},"line_alpha":{"type":"value","value":0.5},"fill_color":{"type":"value","value":"blue"},"fill_alpha":{"type":"value","value":0.5},"hatch_color":{"type":"value","value":"blue"},"hatch_alpha":{"type":"value","value":0.5}}},"nonselection_glyph":{"type":"object","name":"Scatter","id":"p1047","attributes":{"x":{"type":"field","field":"x"},"y":{"type":"field","field":"y"},"size":{"type":"value","value":8},"line_color":{"type":"value","value":"blue"},"line_alpha":{"type":"value","value":0.1},"fill_color":{"type":"value","value":"blue"},"fill_alpha":{"type":"value","value":0.1},"hatch_color":{"type":"value","value":"blue"},"hatch_alpha":{"type":"value","value":0.1}}},"muted_glyph":{"type":"object","name":"Scatter","id":"p1048","attributes":{"x":{"type":"field","field":"x"},"y":{"type":"field","field":"y"},"size":{"type":"value","value":8},"line_color":{"type":"value","value":"blue"},"line_alpha":{"type":"value","value":0.2},"fill_color":{"type":"value","value":"blue"},"fill_alpha":{"type":"value","value":0.2},"hatch_color":{"type":"value","value":"blue"},"hatch_alpha":{"type":"value","value":0.2}}}}},{"type":"object","name":"GlyphRenderer","id":"p1058","attributes":{"data_source":{"type":"object","name":"ColumnDataSource","id":"p1052","attributes":{"selected":{"type":"object","name":"Selection","id":"p1053","attributes":{"indices":[],"line_indices":[]}},"selection_policy":{"type":"object","name":"UnionRenderers","id":"p1054"},"data":{"type":"map","entries":[["x",[0,-2.9462913903743186]],["y",[0,-6.697804399753717]]]}}},"view":{"type":"object","name":"CDSView","id":"p1059","attributes":{"filter":{"type":"object","name":"AllIndices","id":"p1060"}}},"glyph":{"type":"object","name":"Line","id":"p1055","attributes":{"x":{"type":"field","field":"x"},"y":{"type":"field","field":"y"},"line_color":"red","line_alpha":0.6,"line_width":3}},"nonselection_glyph":{"type":"object","name":"Line","id":"p1056","attributes":{"x":{"type":"field","field":"x"},"y":{"type":"field","field":"y"},"line_color":"red","line_alpha":0.1,"line_width":3}},"muted_glyph":{"type":"object","name":"Line","id":"p1057","attributes":{"x":{"type":"field","field":"x"},"y":{"type":"field","field":"y"},"line_color":"red","line_alpha":0.2,"line_width":3}}}},{"type":"object","name":"GlyphRenderer","id":"p1067","attributes":{"data_source":{"type":"object","name":"ColumnDataSource","id":"p1061","attributes":{"selected":{"type":"object","name":"Selection","id":"p1062","attributes":{"indices":[],"line_indices":[]}},"selection_policy":{"type":"object","name":"UnionRenderers","id":"p1063"},"data":{"type":"map","entries":[["x",[0,2.9462913903743186]],["y",[0,6.697804399753717]]]}}},"view":{"type":"object","name":"CDSView","id":"p1068","attributes":{"filter":{"type":"object","name":"AllIndices","id":"p1069"}}},"glyph":{"type":"object","name":"Line","id":"p1064","attributes":{"x":{"type":"field","field":"x"},"y":{"type":"field","field":"y"},"line_alpha":0.6,"line_width":3}},"nonselection_glyph":{"type":"object","name":"Line","id":"p1065","attributes":{"x":{"type":"field","field":"x"},"y":{"type":"field","field":"y"},"line_alpha":0.1,"line_width":3}},"muted_glyph":{"type":"object","name":"Line","id":"p1066","attributes":{"x":{"type":"field","field":"x"},"y":{"type":"field","field":"y"},"line_alpha":0.2,"line_width":3}}}},{"type":"object","name":"GlyphRenderer","id":"p1076","attributes":{"data_source":{"type":"object","name":"ColumnDataSource","id":"p1070","attributes":{"selected":{"type":"object","name":"Selection","id":"p1071","attributes":{"indices":[],"line_indices":[]}},"selection_policy":{"type":"object","name":"UnionRenderers","id":"p1072"},"data":{"type":"map","entries":[["x",[0,0.8696504579898086]],["y",[0,-0.3825497885702174]]]}}},"view":{"type":"object","name":"CDSView","id":"p1077","attributes":{"filter":{"type":"object","name":"AllIndices","id":"p1078"}}},"glyph":{"type":"object","name":"Line","id":"p1073","attributes":{"x":{"type":"field","field":"x"},"y":{"type":"field","field":"y"},"line_color":"orange","line_alpha":0.9,"line_width":3}},"nonselection_glyph":{"type":"object","name":"Line","id":"p1074","attributes":{"x":{"type":"field","field":"x"},"y":{"type":"field","field":"y"},"line_color":"orange","line_alpha":0.1,"line_width":3}},"muted_glyph":{"type":"object","name":"Line","id":"p1075","attributes":{"x":{"type":"field","field":"x"},"y":{"type":"field","field":"y"},"line_color":"orange","line_alpha":0.2,"line_width":3}}}},{"type":"object","name":"GlyphRenderer","id":"p1085","attributes":{"data_source":{"type":"object","name":"ColumnDataSource","id":"p1079","attributes":{"selected":{"type":"object","name":"Selection","id":"p1080","attributes":{"indices":[],"line_indices":[]}},"selection_policy":{"type":"object","name":"UnionRenderers","id":"p1081"},"data":{"type":"map","entries":[["x",[0,-0.8696504579898086]],["y",[0,0.3825497885702174]]]}}},"view":{"type":"object","name":"CDSView","id":"p1086","attributes":{"filter":{"type":"object","name":"AllIndices","id":"p1087"}}},"glyph":{"type":"object","name":"Line","id":"p1082","attributes":{"x":{"type":"field","field":"x"},"y":{"type":"field","field":"y"},"line_color":"cyan","line_alpha":0.9,"line_width":3}},"nonselection_glyph":{"type":"object","name":"Line","id":"p1083","attributes":{"x":{"type":"field","field":"x"},"y":{"type":"field","field":"y"},"line_color":"cyan","line_alpha":0.1,"line_width":3}},"muted_glyph":{"type":"object","name":"Line","id":"p1084","attributes":{"x":{"type":"field","field":"x"},"y":{"type":"field","field":"y"},"line_color":"cyan","line_alpha":0.2,"line_width":3}}}}],"toolbar":{"type":"object","name":"Toolbar","id":"p1017","attributes":{"tools":[{"type":"object","name":"PanTool","id":"p1030"},{"type":"object","name":"WheelZoomTool","id":"p1031","attributes":{"renderers":"auto"}},{"type":"object","name":"BoxZoomTool","id":"p1032","attributes":{"dimensions":"both","overlay":{"type":"object","name":"BoxAnnotation","id":"p1033","attributes":{"syncable":false,"line_color":"black","line_alpha":1.0,"line_width":2,"line_dash":[4,4],"fill_color":"lightgrey","fill_alpha":0.5,"level":"overlay","visible":false,"left":{"type":"number","value":"nan"},"right":{"type":"number","value":"nan"},"top":{"type":"number","value":"nan"},"bottom":{"type":"number","value":"nan"},"left_units":"canvas","right_units":"canvas","top_units":"canvas","bottom_units":"canvas","handles":{"type":"object","name":"BoxInteractionHandles","id":"p1039","attributes":{"all":{"type":"object","name":"AreaVisuals","id":"p1038","attributes":{"fill_color":"white","hover_fill_color":"lightgray"}}}}}}}},{"type":"object","name":"SaveTool","id":"p1040"},{"type":"object","name":"ResetTool","id":"p1041"},{"type":"object","name":"HelpTool","id":"p1042"}]}},"left":[{"type":"object","name":"LinearAxis","id":"p1025","attributes":{"ticker":{"type":"object","name":"BasicTicker","id":"p1026","attributes":{"mantissas":[1,2,5]}},"formatter":{"type":"object","name":"BasicTickFormatter","id":"p1027"},"axis_label":"y (centered)","major_label_policy":{"type":"object","name":"AllLabels","id":"p1028"}}}],"below":[{"type":"object","name":"LinearAxis","id":"p1020","attributes":{"ticker":{"type":"object","name":"BasicTicker","id":"p1021","attributes":{"mantissas":[1,2,5]}},"formatter":{"type":"object","name":"BasicTickFormatter","id":"p1022"},"axis_label":"x (centered)","major_label_policy":{"type":"object","name":"AllLabels","id":"p1023"}}}],"center":[{"type":"object","name":"Grid","id":"p1024","attributes":{"axis":{"id":"p1020"},"grid_line_color":"gray","grid_line_alpha":0.3}},{"type":"object","name":"Grid","id":"p1029","attributes":{"dimension":1,"axis":{"id":"p1025"},"grid_line_color":"gray","grid_line_alpha":0.3}}],"match_aspect":true}}]}};
  const render_items = [{"docid":"f73174ac-ad4c-40f5-8009-d07230c291a2","roots":{"p1008":"e09fafd1-825f-4475-9fd7-641d44443cf7"},"root_ids":["p1008"]}];
  void root.Bokeh.embed.embed_items_notebook(docs_json, render_items);
  }
  if (root.Bokeh !== undefined) {
    embed_document(root);
  } else {
    let attempts = 0;
    const timer = setInterval(function(root) {
      if (root.Bokeh !== undefined) {
        clearInterval(timer);
        embed_document(root);
      } else {
        attempts++;
        if (attempts > 100) {
          clearInterval(timer);
          console.log("Bokeh: ERROR: Unable to run BokehJS code because BokehJS library is missing");
        }
      }
    }, 10, root)
  }
})(window);</script></div>
</div>
<p>This is the key for PCA method: finding the eigen values and eigen vectors give some special directions where the data varies the most and allow to reduce the dimensionality of the data (in this case, using the first principal component will be enough)</p>
</section>
<section id="example-applications-to-basic-sciences">
<h2><span class="section-number">4.3. </span>Example applications to basic sciences<a class="headerlink" href="#example-applications-to-basic-sciences" title="Link to this heading">#</a></h2>
<section id="biology-molecular-biology">
<h3><span class="section-number">4.3.1. </span>Biology &amp; Molecular Biology<a class="headerlink" href="#biology-molecular-biology" title="Link to this heading">#</a></h3>
<ul class="simple">
<li><p>Higa, G. S., de Oliveira, L. G., Luns, K. F., de Jesus, R. E., Ferreira, R. C., da Silva, J. A., Cunha, P. H., &amp; Pires, D. S. (2024). Application of Principal Component Analysis as a Prediction Model for Feline Sporotrichosis. Animals, 14(12), 1696.  <a class="reference external" href="https://www.mdpi.com/2306-7381/12/1/32">https://www.mdpi.com/2306-7381/12/1/32</a></p></li>
<li><p>Sfriso, P., &amp; Crave, A. (2022). Principal Component Analysis and Related Methods for Investigating the Dynamics of Biological Macromolecules. J - Multidisciplinary Scientific Journal, 5(2), 298–317. <a class="reference external" href="https://www.mdpi.com/2571-8800/5/2/21">https://www.mdpi.com/2571-8800/5/2/21</a></p></li>
</ul>
</section>
<section id="environmental-science-earth-science">
<h3><span class="section-number">4.3.2. </span>Environmental Science &amp; Earth Science<a class="headerlink" href="#environmental-science-earth-science" title="Link to this heading">#</a></h3>
<ul class="simple">
<li><p>Huang, J., Lu, D., Lin, W., &amp; Yang, Q. (2024). Enhancing slope stability prediction through integrated PCA-SSA-SVM modeling: a case study of LongLian expressway. Frontiers in Earth Science, 12. DOI: <a class="reference external" href="https://doi.org/10.3389/feart.2024.1429601">https://doi.org/10.3389/feart.2024.1429601</a></p></li>
<li><p>Kaur, L., &amp; Godara, P. (2025). Application of Principal Component Analysis (PCA) in groundwater quality evaluation: A case study of arid region. Energy &amp; Environment Management, 1(21). <a class="reference external" href="https://ojs.awpublishing.org/eem/article/view/17">https://ojs.awpublishing.org/eem/article/view/17</a></p></li>
<li><p>Kahangwa, C. (2022). Application of Principal Component Analysis, Cluster Analysis, Pollution Index and Geoaccumulation Index in Pollution Assessment with Heavy Metals from Gold Mining Operations, Tanzania. Journal of Geoscience and Environment Protection, 10(4), 303-317. <a class="reference external" href="https://www.scirp.org/journal/paperinformation?paperid=116892">https://www.scirp.org/journal/paperinformation?paperid=116892</a></p></li>
</ul>
</section>
<section id="chemistry">
<h3><span class="section-number">4.3.3. </span>Chemistry<a class="headerlink" href="#chemistry" title="Link to this heading">#</a></h3>
<ul class="simple">
<li><p>Moreira, M., Hillenkamp, M., Divitini, G., Tizei, L. H. G., Ducati, C., Cotta, M. A., Rodrigues, V., &amp; Ugarte, D. (2025). Improving Quantitative EDS Chemical Analysis of Alloy Nanoparticles by PCA Denoising: Part I, Reducing Reconstruction Bias. Microscopy and Microanalysis. DOI: <a class="reference external" href="https://academic.oup.com/mam/article-abstract/28/2/338/6911396?redirectedFrom=fulltext">https://academic.oup.com/mam/article-abstract/28/2/338/6911396?redirectedFrom=fulltext</a></p></li>
</ul>
</section>
<section id="physics">
<h3><span class="section-number">4.3.4. </span>Physics<a class="headerlink" href="#physics" title="Link to this heading">#</a></h3>
<ul class="simple">
<li><p>Yao, Z., Yang, J., &amp; Lin, H.-Q. (2025). Principal component analysis for percolation with and without data preprocessing. Physical Review E, 111(4).DOI: <a class="reference external" href="https://journals.aps.org/pre/abstract/10.1103/PhysRevE.111.045303">https://journals.aps.org/pre/abstract/10.1103/PhysRevE.111.045303</a></p></li>
</ul>
</section>
</section>
<section id="mathematical-foundation">
<h2><span class="section-number">4.4. </span>Mathematical Foundation<a class="headerlink" href="#mathematical-foundation" title="Link to this heading">#</a></h2>
<p>PCA is often viewed as looking for the directions with the most variance. But it can also be formulated as  an optimization problem, and this explains why is the covariance matrix actually there. First, let’s assume all data is centered, that is it has mean 0. Now, let’s also assume that there is a new given unit vector <span class="math notranslate nohighlight">\(u\)</span> that could be more useful that some of the original axis we were using. The projection of the features vector <span class="math notranslate nohighlight">\(x\)</span> on <span class="math notranslate nohighlight">\(u\)</span> is</p>
<div class="amsmath math notranslate nohighlight" id="equation-69accc5a-3140-4a14-a3e2-258a8d56737d">
<span class="eqno">(4.3)<a class="headerlink" href="#equation-69accc5a-3140-4a14-a3e2-258a8d56737d" title="Permalink to this equation">#</a></span>\[\begin{equation}
x' = (u\cdot x) u = (u^T x) u = x_u u.
\end{equation}\]</div>
<p>The dot product represents the amount of information we are retaining, it is maximum when both vectors are in the same direction, and null when they are orthogonal. Since it could be negative, we can define the aount of preserved information as</p>
<div class="amsmath math notranslate nohighlight" id="equation-ff1bae08-8832-4643-91e8-53e86d65a18e">
<span class="eqno">(4.4)<a class="headerlink" href="#equation-ff1bae08-8832-4643-91e8-53e86d65a18e" title="Permalink to this equation">#</a></span>\[\begin{equation}
(u\cdot x)^2,
\end{equation}\]</div>
<p>or, better, the mean preserved info over all samples as</p>
<div class="amsmath math notranslate nohighlight" id="equation-ec15cc98-dfec-4821-a11c-79d7408c2199">
<span class="eqno">(4.5)<a class="headerlink" href="#equation-ec15cc98-dfec-4821-a11c-79d7408c2199" title="Permalink to this equation">#</a></span>\[\begin{equation}
\frac{1}{n}\sum_i^n (u\cdot x_i)^2 = \frac{1}{n} \sum_i (x_{i,u})^2 = \frac{1}{n} \sum_i x_{i,u}^T x_{i,u} = \frac{1}{n} \sum_i (x_i^T u)(u^T x_i) = \frac{1}{n} \sum_i (x_i^T x_i) u u^T = \frac{1}{n} \sum_i (x_i^T x_i) u^2, = C u^2,   
\end{equation}\]</div>
<p>where we have replaced the definition of the covariance matrix</p>
<div class="amsmath math notranslate nohighlight" id="equation-1d122333-7f7d-49a5-b2fd-b16d386e0a56">
<span class="eqno">(4.6)<a class="headerlink" href="#equation-1d122333-7f7d-49a5-b2fd-b16d386e0a56" title="Permalink to this equation">#</a></span>\[\begin{equation}
C = \frac{1}{n} \sum_i (x_i^T x_i).
\end{equation}\]</div>
<p>Therefore, to find the direction that preserve the most information about our original data, we must solve the optimization problem</p>
<div class="amsmath math notranslate nohighlight" id="equation-d27bf8fb-266d-4dd4-a5e2-2b9c3c31f39a">
<span class="eqno">(4.7)<a class="headerlink" href="#equation-d27bf8fb-266d-4dd4-a5e2-2b9c3c31f39a" title="Permalink to this equation">#</a></span>\[\begin{equation}
\max Cu^2
\end{equation}\]</div>
<p>subject to</p>
<div class="amsmath math notranslate nohighlight" id="equation-02d366bb-6dba-44c4-a93c-8727fa155d37">
<span class="eqno">(4.8)<a class="headerlink" href="#equation-02d366bb-6dba-44c4-a93c-8727fa155d37" title="Permalink to this equation">#</a></span>\[\begin{equation}
u^2 = 1.
\end{equation}\]</div>
<p>This can be donde using the method of Lagrange multipliers, and some matrix calculus allows us to write it here simply:</p>
<div class="amsmath math notranslate nohighlight" id="equation-4e36b4a8-e494-4849-aed8-2c0d1bf7c96c">
<span class="eqno">(4.9)<a class="headerlink" href="#equation-4e36b4a8-e494-4849-aed8-2c0d1bf7c96c" title="Permalink to this equation">#</a></span>\[\begin{equation}
\max Cu^2, \text{s.t, \ \ } u^2 = 1,
\end{equation}\]</div>
<p>deriving with respect to u (yes, it is a vector, see matrix calculus), and introducing a Lagrange multiplier <span class="math notranslate nohighlight">\(\lambda\)</span> for the constraint</p>
<div class="amsmath math notranslate nohighlight" id="equation-53234c71-3134-499d-b478-8ef02ef13dbe">
<span class="eqno">(4.10)<a class="headerlink" href="#equation-53234c71-3134-499d-b478-8ef02ef13dbe" title="Permalink to this equation">#</a></span>\[\begin{equation}
\frac{\partial}{\partial u} (Cu^2 - \lambda (u^2 - 1)) = 0,
\end{equation}\]</div>
<div class="amsmath math notranslate nohighlight" id="equation-ba87374b-c2da-47f4-bd54-45d60ebfd675">
<span class="eqno">(4.11)<a class="headerlink" href="#equation-ba87374b-c2da-47f4-bd54-45d60ebfd675" title="Permalink to this equation">#</a></span>\[\begin{equation}
2Cu - 2\lambda u = 0,
\end{equation}\]</div>
<p>and, finally</p>
<div class="amsmath math notranslate nohighlight" id="equation-6e78f2ca-d522-45d1-b6c6-3bfa0c609a5a">
<span class="eqno">(4.12)<a class="headerlink" href="#equation-6e78f2ca-d522-45d1-b6c6-3bfa0c609a5a" title="Permalink to this equation">#</a></span>\[\begin{equation}
Cu = \lambda u, 
\end{equation}\]</div>
<p>which is exactly the eigen-value problem for the covariance matrix. Therefore, the directions that preserve the most information about our original data are exactly the eigen-vectors of the covariance matrix.</p>
</section>
<section id="general-algorithm">
<h2><span class="section-number">4.5. </span>General algorithm<a class="headerlink" href="#general-algorithm" title="Link to this heading">#</a></h2>
<section id="step-1-data-centering">
<h3><span class="section-number">4.5.1. </span>Step 1: Data Centering<a class="headerlink" href="#step-1-data-centering" title="Link to this heading">#</a></h3>
<div class="amsmath math notranslate nohighlight" id="equation-23e84992-a518-4811-8732-afe846fec441">
<span class="eqno">(4.13)<a class="headerlink" href="#equation-23e84992-a518-4811-8732-afe846fec441" title="Permalink to this equation">#</a></span>\[\begin{equation}
\hat X = X - \mu
\end{equation}\]</div>
<p>Where <span class="math notranslate nohighlight">\(\mu\)</span> is the mean of each variable.</p>
<p><strong>Why center the data?</strong></p>
<ul class="simple">
<li><p>Ensures PCA finds directions of maximum variance from the center</p></li>
<li><p>Prevents variables with larger scales from dominating</p></li>
</ul>
</section>
<section id="step-2-covariance-matrix">
<h3><span class="section-number">4.5.2. </span>Step 2: Covariance Matrix<a class="headerlink" href="#step-2-covariance-matrix" title="Link to this heading">#</a></h3>
<div class="amsmath math notranslate nohighlight" id="equation-56e915f0-b652-427b-92fb-04e6928b7ccb">
<span class="eqno">(4.14)<a class="headerlink" href="#equation-56e915f0-b652-427b-92fb-04e6928b7ccb" title="Permalink to this equation">#</a></span>\[\begin{equation}
C = \frac{1}{n-1}\  \hat X^T \times \hat X.
\end{equation}\]</div>
<p>Where <span class="math notranslate nohighlight">\(n\)</span> is the number of observations.</p>
<p><strong>The covariance matrix captures</strong>:</p>
<ul class="simple">
<li><p>How much each variable varies (diagonal elements)</p></li>
<li><p>How variables co-vary together (off-diagonal elements)</p></li>
</ul>
</section>
<section id="step-3-eigendecomposition">
<h3><span class="section-number">4.5.3. </span>Step 3: Eigendecomposition<a class="headerlink" href="#step-3-eigendecomposition" title="Link to this heading">#</a></h3>
<div class="amsmath math notranslate nohighlight" id="equation-1f7b6ef5-49fe-4792-a5da-ddbf4126dffd">
<span class="eqno">(4.15)<a class="headerlink" href="#equation-1f7b6ef5-49fe-4792-a5da-ddbf4126dffd" title="Permalink to this equation">#</a></span>\[\begin{equation}
C\times \vec v = \lambda \vec v
\end{equation}\]</div>
<p>Where:</p>
<ul class="simple">
<li><p><span class="math notranslate nohighlight">\(\vec v\)</span> are eigenvectors (principal component directions)</p></li>
<li><p><span class="math notranslate nohighlight">\(\lambda\)</span> are eigenvalues (variance captured by each component)</p></li>
</ul>
</section>
<section id="step-4-variance-explained">
<h3><span class="section-number">4.5.4. </span>Step 4: Variance Explained<a class="headerlink" href="#step-4-variance-explained" title="Link to this heading">#</a></h3>
<p>Variance Explained by</p>
<div class="amsmath math notranslate nohighlight" id="equation-5e674862-e874-45be-8171-f419d7c95056">
<span class="eqno">(4.16)<a class="headerlink" href="#equation-5e674862-e874-45be-8171-f419d7c95056" title="Permalink to this equation">#</a></span>\[\begin{equation}
PC_i = \frac{\lambda_i}{\sum_j\lambda_j}
\end{equation}\]</div>
<div class="{exercise}Covariance docutils">
<p><strong>Question</strong>: If two variables have a covariance of 0, what does this mean for PCA?</p>
<p>Consider this covariance matrix:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">C</span> <span class="o">=</span> <span class="p">[</span><span class="mf">4.0</span>  <span class="mf">0.0</span><span class="p">]</span>
    <span class="p">[</span><span class="mf">0.0</span>  <span class="mf">1.0</span><span class="p">]</span>
</pre></div>
</div>
<p><strong>Tasks</strong>:</p>
<ol class="arabic simple">
<li><p>What are the eigenvalues?</p></li>
<li><p>What are the eigenvectors?</p></li>
<li><p>How much variance does each PC explain?</p></li>
</ol>
</div>
</section>
</section>
<section id="pca-applied-to-the-previous-data-using-scikit-learn">
<h2><span class="section-number">4.6. </span>PCA applied to the previous data using scikit learn<a class="headerlink" href="#pca-applied-to-the-previous-data-using-scikit-learn" title="Link to this heading">#</a></h2>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span><span class="w"> </span><span class="nn">numpy</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">np</span>

<span class="kn">from</span><span class="w"> </span><span class="nn">sklearn.preprocessing</span><span class="w"> </span><span class="kn">import</span> <span class="n">StandardScaler</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">sklearn.decomposition</span><span class="w"> </span><span class="kn">import</span> <span class="n">PCA</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">pandas</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">pd</span>

<span class="c1"># Data generation</span>
<span class="n">np</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">seed</span><span class="p">(</span><span class="mi">0</span><span class="p">)</span>
<span class="n">n</span> <span class="o">=</span> <span class="mi">200</span>
<span class="n">x</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">normal</span><span class="p">(</span><span class="n">size</span><span class="o">=</span><span class="n">n</span><span class="p">)</span>
<span class="n">y</span> <span class="o">=</span> <span class="mf">2.0</span> <span class="o">*</span> <span class="n">x</span> <span class="o">+</span> <span class="mf">0.8</span> <span class="o">*</span> <span class="n">np</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">normal</span><span class="p">(</span><span class="n">size</span><span class="o">=</span><span class="n">n</span><span class="p">)</span>
<span class="n">X</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">vstack</span><span class="p">([</span><span class="n">x</span><span class="p">,</span> <span class="n">y</span><span class="p">])</span><span class="o">.</span><span class="n">T</span>

<span class="c1">#Xc = X - X.mean(axis=0)</span>
<span class="n">X_scaled</span> <span class="o">=</span> <span class="n">StandardScaler</span><span class="p">()</span><span class="o">.</span><span class="n">fit_transform</span><span class="p">(</span><span class="n">X</span><span class="p">)</span>

<span class="n">pca</span> <span class="o">=</span> <span class="n">PCA</span><span class="p">(</span><span class="n">n_components</span><span class="o">=</span><span class="mi">2</span><span class="p">)</span>
<span class="n">X_pca</span> <span class="o">=</span> <span class="n">pca</span><span class="o">.</span><span class="n">fit_transform</span><span class="p">(</span><span class="n">X_scaled</span><span class="p">)</span>

<span class="n">explained_variance</span> <span class="o">=</span> <span class="n">pca</span><span class="o">.</span><span class="n">explained_variance_ratio_</span>
<span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="se">\n</span><span class="s2">Variance explained by PC1: </span><span class="si">{</span><span class="n">explained_variance</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="si">:</span><span class="s2">.2%</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Variance explained by PC2: </span><span class="si">{</span><span class="n">explained_variance</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="si">:</span><span class="s2">.2%</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Total variance explained: </span><span class="si">{</span><span class="n">np</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="n">explained_variance</span><span class="p">)</span><span class="si">:</span><span class="s2">.2%</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Variance explained by PC1: 97.07%
Variance explained by PC2: 2.93%
Total variance explained: 100.00%
</pre></div>
</div>
</div>
</div>
</section>
<section id="pca-applied-to-real-data">
<h2><span class="section-number">4.7. </span>PCA applied to real data<a class="headerlink" href="#pca-applied-to-real-data" title="Link to this heading">#</a></h2>
<p>Sky data survey. To generate the data, go to <a class="reference external" href="https://skyserver.sdss.org/dr14/en/tools/search/form/searchform.aspx">https://skyserver.sdss.org/dr14/en/tools/search/form/searchform.aspx</a> and write or use the default query. To get the full data, use the following sql command</p>
<div class="highlight-sql notranslate"><div class="highlight"><pre><span></span><span class="w">    </span>
<span class="k">SELECT</span><span class="w"> </span><span class="n">TOP</span><span class="w"> </span><span class="mi">10000</span>
<span class="w">    </span><span class="n">p</span><span class="p">.</span><span class="n">ra</span><span class="p">,</span><span class="w"> </span><span class="n">p</span><span class="p">.</span><span class="nb">dec</span><span class="p">,</span><span class="w"> </span><span class="n">p</span><span class="p">.</span><span class="n">u</span><span class="p">,</span><span class="w"> </span><span class="n">p</span><span class="p">.</span><span class="k">g</span><span class="p">,</span><span class="w"> </span><span class="n">p</span><span class="p">.</span><span class="n">r</span><span class="p">,</span><span class="w"> </span><span class="n">p</span><span class="p">.</span><span class="n">i</span><span class="p">,</span><span class="w"> </span><span class="n">p</span><span class="p">.</span><span class="n">z</span><span class="p">,</span>
<span class="w">    </span><span class="n">p</span><span class="p">.</span><span class="n">run</span><span class="p">,</span><span class="w"> </span><span class="n">p</span><span class="p">.</span><span class="n">rerun</span><span class="p">,</span><span class="w"> </span><span class="n">p</span><span class="p">.</span><span class="n">camcol</span><span class="p">,</span><span class="w"> </span><span class="n">p</span><span class="p">.</span><span class="n">field</span><span class="p">,</span>
<span class="w">    </span><span class="n">s</span><span class="p">.</span><span class="n">specobjid</span><span class="p">,</span><span class="w"> </span><span class="n">s</span><span class="p">.</span><span class="k">class</span><span class="p">,</span><span class="w"> </span><span class="n">s</span><span class="p">.</span><span class="n">z</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="n">redshift</span>
<span class="k">FROM</span><span class="w"> </span><span class="n">PhotoObj</span><span class="w"> </span><span class="k">AS</span><span class="w"> </span><span class="n">p</span>
<span class="k">JOIN</span><span class="w"> </span><span class="n">SpecObj</span><span class="w"> </span><span class="k">AS</span><span class="w"> </span><span class="n">s</span><span class="w"> </span><span class="k">ON</span><span class="w"> </span><span class="n">s</span><span class="p">.</span><span class="n">bestobjid</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">p</span><span class="p">.</span><span class="n">objid</span>
<span class="k">WHERE</span>
<span class="w">    </span><span class="n">p</span><span class="p">.</span><span class="n">u</span><span class="w"> </span><span class="k">BETWEEN</span><span class="w"> </span><span class="mi">0</span><span class="w"> </span><span class="k">AND</span><span class="w"> </span><span class="mi">19</span><span class="p">.</span><span class="mi">6</span>
<span class="w">    </span><span class="k">AND</span><span class="w"> </span><span class="n">p</span><span class="p">.</span><span class="k">g</span><span class="w"> </span><span class="k">BETWEEN</span><span class="w"> </span><span class="mi">0</span><span class="w"> </span><span class="k">AND</span><span class="w"> </span><span class="mi">20</span>
</pre></div>
</div>
<div class="cell tag_hide-input docutils container">
<details class="admonition hide above-input">
<summary aria-label="Toggle hidden content">
<p class="collapsed admonition-title">Show code cell source</p>
<p class="expanded admonition-title">Hide code cell source</p>
</summary>
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span><span class="w"> </span><span class="nn">pandas</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">pd</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">numpy</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">np</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">matplotlib.pyplot</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">plt</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">seaborn</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">sns</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">sklearn.preprocessing</span><span class="w"> </span><span class="kn">import</span> <span class="n">StandardScaler</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">sklearn.decomposition</span><span class="w"> </span><span class="kn">import</span> <span class="n">PCA</span>

<span class="c1"># --- 1. DATA ACQUISITION ---</span>
<span class="c1"># Load the data from the CSV file you downloaded from the SDSS SkyServer.</span>
<span class="c1">#file_path = &#39;SDSS_DR14.csv&#39; # Or whatever you named the downloaded file</span>
<span class="n">file_path</span> <span class="o">=</span> <span class="s1">&#39;skyserverdata2.csv&#39;</span> <span class="c1"># Or whatever you named the downloaded file</span>
<span class="c1"># url = &quot;https://drive.google.com/file/d/1ZLnCVim0P1Ktewyhd1VcRNWTDCte-W0O/view?usp=drive_link&quot;</span>

<span class="k">try</span><span class="p">:</span>
    <span class="c1"># Use skiprows=1 to ignore the first line (&quot;#Table1&quot;)</span>
    <span class="n">sdss_df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">file_path</span><span class="p">,</span> <span class="n">skiprows</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
    <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;SDSS dataset loaded successfully and parsed correctly!&quot;</span><span class="p">)</span>

<span class="k">except</span> <span class="ne">FileNotFoundError</span><span class="p">:</span>
    <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;File not found: &#39;</span><span class="si">{</span><span class="n">file_path</span><span class="si">}</span><span class="s2">&#39;. Please make sure you have downloaded the data and placed it in the correct directory.&quot;</span><span class="p">)</span>
    <span class="n">sdss_df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">()</span>

<span class="c1"># Proceed only if the DataFrame was loaded successfully</span>
<span class="k">if</span> <span class="ow">not</span> <span class="n">sdss_df</span><span class="o">.</span><span class="n">empty</span><span class="p">:</span>
    <span class="c1"># --- 2. DATA EXPLORATION AND PREPARATION ---</span>
    <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;</span><span class="se">\n</span><span class="s2">Dataset Shape:&quot;</span><span class="p">,</span> <span class="n">sdss_df</span><span class="o">.</span><span class="n">shape</span><span class="p">)</span>
    <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;</span><span class="se">\n</span><span class="s2">Columns:&quot;</span><span class="p">,</span> <span class="n">sdss_df</span><span class="o">.</span><span class="n">columns</span><span class="p">)</span>

    <span class="c1"># Define our features and the target variable for coloring the plot</span>
    <span class="n">features</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;u&#39;</span><span class="p">,</span> <span class="s1">&#39;g&#39;</span><span class="p">,</span> <span class="s1">&#39;r&#39;</span><span class="p">,</span> <span class="s1">&#39;i&#39;</span><span class="p">,</span> <span class="s1">&#39;z&#39;</span><span class="p">]</span>
    <span class="n">target</span> <span class="o">=</span> <span class="s1">&#39;class&#39;</span>

    <span class="c1"># --- Safety Check ---</span>
    <span class="c1"># Verify that all required columns exist in the DataFrame before proceeding.</span>
    <span class="k">if</span> <span class="n">target</span> <span class="ow">in</span> <span class="n">sdss_df</span><span class="o">.</span><span class="n">columns</span> <span class="ow">and</span> <span class="nb">all</span><span class="p">(</span><span class="n">col</span> <span class="ow">in</span> <span class="n">sdss_df</span><span class="o">.</span><span class="n">columns</span> <span class="k">for</span> <span class="n">col</span> <span class="ow">in</span> <span class="n">features</span><span class="p">):</span>
        
        <span class="n">X</span> <span class="o">=</span> <span class="n">sdss_df</span><span class="p">[</span><span class="n">features</span><span class="p">]</span>
        <span class="n">y</span> <span class="o">=</span> <span class="n">sdss_df</span><span class="p">[</span><span class="n">target</span><span class="p">]</span>

        <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;</span><span class="se">\n</span><span class="s2">Number of samples for each class:&quot;</span><span class="p">)</span>
        <span class="nb">print</span><span class="p">(</span><span class="n">y</span><span class="o">.</span><span class="n">value_counts</span><span class="p">())</span>

        <span class="c1"># Standardize the feature data</span>
        <span class="n">X_scaled</span> <span class="o">=</span> <span class="n">StandardScaler</span><span class="p">()</span><span class="o">.</span><span class="n">fit_transform</span><span class="p">(</span><span class="n">X</span><span class="p">)</span>
        <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;</span><span class="se">\n</span><span class="s2">Data has been standardized.&quot;</span><span class="p">)</span>

        <span class="c1"># --- 3. APPLYING PCA ---</span>
        <span class="n">pca</span> <span class="o">=</span> <span class="n">PCA</span><span class="p">(</span><span class="n">n_components</span><span class="o">=</span><span class="mi">2</span><span class="p">)</span>
        <span class="n">X_pca</span> <span class="o">=</span> <span class="n">pca</span><span class="o">.</span><span class="n">fit_transform</span><span class="p">(</span><span class="n">X_scaled</span><span class="p">)</span>

        <span class="n">explained_variance</span> <span class="o">=</span> <span class="n">pca</span><span class="o">.</span><span class="n">explained_variance_ratio_</span>
        <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="se">\n</span><span class="s2">Variance explained by PC1: </span><span class="si">{</span><span class="n">explained_variance</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="si">:</span><span class="s2">.2%</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
        <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Variance explained by PC2: </span><span class="si">{</span><span class="n">explained_variance</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="si">:</span><span class="s2">.2%</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
        <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Total variance explained: </span><span class="si">{</span><span class="n">np</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="n">explained_variance</span><span class="p">)</span><span class="si">:</span><span class="s2">.2%</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>

        <span class="c1"># --- 4. VISUALIZATION AND INTERPRETATION ---</span>

        <span class="c1"># Create a new DataFrame for easier plotting with seaborn</span>
        <span class="n">pca_df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">data</span><span class="o">=</span><span class="n">X_pca</span><span class="p">,</span> <span class="n">columns</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;PC1&#39;</span><span class="p">,</span> <span class="s1">&#39;PC2&#39;</span><span class="p">])</span>
        <span class="n">pca_df</span><span class="p">[</span><span class="s1">&#39;class&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">y</span><span class="o">.</span><span class="n">values</span> <span class="c1"># Add the class labels for coloring</span>

        <span class="c1"># --- PLOT 1: SCORES PLOT (COLORED) ---</span>
        <span class="c1"># Scores mean the data is plotted on the new coordinate system : PC1xPC2</span>
        <span class="n">plt</span><span class="o">.</span><span class="n">figure</span><span class="p">(</span><span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">12</span><span class="p">,</span> <span class="mi">9</span><span class="p">))</span>
        <span class="n">sns</span><span class="o">.</span><span class="n">scatterplot</span><span class="p">(</span>
            <span class="n">x</span><span class="o">=</span><span class="s1">&#39;PC1&#39;</span><span class="p">,</span> <span class="n">y</span><span class="o">=</span><span class="s1">&#39;PC2&#39;</span><span class="p">,</span>
            <span class="n">hue</span><span class="o">=</span><span class="s1">&#39;class&#39;</span><span class="p">,</span>
            <span class="n">data</span><span class="o">=</span><span class="n">pca_df</span><span class="p">,</span>
            <span class="n">palette</span><span class="o">=</span><span class="s1">&#39;viridis&#39;</span><span class="p">,</span>
            <span class="n">alpha</span><span class="o">=</span><span class="mf">0.7</span><span class="p">,</span>
            <span class="n">s</span><span class="o">=</span><span class="mi">40</span> <span class="c1"># marker size</span>
        <span class="p">)</span>
        <span class="n">plt</span><span class="o">.</span><span class="n">title</span><span class="p">(</span><span class="s1">&#39;PCA of SDSS Astronomical Objects (based on colors)&#39;</span><span class="p">,</span> <span class="n">fontsize</span><span class="o">=</span><span class="mi">16</span><span class="p">)</span>
        <span class="n">plt</span><span class="o">.</span><span class="n">xlabel</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;Principal Component 1 (</span><span class="si">{</span><span class="n">explained_variance</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="si">:</span><span class="s1">.2%</span><span class="si">}</span><span class="s1">)&#39;</span><span class="p">,</span> <span class="n">fontsize</span><span class="o">=</span><span class="mi">12</span><span class="p">)</span>
        <span class="n">plt</span><span class="o">.</span><span class="n">ylabel</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;Principal Component 2 (</span><span class="si">{</span><span class="n">explained_variance</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="si">:</span><span class="s1">.2%</span><span class="si">}</span><span class="s1">)&#39;</span><span class="p">,</span> <span class="n">fontsize</span><span class="o">=</span><span class="mi">12</span><span class="p">)</span>
        <span class="n">plt</span><span class="o">.</span><span class="n">legend</span><span class="p">(</span><span class="n">title</span><span class="o">=</span><span class="s1">&#39;Object Class&#39;</span><span class="p">)</span>
        <span class="n">plt</span><span class="o">.</span><span class="n">grid</span><span class="p">(</span><span class="kc">True</span><span class="p">)</span>
        <span class="n">plt</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>

        <span class="c1"># --- PLOT 2: LOADINGS PLOT ---</span>
        <span class="n">plt</span><span class="o">.</span><span class="n">figure</span><span class="p">(</span><span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">10</span><span class="p">,</span> <span class="mi">8</span><span class="p">))</span>
        <span class="c1"># In scikit-learn, pca.components_ are the loadings. We transpose for easier plotting.</span>
        <span class="n">loadings</span> <span class="o">=</span> <span class="n">pca</span><span class="o">.</span><span class="n">components_</span><span class="o">.</span><span class="n">T</span>
        
        <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">feature</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">features</span><span class="p">):</span>
            <span class="n">plt</span><span class="o">.</span><span class="n">arrow</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="n">loadings</span><span class="p">[</span><span class="n">i</span><span class="p">,</span> <span class="mi">0</span><span class="p">],</span> <span class="n">loadings</span><span class="p">[</span><span class="n">i</span><span class="p">,</span> <span class="mi">1</span><span class="p">],</span> <span class="n">head_width</span><span class="o">=</span><span class="mf">0.05</span><span class="p">,</span> <span class="n">head_length</span><span class="o">=</span><span class="mf">0.05</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="s1">&#39;red&#39;</span><span class="p">,</span> <span class="n">alpha</span><span class="o">=</span><span class="mf">0.8</span><span class="p">)</span>
            <span class="n">plt</span><span class="o">.</span><span class="n">text</span><span class="p">(</span><span class="n">loadings</span><span class="p">[</span><span class="n">i</span><span class="p">,</span> <span class="mi">0</span><span class="p">]</span><span class="o">*</span><span class="mf">1.15</span><span class="p">,</span> <span class="n">loadings</span><span class="p">[</span><span class="n">i</span><span class="p">,</span> <span class="mi">1</span><span class="p">]</span><span class="o">*</span><span class="mf">1.15</span><span class="p">,</span> <span class="n">feature</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="s1">&#39;black&#39;</span><span class="p">,</span> <span class="n">ha</span><span class="o">=</span><span class="s1">&#39;center&#39;</span><span class="p">,</span> <span class="n">va</span><span class="o">=</span><span class="s1">&#39;center&#39;</span><span class="p">,</span> <span class="n">fontsize</span><span class="o">=</span><span class="mi">14</span><span class="p">)</span>

        <span class="n">plt</span><span class="o">.</span><span class="n">xlim</span><span class="p">(</span><span class="o">-</span><span class="mf">0.8</span><span class="p">,</span> <span class="mf">0.8</span><span class="p">)</span>
        <span class="n">plt</span><span class="o">.</span><span class="n">ylim</span><span class="p">(</span><span class="o">-</span><span class="mf">0.8</span><span class="p">,</span> <span class="mf">0.8</span><span class="p">)</span>
        <span class="n">plt</span><span class="o">.</span><span class="n">xlabel</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;PC1 (</span><span class="si">{</span><span class="n">explained_variance</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="si">:</span><span class="s1">.2%</span><span class="si">}</span><span class="s1">)&#39;</span><span class="p">)</span>
        <span class="n">plt</span><span class="o">.</span><span class="n">ylabel</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;PC2 (</span><span class="si">{</span><span class="n">explained_variance</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="si">:</span><span class="s1">.2%</span><span class="si">}</span><span class="s1">)&#39;</span><span class="p">)</span>
        <span class="n">plt</span><span class="o">.</span><span class="n">title</span><span class="p">(</span><span class="s1">&#39;Loadings Plot&#39;</span><span class="p">,</span> <span class="n">fontsize</span><span class="o">=</span><span class="mi">16</span><span class="p">)</span>
        <span class="n">plt</span><span class="o">.</span><span class="n">gca</span><span class="p">()</span><span class="o">.</span><span class="n">set_aspect</span><span class="p">(</span><span class="s1">&#39;equal&#39;</span><span class="p">,</span> <span class="n">adjustable</span><span class="o">=</span><span class="s1">&#39;box&#39;</span><span class="p">)</span>
        <span class="n">plt</span><span class="o">.</span><span class="n">grid</span><span class="p">()</span>
        <span class="n">plt</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>

    <span class="k">else</span><span class="p">:</span>
        <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;</span><span class="se">\n</span><span class="s2">--- ERROR ---&quot;</span><span class="p">)</span>
        <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;The target column &#39;</span><span class="si">{</span><span class="n">target</span><span class="si">}</span><span class="s2">&#39; or one of the feature columns was not found in the data.&quot;</span><span class="p">)</span>
        <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Please ensure your downloaded file contains all necessary columns.&quot;</span><span class="p">)</span>
        <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Required columns: &#39;u&#39;, &#39;g&#39;, &#39;r&#39;, &#39;i&#39;, &#39;z&#39;, &#39;class&#39;&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
</details>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>File not found: &#39;skyserverdata2.csv&#39;. Please make sure you have downloaded the data and placed it in the correct directory.
</pre></div>
</div>
</div>
</div>
<p>This is a bokeh plotting version with some additional info</p>
<div class="cell tag_hide-input docutils container">
<details class="admonition hide above-input">
<summary aria-label="Toggle hidden content">
<p class="collapsed admonition-title">Show code cell source</p>
<p class="expanded admonition-title">Hide code cell source</p>
</summary>
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span><span class="w"> </span><span class="nn">pandas</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">pd</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">numpy</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">np</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">sklearn.preprocessing</span><span class="w"> </span><span class="kn">import</span> <span class="n">StandardScaler</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">sklearn.decomposition</span><span class="w"> </span><span class="kn">import</span> <span class="n">PCA</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">bokeh.plotting</span><span class="w"> </span><span class="kn">import</span> <span class="n">figure</span><span class="p">,</span> <span class="n">show</span><span class="p">,</span> <span class="n">output_notebook</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">bokeh.io</span><span class="w"> </span><span class="kn">import</span> <span class="n">push_notebook</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">bokeh.palettes</span><span class="w"> </span><span class="kn">import</span> <span class="n">Category10</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">bokeh.models</span><span class="w"> </span><span class="kn">import</span> <span class="n">ColumnDataSource</span><span class="p">,</span> <span class="n">HoverTool</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">bokeh.layouts</span><span class="w"> </span><span class="kn">import</span> <span class="n">column</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">warnings</span>
<span class="n">warnings</span><span class="o">.</span><span class="n">filterwarnings</span><span class="p">(</span><span class="s1">&#39;ignore&#39;</span><span class="p">)</span>

<span class="c1"># Enable Bokeh output in notebook</span>
<span class="n">output_notebook</span><span class="p">()</span>

<span class="c1"># --- 1. DATA ACQUISITION ---</span>
<span class="n">file_path</span> <span class="o">=</span> <span class="s1">&#39;skyserverdata2.csv&#39;</span>  <span class="c1"># Or whatever you named the downloaded file</span>
<span class="k">try</span><span class="p">:</span>
    <span class="c1"># Use skiprows=1 to ignore the first line (&quot;#Table1&quot;)</span>
    <span class="n">sdss_df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">file_path</span><span class="p">,</span> <span class="n">skiprows</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
    <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;SDSS dataset loaded successfully and parsed correctly!&quot;</span><span class="p">)</span>
<span class="k">except</span> <span class="ne">FileNotFoundError</span><span class="p">:</span>
    <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;File not found: &#39;</span><span class="si">{</span><span class="n">file_path</span><span class="si">}</span><span class="s2">&#39;. Please make sure you have downloaded the data and placed it in the correct directory.&quot;</span><span class="p">)</span>
    <span class="n">sdss_df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">()</span>

<span class="c1"># Proceed only if the DataFrame was loaded successfully</span>
<span class="k">if</span> <span class="ow">not</span> <span class="n">sdss_df</span><span class="o">.</span><span class="n">empty</span><span class="p">:</span>
    <span class="c1"># --- 2. DATA EXPLORATION AND PREPARATION ---</span>
    <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;</span><span class="se">\n</span><span class="s2">Dataset Shape:&quot;</span><span class="p">,</span> <span class="n">sdss_df</span><span class="o">.</span><span class="n">shape</span><span class="p">)</span>
    <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;</span><span class="se">\n</span><span class="s2">Columns:&quot;</span><span class="p">,</span> <span class="n">sdss_df</span><span class="o">.</span><span class="n">columns</span><span class="o">.</span><span class="n">tolist</span><span class="p">())</span>

    <span class="c1"># Define our features and the target variable for coloring the plot</span>
    <span class="n">features</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;u&#39;</span><span class="p">,</span> <span class="s1">&#39;g&#39;</span><span class="p">,</span> <span class="s1">&#39;r&#39;</span><span class="p">,</span> <span class="s1">&#39;i&#39;</span><span class="p">,</span> <span class="s1">&#39;z&#39;</span><span class="p">]</span>
    <span class="n">target</span> <span class="o">=</span> <span class="s1">&#39;class&#39;</span>

    <span class="c1"># --- Safety Check ---</span>
    <span class="k">if</span> <span class="n">target</span> <span class="ow">in</span> <span class="n">sdss_df</span><span class="o">.</span><span class="n">columns</span> <span class="ow">and</span> <span class="nb">all</span><span class="p">(</span><span class="n">col</span> <span class="ow">in</span> <span class="n">sdss_df</span><span class="o">.</span><span class="n">columns</span> <span class="k">for</span> <span class="n">col</span> <span class="ow">in</span> <span class="n">features</span><span class="p">):</span>
        
        <span class="n">X</span> <span class="o">=</span> <span class="n">sdss_df</span><span class="p">[</span><span class="n">features</span><span class="p">]</span>
        <span class="n">y</span> <span class="o">=</span> <span class="n">sdss_df</span><span class="p">[</span><span class="n">target</span><span class="p">]</span>

        <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;</span><span class="se">\n</span><span class="s2">Number of samples for each class:&quot;</span><span class="p">)</span>
        <span class="nb">print</span><span class="p">(</span><span class="n">y</span><span class="o">.</span><span class="n">value_counts</span><span class="p">())</span>

        <span class="c1"># Remove any rows with NaN values</span>
        <span class="n">mask</span> <span class="o">=</span> <span class="o">~</span><span class="p">(</span><span class="n">X</span><span class="o">.</span><span class="n">isnull</span><span class="p">()</span><span class="o">.</span><span class="n">any</span><span class="p">(</span><span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span> <span class="o">|</span> <span class="n">y</span><span class="o">.</span><span class="n">isnull</span><span class="p">())</span>
        <span class="n">X</span> <span class="o">=</span> <span class="n">X</span><span class="p">[</span><span class="n">mask</span><span class="p">]</span>
        <span class="n">y</span> <span class="o">=</span> <span class="n">y</span><span class="p">[</span><span class="n">mask</span><span class="p">]</span>
        <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="se">\n</span><span class="s2">After removing NaN values: </span><span class="si">{</span><span class="nb">len</span><span class="p">(</span><span class="n">X</span><span class="p">)</span><span class="si">}</span><span class="s2"> samples&quot;</span><span class="p">)</span>

        <span class="c1"># Standardize the feature data</span>
        <span class="n">scaler</span> <span class="o">=</span> <span class="n">StandardScaler</span><span class="p">()</span>
        <span class="n">X_scaled</span> <span class="o">=</span> <span class="n">scaler</span><span class="o">.</span><span class="n">fit_transform</span><span class="p">(</span><span class="n">X</span><span class="p">)</span>
        <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;</span><span class="se">\n</span><span class="s2">Data has been standardized.&quot;</span><span class="p">)</span>

        <span class="c1"># --- 3. APPLYING PCA ---</span>
        <span class="n">pca</span> <span class="o">=</span> <span class="n">PCA</span><span class="p">(</span><span class="n">n_components</span><span class="o">=</span><span class="mi">2</span><span class="p">)</span>
        <span class="n">X_pca</span> <span class="o">=</span> <span class="n">pca</span><span class="o">.</span><span class="n">fit_transform</span><span class="p">(</span><span class="n">X_scaled</span><span class="p">)</span>

        <span class="n">explained_variance</span> <span class="o">=</span> <span class="n">pca</span><span class="o">.</span><span class="n">explained_variance_ratio_</span>
        <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="se">\n</span><span class="s2">Variance explained by PC1: </span><span class="si">{</span><span class="n">explained_variance</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="si">:</span><span class="s2">.2%</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
        <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Variance explained by PC2: </span><span class="si">{</span><span class="n">explained_variance</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="si">:</span><span class="s2">.2%</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
        <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Total variance explained: </span><span class="si">{</span><span class="n">np</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="n">explained_variance</span><span class="p">)</span><span class="si">:</span><span class="s2">.2%</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>

        <span class="c1"># --- 4. VISUALIZATION AND INTERPRETATION ---</span>

        <span class="c1"># Create a new DataFrame for easier plotting with Bokeh</span>
        <span class="n">pca_df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">data</span><span class="o">=</span><span class="n">X_pca</span><span class="p">,</span> <span class="n">columns</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;PC1&#39;</span><span class="p">,</span> <span class="s1">&#39;PC2&#39;</span><span class="p">])</span>
        <span class="n">pca_df</span><span class="p">[</span><span class="s1">&#39;class&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">y</span><span class="o">.</span><span class="n">values</span>  <span class="c1"># Add the class labels for coloring</span>

        <span class="c1"># Get unique classes and assign colors</span>
        <span class="n">unique_classes</span> <span class="o">=</span> <span class="nb">sorted</span><span class="p">(</span><span class="n">y</span><span class="o">.</span><span class="n">unique</span><span class="p">())</span>
        <span class="n">n_classes</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">unique_classes</span><span class="p">)</span>
        <span class="n">colors</span> <span class="o">=</span> <span class="n">Category10</span><span class="p">[</span><span class="nb">max</span><span class="p">(</span><span class="mi">3</span><span class="p">,</span> <span class="n">n_classes</span><span class="p">)][:</span><span class="n">n_classes</span><span class="p">]</span>
        <span class="n">color_map</span> <span class="o">=</span> <span class="nb">dict</span><span class="p">(</span><span class="nb">zip</span><span class="p">(</span><span class="n">unique_classes</span><span class="p">,</span> <span class="n">colors</span><span class="p">))</span>
        
        <span class="c1"># Add color column to dataframe</span>
        <span class="n">pca_df</span><span class="p">[</span><span class="s1">&#39;color&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="p">[</span><span class="n">color_map</span><span class="p">[</span><span class="bp">cls</span><span class="p">]</span> <span class="k">for</span> <span class="bp">cls</span> <span class="ow">in</span> <span class="n">pca_df</span><span class="p">[</span><span class="s1">&#39;class&#39;</span><span class="p">]]</span>

        <span class="c1"># --- PLOT 1: SCORES PLOT (COLORED) ---</span>
        <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;</span><span class="se">\n</span><span class="s2">Generating PCA scores plot...&quot;</span><span class="p">)</span>
        
        <span class="c1"># Create hover tool</span>
        <span class="n">hover</span> <span class="o">=</span> <span class="n">HoverTool</span><span class="p">(</span><span class="n">tooltips</span><span class="o">=</span><span class="p">[</span>
            <span class="p">(</span><span class="s2">&quot;Class&quot;</span><span class="p">,</span> <span class="s2">&quot;@class&quot;</span><span class="p">),</span>
            <span class="p">(</span><span class="s2">&quot;PC1&quot;</span><span class="p">,</span> <span class="s2">&quot;@PC1</span><span class="si">{0.00}</span><span class="s2">&quot;</span><span class="p">),</span>
            <span class="p">(</span><span class="s2">&quot;PC2&quot;</span><span class="p">,</span> <span class="s2">&quot;@PC2</span><span class="si">{0.00}</span><span class="s2">&quot;</span><span class="p">)</span>
        <span class="p">])</span>

        <span class="c1"># Prepare the data for the scatter plot (scores plot)</span>
        <span class="n">source</span> <span class="o">=</span> <span class="n">ColumnDataSource</span><span class="p">(</span><span class="n">pca_df</span><span class="p">)</span>

        <span class="n">p1</span> <span class="o">=</span> <span class="n">figure</span><span class="p">(</span>
            <span class="n">width</span><span class="o">=</span><span class="mi">800</span><span class="p">,</span> 
            <span class="n">height</span><span class="o">=</span><span class="mi">600</span><span class="p">,</span> 
            <span class="n">title</span><span class="o">=</span><span class="s2">&quot;PCA of SDSS Astronomical Objects (based on ugriz colors)&quot;</span><span class="p">,</span>
            <span class="n">tools</span><span class="o">=</span><span class="p">[</span><span class="n">hover</span><span class="p">,</span> <span class="s2">&quot;pan&quot;</span><span class="p">,</span> <span class="s2">&quot;wheel_zoom&quot;</span><span class="p">,</span> <span class="s2">&quot;box_zoom&quot;</span><span class="p">,</span> <span class="s2">&quot;reset&quot;</span><span class="p">,</span> <span class="s2">&quot;save&quot;</span><span class="p">]</span>
        <span class="p">)</span>
        
        <span class="c1"># Create scatter plot for each class separately to get proper legend</span>
        <span class="k">for</span> <span class="bp">cls</span> <span class="ow">in</span> <span class="n">unique_classes</span><span class="p">:</span>
            <span class="n">class_data</span> <span class="o">=</span> <span class="n">pca_df</span><span class="p">[</span><span class="n">pca_df</span><span class="p">[</span><span class="s1">&#39;class&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="bp">cls</span><span class="p">]</span>
            <span class="n">class_source</span> <span class="o">=</span> <span class="n">ColumnDataSource</span><span class="p">(</span><span class="n">class_data</span><span class="p">)</span>
            <span class="n">p1</span><span class="o">.</span><span class="n">circle</span><span class="p">(</span>
                <span class="n">x</span><span class="o">=</span><span class="s1">&#39;PC1&#39;</span><span class="p">,</span> <span class="n">y</span><span class="o">=</span><span class="s1">&#39;PC2&#39;</span><span class="p">,</span> 
                <span class="n">source</span><span class="o">=</span><span class="n">class_source</span><span class="p">,</span> 
                <span class="n">size</span><span class="o">=</span><span class="mi">8</span><span class="p">,</span> 
                <span class="n">color</span><span class="o">=</span><span class="n">color_map</span><span class="p">[</span><span class="bp">cls</span><span class="p">],</span>
                <span class="n">legend_label</span><span class="o">=</span><span class="nb">str</span><span class="p">(</span><span class="bp">cls</span><span class="p">),</span> 
                <span class="n">alpha</span><span class="o">=</span><span class="mf">0.7</span>
            <span class="p">)</span>
        
        <span class="n">p1</span><span class="o">.</span><span class="n">xaxis</span><span class="o">.</span><span class="n">axis_label</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;Principal Component 1 (</span><span class="si">{</span><span class="n">explained_variance</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="si">:</span><span class="s1">.1%</span><span class="si">}</span><span class="s1"> variance)&#39;</span>
        <span class="n">p1</span><span class="o">.</span><span class="n">yaxis</span><span class="o">.</span><span class="n">axis_label</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;Principal Component 2 (</span><span class="si">{</span><span class="n">explained_variance</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="si">:</span><span class="s1">.1%</span><span class="si">}</span><span class="s1"> variance)&#39;</span>
        <span class="n">p1</span><span class="o">.</span><span class="n">legend</span><span class="o">.</span><span class="n">title</span> <span class="o">=</span> <span class="s1">&#39;Object Class&#39;</span>
        <span class="n">p1</span><span class="o">.</span><span class="n">legend</span><span class="o">.</span><span class="n">location</span> <span class="o">=</span> <span class="s2">&quot;top_right&quot;</span>
        <span class="n">p1</span><span class="o">.</span><span class="n">legend</span><span class="o">.</span><span class="n">click_policy</span> <span class="o">=</span> <span class="s2">&quot;hide&quot;</span>  <span class="c1"># Allow hiding/showing classes by clicking legend</span>
        
        <span class="n">show</span><span class="p">(</span><span class="n">p1</span><span class="p">)</span>

        <span class="c1"># --- PLOT 2: LOADINGS PLOT ---</span>
        <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;</span><span class="se">\n</span><span class="s2">Generating loadings plot...&quot;</span><span class="p">)</span>
        
        <span class="c1"># In scikit-learn, pca.components_ are the loadings (transposed)</span>
        <span class="n">loadings</span> <span class="o">=</span> <span class="n">pca</span><span class="o">.</span><span class="n">components_</span><span class="o">.</span><span class="n">T</span>  <span class="c1"># Shape: (n_features, n_components)</span>
        
        <span class="c1"># Create the loadings plot</span>
        <span class="n">p2</span> <span class="o">=</span> <span class="n">figure</span><span class="p">(</span>
            <span class="n">width</span><span class="o">=</span><span class="mi">800</span><span class="p">,</span> 
            <span class="n">height</span><span class="o">=</span><span class="mi">600</span><span class="p">,</span> 
            <span class="n">title</span><span class="o">=</span><span class="s2">&quot;PCA Loadings Plot - Feature Contributions&quot;</span><span class="p">,</span>
            <span class="n">tools</span><span class="o">=</span><span class="s2">&quot;pan,wheel_zoom,box_zoom,reset,save&quot;</span>
        <span class="p">)</span>
        
        <span class="c1"># Draw arrows from origin to each loading point</span>
        <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">feature</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">features</span><span class="p">):</span>
            <span class="n">x_end</span> <span class="o">=</span> <span class="n">loadings</span><span class="p">[</span><span class="n">i</span><span class="p">,</span> <span class="mi">0</span><span class="p">]</span>
            <span class="n">y_end</span> <span class="o">=</span> <span class="n">loadings</span><span class="p">[</span><span class="n">i</span><span class="p">,</span> <span class="mi">1</span><span class="p">]</span>
            
            <span class="c1"># Draw arrow line</span>
            <span class="n">p2</span><span class="o">.</span><span class="n">line</span><span class="p">([</span><span class="mi">0</span><span class="p">,</span> <span class="n">x_end</span><span class="p">],</span> <span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="n">y_end</span><span class="p">],</span> <span class="n">line_width</span><span class="o">=</span><span class="mi">2</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="s1">&#39;red&#39;</span><span class="p">,</span> <span class="n">alpha</span><span class="o">=</span><span class="mf">0.8</span><span class="p">)</span>
            
            <span class="c1"># Draw arrow head (simple triangle)</span>
            <span class="n">p2</span><span class="o">.</span><span class="n">triangle</span><span class="p">([</span><span class="n">x_end</span><span class="p">],</span> <span class="p">[</span><span class="n">y_end</span><span class="p">],</span> <span class="n">size</span><span class="o">=</span><span class="mi">10</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="s1">&#39;red&#39;</span><span class="p">,</span> <span class="n">alpha</span><span class="o">=</span><span class="mf">0.8</span><span class="p">)</span>
            
            <span class="c1"># Add feature labels</span>
            <span class="n">p2</span><span class="o">.</span><span class="n">text</span><span class="p">(</span>
                <span class="n">x</span><span class="o">=</span><span class="p">[</span><span class="n">x_end</span> <span class="o">*</span> <span class="mf">1.1</span><span class="p">],</span> <span class="n">y</span><span class="o">=</span><span class="p">[</span><span class="n">y_end</span> <span class="o">*</span> <span class="mf">1.1</span><span class="p">],</span> 
                <span class="n">text</span><span class="o">=</span><span class="p">[</span><span class="n">feature</span><span class="p">],</span> 
                <span class="n">text_font_size</span><span class="o">=</span><span class="s2">&quot;12pt&quot;</span><span class="p">,</span> 
                <span class="n">text_align</span><span class="o">=</span><span class="s2">&quot;center&quot;</span><span class="p">,</span>
                <span class="n">text_baseline</span><span class="o">=</span><span class="s2">&quot;middle&quot;</span>
            <span class="p">)</span>

        <span class="c1"># Add grid lines at origin</span>
        <span class="n">p2</span><span class="o">.</span><span class="n">line</span><span class="p">([</span><span class="o">-</span><span class="mi">1</span><span class="p">,</span> <span class="mi">1</span><span class="p">],</span> <span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="mi">0</span><span class="p">],</span> <span class="n">line_color</span><span class="o">=</span><span class="s1">&#39;gray&#39;</span><span class="p">,</span> <span class="n">line_alpha</span><span class="o">=</span><span class="mf">0.3</span><span class="p">)</span>
        <span class="n">p2</span><span class="o">.</span><span class="n">line</span><span class="p">([</span><span class="mi">0</span><span class="p">,</span> <span class="mi">0</span><span class="p">],</span> <span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">,</span> <span class="mi">1</span><span class="p">],</span> <span class="n">line_color</span><span class="o">=</span><span class="s1">&#39;gray&#39;</span><span class="p">,</span> <span class="n">line_alpha</span><span class="o">=</span><span class="mf">0.3</span><span class="p">)</span>
        
        <span class="c1"># Add unit circle for reference</span>
        <span class="n">theta</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">linspace</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="mi">2</span><span class="o">*</span><span class="n">np</span><span class="o">.</span><span class="n">pi</span><span class="p">,</span> <span class="mi">100</span><span class="p">)</span>
        <span class="n">circle_x</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">cos</span><span class="p">(</span><span class="n">theta</span><span class="p">)</span>
        <span class="n">circle_y</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">sin</span><span class="p">(</span><span class="n">theta</span><span class="p">)</span>
        <span class="n">p2</span><span class="o">.</span><span class="n">line</span><span class="p">(</span><span class="n">circle_x</span><span class="p">,</span> <span class="n">circle_y</span><span class="p">,</span> <span class="n">line_color</span><span class="o">=</span><span class="s1">&#39;gray&#39;</span><span class="p">,</span> <span class="n">line_alpha</span><span class="o">=</span><span class="mf">0.3</span><span class="p">,</span> <span class="n">line_dash</span><span class="o">=</span><span class="s1">&#39;dashed&#39;</span><span class="p">)</span>

        <span class="n">p2</span><span class="o">.</span><span class="n">xaxis</span><span class="o">.</span><span class="n">axis_label</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;PC1 (</span><span class="si">{</span><span class="n">explained_variance</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="si">:</span><span class="s1">.1%</span><span class="si">}</span><span class="s1"> variance)&#39;</span>
        <span class="n">p2</span><span class="o">.</span><span class="n">yaxis</span><span class="o">.</span><span class="n">axis_label</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;PC2 (</span><span class="si">{</span><span class="n">explained_variance</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="si">:</span><span class="s1">.1%</span><span class="si">}</span><span class="s1"> variance)&#39;</span>
        
        <span class="c1"># Set equal aspect ratio and appropriate ranges</span>
        <span class="n">max_range</span> <span class="o">=</span> <span class="nb">max</span><span class="p">(</span><span class="nb">abs</span><span class="p">(</span><span class="n">loadings</span><span class="o">.</span><span class="n">max</span><span class="p">()),</span> <span class="nb">abs</span><span class="p">(</span><span class="n">loadings</span><span class="o">.</span><span class="n">min</span><span class="p">()))</span> <span class="o">*</span> <span class="mf">1.2</span>
        <span class="n">p2</span><span class="o">.</span><span class="n">x_range</span><span class="o">.</span><span class="n">start</span> <span class="o">=</span> <span class="o">-</span><span class="n">max_range</span>
        <span class="n">p2</span><span class="o">.</span><span class="n">x_range</span><span class="o">.</span><span class="n">end</span> <span class="o">=</span> <span class="n">max_range</span>
        <span class="n">p2</span><span class="o">.</span><span class="n">y_range</span><span class="o">.</span><span class="n">start</span> <span class="o">=</span> <span class="o">-</span><span class="n">max_range</span>
        <span class="n">p2</span><span class="o">.</span><span class="n">y_range</span><span class="o">.</span><span class="n">end</span> <span class="o">=</span> <span class="n">max_range</span>
        
        <span class="n">show</span><span class="p">(</span><span class="n">p2</span><span class="p">)</span>

        <span class="c1"># --- INTERPRETATION ---</span>
        <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;</span><span class="se">\n</span><span class="s2">&quot;</span> <span class="o">+</span> <span class="s2">&quot;=&quot;</span><span class="o">*</span><span class="mi">60</span><span class="p">)</span>
        <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;INTERPRETATION OF RESULTS:&quot;</span><span class="p">)</span>
        <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;=&quot;</span><span class="o">*</span><span class="mi">60</span><span class="p">)</span>
        
        <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="se">\n</span><span class="s2">PCA Summary:&quot;</span><span class="p">)</span>
        <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;- Total variance explained by first 2 components: </span><span class="si">{</span><span class="n">np</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="n">explained_variance</span><span class="p">)</span><span class="si">:</span><span class="s2">.1%</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
        <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;- PC1 explains </span><span class="si">{</span><span class="n">explained_variance</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="si">:</span><span class="s2">.1%</span><span class="si">}</span><span class="s2"> of variance&quot;</span><span class="p">)</span>
        <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;- PC2 explains </span><span class="si">{</span><span class="n">explained_variance</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="si">:</span><span class="s2">.1%</span><span class="si">}</span><span class="s2"> of variance&quot;</span><span class="p">)</span>
        
        <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="se">\n</span><span class="s2">Loadings (feature contributions):&quot;</span><span class="p">)</span>
        <span class="n">loadings_df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">loadings</span><span class="p">,</span> <span class="n">columns</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;PC1&#39;</span><span class="p">,</span> <span class="s1">&#39;PC2&#39;</span><span class="p">],</span> <span class="n">index</span><span class="o">=</span><span class="n">features</span><span class="p">)</span>
        <span class="nb">print</span><span class="p">(</span><span class="n">loadings_df</span><span class="o">.</span><span class="n">round</span><span class="p">(</span><span class="mi">3</span><span class="p">))</span>
        
        <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="se">\n</span><span class="s2">Top contributors to PC1:&quot;</span><span class="p">)</span>
        <span class="n">pc1_contrib</span> <span class="o">=</span> <span class="nb">abs</span><span class="p">(</span><span class="n">loadings_df</span><span class="p">[</span><span class="s1">&#39;PC1&#39;</span><span class="p">])</span><span class="o">.</span><span class="n">sort_values</span><span class="p">(</span><span class="n">ascending</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
        <span class="k">for</span> <span class="n">feature</span><span class="p">,</span> <span class="n">contrib</span> <span class="ow">in</span> <span class="n">pc1_contrib</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
            <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;  </span><span class="si">{</span><span class="n">feature</span><span class="si">}</span><span class="s2">: </span><span class="si">{</span><span class="n">contrib</span><span class="si">:</span><span class="s2">.3f</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
            
        <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="se">\n</span><span class="s2">Top contributors to PC2:&quot;</span><span class="p">)</span>
        <span class="n">pc2_contrib</span> <span class="o">=</span> <span class="nb">abs</span><span class="p">(</span><span class="n">loadings_df</span><span class="p">[</span><span class="s1">&#39;PC2&#39;</span><span class="p">])</span><span class="o">.</span><span class="n">sort_values</span><span class="p">(</span><span class="n">ascending</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
        <span class="k">for</span> <span class="n">feature</span><span class="p">,</span> <span class="n">contrib</span> <span class="ow">in</span> <span class="n">pc2_contrib</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
            <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;  </span><span class="si">{</span><span class="n">feature</span><span class="si">}</span><span class="s2">: </span><span class="si">{</span><span class="n">contrib</span><span class="si">:</span><span class="s2">.3f</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
    
    <span class="k">else</span><span class="p">:</span>
        <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;</span><span class="se">\n</span><span class="s2">--- ERROR ---&quot;</span><span class="p">)</span>
        <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;The target column &#39;</span><span class="si">{</span><span class="n">target</span><span class="si">}</span><span class="s2">&#39; or one of the feature columns was not found in the data.&quot;</span><span class="p">)</span>
        <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Please ensure your downloaded file contains all necessary columns.&quot;</span><span class="p">)</span>
        <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Required columns: &#39;u&#39;, &#39;g&#39;, &#39;r&#39;, &#39;i&#39;, &#39;z&#39;, &#39;class&#39;&quot;</span><span class="p">)</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="n">sdss_df</span><span class="o">.</span><span class="n">empty</span><span class="p">:</span>
            <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Available columns: </span><span class="si">{</span><span class="n">sdss_df</span><span class="o">.</span><span class="n">columns</span><span class="o">.</span><span class="n">tolist</span><span class="p">()</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>

<span class="k">else</span><span class="p">:</span>
    <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Cannot proceed without data. Please check the file path and try again.&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
</details>
<div class="cell_output docutils container">
<div class="output text_html">    <style>
        .bk-notebook-logo {
            display: block;
            width: 20px;
            height: 20px;
            background-image: url(data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAABQAAAAUCAYAAACNiR0NAAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAALEgAACxIB0t1+/AAAABx0RVh0U29mdHdhcmUAQWRvYmUgRmlyZXdvcmtzIENTNui8sowAAAOkSURBVDiNjZRtaJVlGMd/1/08zzln5zjP1LWcU9N0NkN8m2CYjpgQYQXqSs0I84OLIC0hkEKoPtiH3gmKoiJDU7QpLgoLjLIQCpEsNJ1vqUOdO7ppbuec5+V+rj4ctwzd8IIbbi6u+8f1539dt3A78eXC7QizUF7gyV1fD1Yqg4JWz84yffhm0qkFqBogB9rM8tZdtwVsPUhWhGcFJngGeWrPzHm5oaMmkfEg1usvLFyc8jLRqDOMru7AyC8saQr7GG7f5fvDeH7Ej8CM66nIF+8yngt6HWaKh7k49Soy9nXurCi1o3qUbS3zWfrYeQDTB/Qj6kX6Ybhw4B+bOYoLKCC9H3Nu/leUTZ1JdRWkkn2ldcCamzrcf47KKXdAJllSlxAOkRgyHsGC/zRday5Qld9DyoM4/q/rUoy/CXh3jzOu3bHUVZeU+DEn8FInkPBFlu3+nW3Nw0mk6vCDiWg8CeJaxEwuHS3+z5RgY+YBR6V1Z1nxSOfoaPa4LASWxxdNp+VWTk7+4vzaou8v8PN+xo+KY2xsw6une2frhw05CTYOmQvsEhjhWjn0bmXPjpE1+kplmmkP3suftwTubK9Vq22qKmrBhpY4jvd5afdRA3wGjFAgcnTK2s4hY0/GPNIb0nErGMCRxWOOX64Z8RAC4oCXdklmEvcL8o0BfkNK4lUg9HTl+oPlQxdNo3Mg4Nv175e/1LDGzZen30MEjRUtmXSfiTVu1kK8W4txyV6BMKlbgk3lMwYCiusNy9fVfvvwMxv8Ynl6vxoByANLTWplvuj/nF9m2+PDtt1eiHPBr1oIfhCChQMBw6Aw0UulqTKZdfVvfG7VcfIqLG9bcldL/+pdWTLxLUy8Qq38heUIjh4XlzZxzQm19lLFlr8vdQ97rjZVOLf8nclzckbcD4wxXMidpX30sFd37Fv/GtwwhzhxGVAprjbg0gCAEeIgwCZyTV2Z1REEW8O4py0wsjeloKoMr6iCY6dP92H6Vw/oTyICIthibxjm/DfN9lVz8IqtqKYLUXfoKVMVQVVJOElGjrnnUt9T9wbgp8AyYKaGlqingHZU/uG2NTZSVqwHQTWkx9hxjkpWDaCg6Ckj5qebgBVbT3V3NNXMSiWSDdGV3hrtzla7J+duwPOToIg42ChPQOQjspnSlp1V+Gjdged7+8UN5CRAV7a5EdFNwCjEaBR27b3W890TE7g24NAP/mMDXRWrGoFPQI9ls/MWO2dWFAar/xcOIImbbpA3zgAAAABJRU5ErkJggg==);
        }
    </style>
    <div>
        <a href="https://bokeh.org" target="_blank" class="bk-notebook-logo"></a>
        <span id="d7576407-e53e-4302-8607-a4d065d798d5">Loading BokehJS ...</span>
    </div>
</div><script type="application/javascript">'use strict';
(function(root) {
  function now() {
    return new Date();
  }

  const force = true;

  if (typeof root._bokeh_onload_callbacks === "undefined" || force === true) {
    root._bokeh_onload_callbacks = [];
    root._bokeh_is_loading = undefined;
  }

const JS_MIME_TYPE = 'application/javascript';
  const HTML_MIME_TYPE = 'text/html';
  const EXEC_MIME_TYPE = 'application/vnd.bokehjs_exec.v0+json';
  const CLASS_NAME = 'output_bokeh rendered_html';

  /**
   * Render data to the DOM node
   */
  function render(props, node) {
    const script = document.createElement("script");
    node.appendChild(script);
  }

  /**
   * Handle when an output is cleared or removed
   */
  function handleClearOutput(event, handle) {
    function drop(id) {
      const view = Bokeh.index.get_by_id(id)
      if (view != null) {
        view.model.document.clear()
        Bokeh.index.delete(view)
      }
    }

    const cell = handle.cell;

    const id = cell.output_area._bokeh_element_id;
    const server_id = cell.output_area._bokeh_server_id;

    // Clean up Bokeh references
    if (id != null) {
      drop(id)
    }

    if (server_id !== undefined) {
      // Clean up Bokeh references
      const cmd_clean = "from bokeh.io.state import curstate; print(curstate().uuid_to_server['" + server_id + "'].get_sessions()[0].document.roots[0]._id)";
      cell.notebook.kernel.execute(cmd_clean, {
        iopub: {
          output: function(msg) {
            const id = msg.content.text.trim()
            drop(id)
          }
        }
      });
      // Destroy server and session
      const cmd_destroy = "import bokeh.io.notebook as ion; ion.destroy_server('" + server_id + "')";
      cell.notebook.kernel.execute(cmd_destroy);
    }
  }

  /**
   * Handle when a new output is added
   */
  function handleAddOutput(event, handle) {
    const output_area = handle.output_area;
    const output = handle.output;

    // limit handleAddOutput to display_data with EXEC_MIME_TYPE content only
    if ((output.output_type != "display_data") || (!Object.prototype.hasOwnProperty.call(output.data, EXEC_MIME_TYPE))) {
      return
    }

    const toinsert = output_area.element.find("." + CLASS_NAME.split(' ')[0]);

    if (output.metadata[EXEC_MIME_TYPE]["id"] !== undefined) {
      toinsert[toinsert.length - 1].firstChild.textContent = output.data[JS_MIME_TYPE];
      // store reference to embed id on output_area
      output_area._bokeh_element_id = output.metadata[EXEC_MIME_TYPE]["id"];
    }
    if (output.metadata[EXEC_MIME_TYPE]["server_id"] !== undefined) {
      const bk_div = document.createElement("div");
      bk_div.innerHTML = output.data[HTML_MIME_TYPE];
      const script_attrs = bk_div.children[0].attributes;
      for (let i = 0; i < script_attrs.length; i++) {
        toinsert[toinsert.length - 1].firstChild.setAttribute(script_attrs[i].name, script_attrs[i].value);
        toinsert[toinsert.length - 1].firstChild.textContent = bk_div.children[0].textContent
      }
      // store reference to server id on output_area
      output_area._bokeh_server_id = output.metadata[EXEC_MIME_TYPE]["server_id"];
    }
  }

  function register_renderer(events, OutputArea) {

    function append_mime(data, metadata, element) {
      // create a DOM node to render to
      const toinsert = this.create_output_subarea(
        metadata,
        CLASS_NAME,
        EXEC_MIME_TYPE
      );
      this.keyboard_manager.register_events(toinsert);
      // Render to node
      const props = {data: data, metadata: metadata[EXEC_MIME_TYPE]};
      render(props, toinsert[toinsert.length - 1]);
      element.append(toinsert);
      return toinsert
    }

    /* Handle when an output is cleared or removed */
    events.on('clear_output.CodeCell', handleClearOutput);
    events.on('delete.Cell', handleClearOutput);

    /* Handle when a new output is added */
    events.on('output_added.OutputArea', handleAddOutput);

    /**
     * Register the mime type and append_mime function with output_area
     */
    OutputArea.prototype.register_mime_type(EXEC_MIME_TYPE, append_mime, {
      /* Is output safe? */
      safe: true,
      /* Index of renderer in `output_area.display_order` */
      index: 0
    });
  }

  // register the mime type if in Jupyter Notebook environment and previously unregistered
  if (root.Jupyter !== undefined) {
    const events = require('base/js/events');
    const OutputArea = require('notebook/js/outputarea').OutputArea;

    if (OutputArea.prototype.mime_types().indexOf(EXEC_MIME_TYPE) == -1) {
      register_renderer(events, OutputArea);
    }
  }
  if (typeof (root._bokeh_timeout) === "undefined" || force === true) {
    root._bokeh_timeout = Date.now() + 5000;
    root._bokeh_failed_load = false;
  }

  const NB_LOAD_WARNING = {'data': {'text/html':
     "<div style='background-color: #fdd'>\n"+
     "<p>\n"+
     "BokehJS does not appear to have successfully loaded. If loading BokehJS from CDN, this \n"+
     "may be due to a slow or bad network connection. Possible fixes:\n"+
     "</p>\n"+
     "<ul>\n"+
     "<li>re-rerun `output_notebook()` to attempt to load from CDN again, or</li>\n"+
     "<li>use INLINE resources instead, as so:</li>\n"+
     "</ul>\n"+
     "<code>\n"+
     "from bokeh.resources import INLINE\n"+
     "output_notebook(resources=INLINE)\n"+
     "</code>\n"+
     "</div>"}};

  function display_loaded(error = null) {
    const el = document.getElementById("d7576407-e53e-4302-8607-a4d065d798d5");
    if (el != null) {
      const html = (() => {
        if (typeof root.Bokeh === "undefined") {
          if (error == null) {
            return "BokehJS is loading ...";
          } else {
            return "BokehJS failed to load.";
          }
        } else {
          const prefix = `BokehJS ${root.Bokeh.version}`;
          if (error == null) {
            return `${prefix} successfully loaded.`;
          } else {
            return `${prefix} <b>encountered errors</b> while loading and may not function as expected.`;
          }
        }
      })();
      el.innerHTML = html;

      if (error != null) {
        const wrapper = document.createElement("div");
        wrapper.style.overflow = "auto";
        wrapper.style.height = "5em";
        wrapper.style.resize = "vertical";
        const content = document.createElement("div");
        content.style.fontFamily = "monospace";
        content.style.whiteSpace = "pre-wrap";
        content.style.backgroundColor = "rgb(255, 221, 221)";
        content.textContent = error.stack ?? error.toString();
        wrapper.append(content);
        el.append(wrapper);
      }
    } else if (Date.now() < root._bokeh_timeout) {
      setTimeout(() => display_loaded(error), 100);
    }
  }

  function run_callbacks() {
    try {
      root._bokeh_onload_callbacks.forEach(function(callback) {
        if (callback != null)
          callback();
      });
    } finally {
      delete root._bokeh_onload_callbacks
    }
    console.debug("Bokeh: all callbacks have finished");
  }

  function load_libs(css_urls, js_urls, callback) {
    if (css_urls == null) css_urls = [];
    if (js_urls == null) js_urls = [];

    root._bokeh_onload_callbacks.push(callback);
    if (root._bokeh_is_loading > 0) {
      console.debug("Bokeh: BokehJS is being loaded, scheduling callback at", now());
      return null;
    }
    if (js_urls == null || js_urls.length === 0) {
      run_callbacks();
      return null;
    }
    console.debug("Bokeh: BokehJS not loaded, scheduling load and callback at", now());
    root._bokeh_is_loading = css_urls.length + js_urls.length;

    function on_load() {
      root._bokeh_is_loading--;
      if (root._bokeh_is_loading === 0) {
        console.debug("Bokeh: all BokehJS libraries/stylesheets loaded");
        run_callbacks()
      }
    }

    function on_error(url) {
      console.error("failed to load " + url);
    }

    for (let i = 0; i < css_urls.length; i++) {
      const url = css_urls[i];
      const element = document.createElement("link");
      element.onload = on_load;
      element.onerror = on_error.bind(null, url);
      element.rel = "stylesheet";
      element.type = "text/css";
      element.href = url;
      console.debug("Bokeh: injecting link tag for BokehJS stylesheet: ", url);
      document.body.appendChild(element);
    }

    for (let i = 0; i < js_urls.length; i++) {
      const url = js_urls[i];
      const element = document.createElement('script');
      element.onload = on_load;
      element.onerror = on_error.bind(null, url);
      element.async = false;
      element.src = url;
      console.debug("Bokeh: injecting script tag for BokehJS library: ", url);
      document.head.appendChild(element);
    }
  };

  function inject_raw_css(css) {
    const element = document.createElement("style");
    element.appendChild(document.createTextNode(css));
    document.body.appendChild(element);
  }

  const js_urls = ["https://cdn.bokeh.org/bokeh/release/bokeh-3.8.0.min.js", "https://cdn.bokeh.org/bokeh/release/bokeh-gl-3.8.0.min.js", "https://cdn.bokeh.org/bokeh/release/bokeh-widgets-3.8.0.min.js", "https://cdn.bokeh.org/bokeh/release/bokeh-tables-3.8.0.min.js", "https://cdn.bokeh.org/bokeh/release/bokeh-mathjax-3.8.0.min.js"];
  const css_urls = [];

  const inline_js = [    function(Bokeh) {
      Bokeh.set_log_level("info");
    },
function(Bokeh) {
    }
  ];

  function run_inline_js() {
    if (root.Bokeh !== undefined || force === true) {
      try {
            for (let i = 0; i < inline_js.length; i++) {
      inline_js[i].call(root, root.Bokeh);
    }

      } catch (error) {display_loaded(error);throw error;
      }if (force === true) {
        display_loaded();
      }} else if (Date.now() < root._bokeh_timeout) {
      setTimeout(run_inline_js, 100);
    } else if (!root._bokeh_failed_load) {
      console.log("Bokeh: BokehJS failed to load within specified timeout.");
      root._bokeh_failed_load = true;
    } else if (force !== true) {
      const cell = $(document.getElementById("d7576407-e53e-4302-8607-a4d065d798d5")).parents('.cell').data().cell;
      cell.output_area.append_execute_result(NB_LOAD_WARNING)
    }
  }

  if (root._bokeh_is_loading === 0) {
    console.debug("Bokeh: BokehJS loaded, going straight to plotting");
    run_inline_js();
  } else {
    load_libs(css_urls, js_urls, function() {
      console.debug("Bokeh: BokehJS plotting callback run at", now());
      run_inline_js();
    });
  }
}(window));</script><div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>File not found: &#39;skyserverdata2.csv&#39;. Please make sure you have downloaded the data and placed it in the correct directory.
Cannot proceed without data. Please check the file path and try again.
</pre></div>
</div>
</div>
</div>
<div class="exercise admonition" id="lectures/03-PCA/Intro-PCA-exercise-1">

<p class="admonition-title"><span class="caption-number">Exercise 4.2 </span> (The Importance of Standardization)</p>
<section id="exercise-content">
<p>Take the original Python script and find the line where the data is scaled. Comment it and perform again the PCA analysis. Is there any consequence? Replot and analyze. How much variance is capture now by PC1 and PC2? Do you still see the quasars separation in the scores plot?</p>
</section>
</div>
<div class="exercise admonition" id="lectures/03-PCA/Intro-PCA-exercise-2">

<p class="admonition-title"><span class="caption-number">Exercise 4.3 </span> (Incorporating a Physical Measurement)</p>
<section id="exercise-content">
<p>Investigate how adding a non-color, physical feature like ‘redshift’ changes the PCA results and the interpretability of the components. Redshift is a measure of how much an object’s light has been stretched as it travels through the expanding universe; it is strongly correlated with distance…  In the original script, add ‘redshift’ to the list of features to be included in the analysis. Re-run the script.</p>
<ul class="simple">
<li><p>What is the new total explained variance captured by PC1 and PC2? Has it increased or decreased?</p></li>
<li><p>Look at the new Scores Plot. Is the separation between the three classes (STAR, GALAXY, QSO) better or worse than in the original plot? Pay close attention to the Star/Galaxy overlap.</p></li>
<li><p>Examine the new Loadings Plot. Where does the new ‘redshift’ arrow point? What does its direction and length tell you about its contribution to PC1 and PC2?</p></li>
<li><p>Based on the new loadings, what do PC1 and PC2 now represent? Has the interpretation of “overall brightness” or “color contrast” changed? What physical reason might explain why adding redshift was so effective (or ineffective) at separating the classes?</p></li>
</ul>
</section>
</div>
<div class="exercise admonition" id="lectures/03-PCA/Intro-PCA-exercise-3">

<p class="admonition-title"><span class="caption-number">Exercise 4.4 </span> (What about more components?)</p>
<section id="exercise-content">
<ul class="simple">
<li><p>Plot the scores for the first 3 and 4 principal components.</p></li>
<li><p>Analyze how the classes (STAR, GALAXY, QSO) are distributed in the 3D space</p></li>
<li><p>Calculate the cumulative explained variance for the first 3 or 4 components. How much of the total variance is explained by these components?</p></li>
<li><p>Does adding more components improves the separation between classes and whether the additional components add significant information?</p></li>
</ul>
</section>
</div>
<div class="exercise admonition" id="lectures/03-PCA/Intro-PCA-exercise-4">

<p class="admonition-title"><span class="caption-number">Exercise 4.5 </span> (Compare the performance of PCA with other dimensionality reduction methods like t-SNE or UMAP.)</p>
<section id="exercise-content">
<ul class="simple">
<li><p>Apply t-SNE or UMAP to the original feature set (without PCA) and visualize the results.</p></li>
<li><p>Compare the ability of t-SNE/UMAP to separate the classes with PCA.</p></li>
<li><p>Discuss the advantages and limitations of PCA versus non-linear dimensionality reduction methods like t-SNE and UMAP.</p></li>
</ul>
</section>
</div>
</section>
<section id="interpretation-guidelines">
<h2><span class="section-number">4.8. </span>Interpretation Guidelines<a class="headerlink" href="#interpretation-guidelines" title="Link to this heading">#</a></h2>
<section id="choosing-number-of-components">
<h3><span class="section-number">4.8.1. </span>Choosing Number of Components<a class="headerlink" href="#choosing-number-of-components" title="Link to this heading">#</a></h3>
<p><strong>Method 1: Cumulative Variance</strong></p>
<ul class="simple">
<li><p>Keep components explaining 80-95% of variance</p></li>
<li><p>Depends on application requirements</p></li>
</ul>
<p><strong>Method 2: Scree Plot (Elbow Method)</strong></p>
<ul class="simple">
<li><p>Plot eigenvalues vs. component number</p></li>
<li><p>Look for “elbow” where decrease slows</p></li>
</ul>
<p><strong>Method 3: Kaiser Criterion</strong></p>
<ul class="simple">
<li><p>Keep components with eigenvalues &gt; 1 (if data is standardized)</p></li>
<li><p>Only applies when variables are standardized</p></li>
</ul>
</section>
<section id="loading-interpretation">
<h3><span class="section-number">4.8.2. </span>Loading Interpretation<a class="headerlink" href="#loading-interpretation" title="Link to this heading">#</a></h3>
<p><strong>Loading Values</strong>:</p>
<ul class="simple">
<li><p><strong>|loading| &gt; 0.7</strong>: Strong relationship</p></li>
<li><p><strong>0.3 &lt; |loading| &lt; 0.7</strong>: Moderate relationship</p></li>
<li><p><strong>|loading| &lt; 0.3</strong>: Weak relationship</p></li>
</ul>
<p><strong>Signs Matter</strong>:</p>
<ul class="simple">
<li><p><strong>Positive loading</strong>: Variable increases with PC</p></li>
<li><p><strong>Negative loading</strong>: Variable decreases with PC</p></li>
</ul>
</section>
</section>
<section id="common-pitfalls">
<h2><span class="section-number">4.9. </span>Common Pitfalls<a class="headerlink" href="#common-pitfalls" title="Link to this heading">#</a></h2>
<div class="pst-scrollable-table-container"><table class="table">
<thead>
<tr class="row-odd"><th class="head"><p><strong>Mistake</strong></p></th>
<th class="head"><p><strong>Correct</strong></p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p>Treating PCs as original variables</p></td>
<td><p>PCs are linear combinations of original variables</p></td>
</tr>
<tr class="row-odd"><td><p>Ignoring scaling/standardization</p></td>
<td><p>Consider whether to standardize based on variable units</p></td>
</tr>
<tr class="row-even"><td><p>Over-interpreting small components</p></td>
<td><p>Focus on components explaining meaningful variance</p></td>
</tr>
</tbody>
</table>
</div>
</section>
<section id="advanced-details">
<h2><span class="section-number">4.10. </span>Advanced Details<a class="headerlink" href="#advanced-details" title="Link to this heading">#</a></h2>
<section id="pca-assumptions-and-limitations">
<h3><span class="section-number">4.10.1. </span>PCA Assumptions and Limitations<a class="headerlink" href="#pca-assumptions-and-limitations" title="Link to this heading">#</a></h3>
<p><strong>Assumptions</strong>:</p>
<ul class="simple">
<li><p>Linear relationships between variables</p></li>
<li><p>Data follows (approximately) multivariate normal distribution</p></li>
<li><p>Variables are continuous</p></li>
</ul>
<p><strong>Limitations</strong>:</p>
<ul class="simple">
<li><p><strong>Linear combinations only</strong>: Cannot capture nonlinear patterns</p></li>
<li><p><strong>Variance-based</strong>: May not preserve class separability</p></li>
<li><p><strong>Global method</strong>: Same transformation for all data points</p></li>
</ul>
<p><strong>When PCA May Not Work Well</strong>:</p>
<ul class="simple">
<li><p>Highly nonlinear data (consider kernel PCA)</p></li>
<li><p>Categorical variables (consider correspondence analysis)</p></li>
<li><p>When rare events are important (PCA focuses on major patterns)</p></li>
</ul>
</section>
<section id="robust-pca">
<h3><span class="section-number">4.10.2. </span>Robust PCA<a class="headerlink" href="#robust-pca" title="Link to this heading">#</a></h3>
<p><strong>Problem</strong>: Standard PCA is sensitive to outliers.</p>
<p><strong>Solutions</strong>:</p>
<ul class="simple">
<li><p><strong>Robust PCA</strong>: Use median instead of mean, robust covariance estimation</p></li>
<li><p><strong>Sparse PCA</strong>: Assume many loadings are exactly zero</p></li>
<li><p><strong>Kernel PCA</strong>: Handle nonlinear relationships</p></li>
</ul>
</section>
</section>
<hr class="docutils" />
<section id="a-detailed-worked-example-the-iris-dataset">
<h2><span class="section-number">4.11. </span>A detailed Worked Example - The Iris Dataset<a class="headerlink" href="#a-detailed-worked-example-the-iris-dataset" title="Link to this heading">#</a></h2>
<p>The Iris dataset is a classic in machine learning and statistics. It contains 150 samples from three species of Iris flowers (Setosa, Versicolor, and Virginica). For each sample, four features were measured: sepal length, sepal width, petal length, and petal width.</p>
<p>Our goal is to see if we can use PCA to ‘summarize’ these four measurements and visualize the separation between the species.</p>
<section id="setup-importing-libraries">
<h3><span class="section-number">4.11.1. </span>Setup - Importing Libraries<a class="headerlink" href="#setup-importing-libraries" title="Link to this heading">#</a></h3>
<p>First, let’s import all the necessary libraries for data manipulation, numerical computation, and plotting.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span><span class="w"> </span><span class="nn">pandas</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">pd</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">numpy</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">np</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">matplotlib.pyplot</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">plt</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">seaborn</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">sns</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">sklearn.preprocessing</span><span class="w"> </span><span class="kn">import</span> <span class="n">StandardScaler</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">sklearn.decomposition</span><span class="w"> </span><span class="kn">import</span> <span class="n">PCA</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">sklearn.datasets</span><span class="w"> </span><span class="kn">import</span> <span class="n">load_iris</span>

<span class="c1"># Set some default plotting styles for better looking visuals</span>
<span class="n">sns</span><span class="o">.</span><span class="n">set_style</span><span class="p">(</span><span class="s1">&#39;whitegrid&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
</section>
<section id="loading-data">
<h3><span class="section-number">4.11.2. </span>Loading data<a class="headerlink" href="#loading-data" title="Link to this heading">#</a></h3>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Load the data</span>
<span class="n">iris</span> <span class="o">=</span> <span class="n">load_iris</span><span class="p">()</span>
<span class="n">X</span> <span class="o">=</span> <span class="n">iris</span><span class="o">.</span><span class="n">data</span> <span class="c1"># The feature matrix</span>
<span class="n">y</span> <span class="o">=</span> <span class="n">iris</span><span class="o">.</span><span class="n">target</span> <span class="c1"># The labels (species)</span>
<span class="n">feature_names</span> <span class="o">=</span> <span class="n">iris</span><span class="o">.</span><span class="n">feature_names</span>
<span class="n">target_names</span> <span class="o">=</span> <span class="n">iris</span><span class="o">.</span><span class="n">target_names</span>

<span class="c1"># Let&#39;s look at the first 5 rows of the data</span>
<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Feature Names:&quot;</span><span class="p">,</span> <span class="n">feature_names</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;</span><span class="se">\n</span><span class="s2">First 5 rows of data:</span><span class="se">\n</span><span class="s2">&quot;</span><span class="p">,</span> <span class="n">X</span><span class="p">[:</span><span class="mi">5</span><span class="p">])</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Feature Names: [&#39;sepal length (cm)&#39;, &#39;sepal width (cm)&#39;, &#39;petal length (cm)&#39;, &#39;petal width (cm)&#39;]

First 5 rows of data:
 [[5.1 3.5 1.4 0.2]
 [4.9 3.  1.4 0.2]
 [4.7 3.2 1.3 0.2]
 [4.6 3.1 1.5 0.2]
 [5.  3.6 1.4 0.2]]
</pre></div>
</div>
</div>
</div>
</section>
<section id="step-1-standardize-the-data">
<h3><span class="section-number">4.11.3. </span>Step 1: Standardize the Data<a class="headerlink" href="#step-1-standardize-the-data" title="Link to this heading">#</a></h3>
<p>PCA is sensitive to the scale of the features. We need to standardize our data so that each feature has a mean of 0 and a standard deviation of 1.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">X_scaled</span> <span class="o">=</span> <span class="n">StandardScaler</span><span class="p">()</span><span class="o">.</span><span class="n">fit_transform</span><span class="p">(</span><span class="n">X</span><span class="p">)</span>

<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;First 5 rows of scaled data:</span><span class="se">\n</span><span class="s2">&quot;</span><span class="p">,</span> <span class="n">X_scaled</span><span class="p">[:</span><span class="mi">5</span><span class="p">])</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>First 5 rows of scaled data:
 [[-0.90068117  1.01900435 -1.34022653 -1.3154443 ]
 [-1.14301691 -0.13197948 -1.34022653 -1.3154443 ]
 [-1.38535265  0.32841405 -1.39706395 -1.3154443 ]
 [-1.50652052  0.09821729 -1.2833891  -1.3154443 ]
 [-1.02184904  1.24920112 -1.34022653 -1.3154443 ]]
</pre></div>
</div>
</div>
</div>
</section>
<section id="step-2-perform-pca">
<h3><span class="section-number">4.11.4. </span>Step 2: Perform PCA<a class="headerlink" href="#step-2-perform-pca" title="Link to this heading">#</a></h3>
<p>Now we apply PCA. We’ll ask <code class="docutils literal notranslate"><span class="pre">scikit-learn</span></code> to find the first two principal components.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Initialize PCA and fit the scaled data</span>
<span class="c1"># n_components specifies how many dimensions we want to reduce to.</span>
<span class="n">pca</span> <span class="o">=</span> <span class="n">PCA</span><span class="p">(</span><span class="n">n_components</span><span class="o">=</span><span class="mi">2</span><span class="p">)</span>

<span class="c1"># Fit the model and transform the data to the new coordinate system</span>
<span class="n">X_pca</span> <span class="o">=</span> <span class="n">pca</span><span class="o">.</span><span class="n">fit_transform</span><span class="p">(</span><span class="n">X_scaled</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
</section>
<section id="step-3-analyze-and-visualize-the-results">
<h3><span class="section-number">4.11.5. </span>Step 3: Analyze and Visualize the Results<a class="headerlink" href="#step-3-analyze-and-visualize-the-results" title="Link to this heading">#</a></h3>
<section id="scree-plot">
<h4><span class="section-number">4.11.5.1. </span>Scree Plot<a class="headerlink" href="#scree-plot" title="Link to this heading">#</a></h4>
<p>First, let’s see how much variance our two new components capture. A scree plot is perfect for this.</p>
<div class="cell tag_hide-input docutils container">
<details class="admonition hide above-input">
<summary aria-label="Toggle hidden content">
<p class="collapsed admonition-title">Show code cell source</p>
<p class="expanded admonition-title">Hide code cell source</p>
</summary>
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">explained_variance</span> <span class="o">=</span> <span class="n">pca</span><span class="o">.</span><span class="n">explained_variance_ratio_</span>

<span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Variance explained by PC1: </span><span class="si">{</span><span class="n">explained_variance</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="si">:</span><span class="s2">.2%</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Variance explained by PC2: </span><span class="si">{</span><span class="n">explained_variance</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="si">:</span><span class="s2">.2%</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Total variance explained by first two components: </span><span class="si">{</span><span class="n">np</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="n">explained_variance</span><span class="p">)</span><span class="si">:</span><span class="s2">.2%</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>

<span class="c1"># To make a full scree plot, we can re-run PCA without specifying n_components</span>
<span class="n">pca_full</span> <span class="o">=</span> <span class="n">PCA</span><span class="p">()</span><span class="o">.</span><span class="n">fit</span><span class="p">(</span><span class="n">X_scaled</span><span class="p">)</span>

<span class="n">plt</span><span class="o">.</span><span class="n">figure</span><span class="p">(</span><span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">8</span><span class="p">,</span> <span class="mi">6</span><span class="p">))</span>
<span class="n">plt</span><span class="o">.</span><span class="n">bar</span><span class="p">(</span><span class="nb">range</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="nb">len</span><span class="p">(</span><span class="n">pca_full</span><span class="o">.</span><span class="n">explained_variance_ratio_</span><span class="p">)</span> <span class="o">+</span> <span class="mi">1</span><span class="p">),</span> <span class="n">pca_full</span><span class="o">.</span><span class="n">explained_variance_ratio_</span> <span class="o">*</span> <span class="mi">100</span><span class="p">,</span> <span class="n">alpha</span><span class="o">=</span><span class="mf">0.7</span><span class="p">,</span> <span class="n">align</span><span class="o">=</span><span class="s1">&#39;center&#39;</span><span class="p">,</span> <span class="n">label</span><span class="o">=</span><span class="s1">&#39;Individual explained variance&#39;</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">step</span><span class="p">(</span><span class="nb">range</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="nb">len</span><span class="p">(</span><span class="n">pca_full</span><span class="o">.</span><span class="n">explained_variance_ratio_</span><span class="p">)</span> <span class="o">+</span> <span class="mi">1</span><span class="p">),</span> <span class="n">np</span><span class="o">.</span><span class="n">cumsum</span><span class="p">(</span><span class="n">pca_full</span><span class="o">.</span><span class="n">explained_variance_ratio_</span><span class="p">)</span> <span class="o">*</span> <span class="mi">100</span><span class="p">,</span> <span class="n">where</span><span class="o">=</span><span class="s1">&#39;mid&#39;</span><span class="p">,</span> <span class="n">label</span><span class="o">=</span><span class="s1">&#39;Cumulative explained variance&#39;</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">ylabel</span><span class="p">(</span><span class="s1">&#39;Explained Variance Percentage&#39;</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">xlabel</span><span class="p">(</span><span class="s1">&#39;Principal Component Index&#39;</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">title</span><span class="p">(</span><span class="s1">&#39;Scree Plot for Iris Dataset&#39;</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">xticks</span><span class="p">(</span><span class="nb">range</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="nb">len</span><span class="p">(</span><span class="n">pca_full</span><span class="o">.</span><span class="n">explained_variance_ratio_</span><span class="p">)</span> <span class="o">+</span> <span class="mi">1</span><span class="p">))</span>
<span class="n">plt</span><span class="o">.</span><span class="n">legend</span><span class="p">(</span><span class="n">loc</span><span class="o">=</span><span class="s1">&#39;best&#39;</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>
</pre></div>
</div>
</div>
</details>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Variance explained by PC1: 72.96%
Variance explained by PC2: 22.85%
Total variance explained by first two components: 95.81%
</pre></div>
</div>
<img alt="../../_images/da0c089067024b8ff34e8e6afb7a9764ad483da2ce1ecaba4a793bf676057715.png" src="../../_images/da0c089067024b8ff34e8e6afb7a9764ad483da2ce1ecaba4a793bf676057715.png" />
</div>
</div>
<p><strong>Observation:</strong> The first two components capture over 95% of the total variance in the data! This means our 2D plot will be a very good representation of the original 4D data.</p>
</section>
<section id="scores-plot">
<h4><span class="section-number">4.11.5.2. </span>Scores Plot<a class="headerlink" href="#scores-plot" title="Link to this heading">#</a></h4>
<p>Next, we create a scores plot. This is a scatter plot of our samples in the new PCA space. We will color each point according to its true species.</p>
<div class="cell tag_hide-input docutils container">
<details class="admonition hide above-input">
<summary aria-label="Toggle hidden content">
<p class="collapsed admonition-title">Show code cell source</p>
<p class="expanded admonition-title">Hide code cell source</p>
</summary>
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">plt</span><span class="o">.</span><span class="n">figure</span><span class="p">(</span><span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">10</span><span class="p">,</span> <span class="mi">8</span><span class="p">))</span>
<span class="n">colors</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;navy&#39;</span><span class="p">,</span> <span class="s1">&#39;turquoise&#39;</span><span class="p">,</span> <span class="s1">&#39;darkorange&#39;</span><span class="p">]</span>

<span class="k">for</span> <span class="n">color</span><span class="p">,</span> <span class="n">i</span><span class="p">,</span> <span class="n">target_name</span> <span class="ow">in</span> <span class="nb">zip</span><span class="p">(</span><span class="n">colors</span><span class="p">,</span> <span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">2</span><span class="p">],</span> <span class="n">target_names</span><span class="p">):</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">scatter</span><span class="p">(</span><span class="n">X_pca</span><span class="p">[</span><span class="n">y</span> <span class="o">==</span> <span class="n">i</span><span class="p">,</span> <span class="mi">0</span><span class="p">],</span> <span class="n">X_pca</span><span class="p">[</span><span class="n">y</span> <span class="o">==</span> <span class="n">i</span><span class="p">,</span> <span class="mi">1</span><span class="p">],</span> <span class="n">color</span><span class="o">=</span><span class="n">color</span><span class="p">,</span> <span class="n">alpha</span><span class="o">=</span><span class="mf">.8</span><span class="p">,</span> <span class="n">lw</span><span class="o">=</span><span class="mi">2</span><span class="p">,</span>
                <span class="n">label</span><span class="o">=</span><span class="n">target_name</span><span class="p">)</span>

<span class="n">plt</span><span class="o">.</span><span class="n">legend</span><span class="p">(</span><span class="n">loc</span><span class="o">=</span><span class="s1">&#39;best&#39;</span><span class="p">,</span> <span class="n">shadow</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">scatterpoints</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">title</span><span class="p">(</span><span class="s1">&#39;PCA Scores Plot of Iris Dataset&#39;</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">xlabel</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;Principal Component 1 (</span><span class="si">{</span><span class="n">explained_variance</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="si">:</span><span class="s1">.2%</span><span class="si">}</span><span class="s1">)&#39;</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">ylabel</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;Principal Component 2 (</span><span class="si">{</span><span class="n">explained_variance</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="si">:</span><span class="s1">.2%</span><span class="si">}</span><span class="s1">)&#39;</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>
</pre></div>
</div>
</div>
</details>
<div class="cell_output docutils container">
<img alt="../../_images/27a2233f498cc0c593f49092632689facc0ac5d29ec66e674e4ae6176118e4af.png" src="../../_images/27a2233f498cc0c593f49092632689facc0ac5d29ec66e674e4ae6176118e4af.png" />
</div>
</div>
<p><strong>Observation:</strong> The three species are very well separated in the PCA plot. The <em>Setosa</em> species is a distinct cluster, while <em>Versicolor</em> and <em>Virginica</em> are also mostly separated, though they have some overlap.</p>
</section>
<section id="biplot">
<h4><span class="section-number">4.11.5.3. </span>Biplot<a class="headerlink" href="#biplot" title="Link to this heading">#</a></h4>
<p>Finally, the biplot helps us understand <em>why</em> the samples are separated. It overlays the original feature vectors (loadings) on top of the scores plot. This tells us how the original variables contribute to the principal components.</p>
<p>We will use a helper function to create a clean biplot.</p>
<div class="cell tag_hide-input docutils container">
<details class="admonition hide above-input">
<summary aria-label="Toggle hidden content">
<p class="collapsed admonition-title">Show code cell source</p>
<p class="expanded admonition-title">Hide code cell source</p>
</summary>
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="k">def</span><span class="w"> </span><span class="nf">biplot</span><span class="p">(</span><span class="n">score</span><span class="p">,</span> <span class="n">coeff</span><span class="p">,</span> <span class="n">labels</span><span class="o">=</span><span class="kc">None</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Creates a biplot visualization.</span>
<span class="sd">    </span>
<span class="sd">    score: The transformed data (scores), e.g., X_pca.</span>
<span class="sd">    coeff: The eigenvectors (loadings), e.g., pca.components_.T.</span>
<span class="sd">    labels: The names of the original features.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">figure</span><span class="p">(</span><span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">12</span><span class="p">,</span> <span class="mi">10</span><span class="p">))</span>
    <span class="n">xs</span> <span class="o">=</span> <span class="n">score</span><span class="p">[:,</span> <span class="mi">0</span><span class="p">]</span>
    <span class="n">ys</span> <span class="o">=</span> <span class="n">score</span><span class="p">[:,</span> <span class="mi">1</span><span class="p">]</span>
    <span class="n">n</span> <span class="o">=</span> <span class="n">coeff</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
    
    <span class="c1"># Plot the scores</span>
    <span class="k">for</span> <span class="n">color</span><span class="p">,</span> <span class="n">i</span><span class="p">,</span> <span class="n">target_name</span> <span class="ow">in</span> <span class="nb">zip</span><span class="p">(</span><span class="n">colors</span><span class="p">,</span> <span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">2</span><span class="p">],</span> <span class="n">target_names</span><span class="p">):</span>
        <span class="n">plt</span><span class="o">.</span><span class="n">scatter</span><span class="p">(</span><span class="n">xs</span><span class="p">[</span><span class="n">y</span> <span class="o">==</span> <span class="n">i</span><span class="p">],</span> <span class="n">ys</span><span class="p">[</span><span class="n">y</span> <span class="o">==</span> <span class="n">i</span><span class="p">],</span> <span class="n">color</span><span class="o">=</span><span class="n">color</span><span class="p">,</span> <span class="n">alpha</span><span class="o">=</span><span class="mf">0.7</span><span class="p">,</span> <span class="n">label</span><span class="o">=</span><span class="n">target_name</span><span class="p">)</span>

    <span class="c1"># Plot the loadings</span>
    <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">n</span><span class="p">):</span>
        <span class="n">plt</span><span class="o">.</span><span class="n">arrow</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="n">coeff</span><span class="p">[</span><span class="n">i</span><span class="p">,</span> <span class="mi">0</span><span class="p">]</span><span class="o">*</span><span class="mi">4</span><span class="p">,</span> <span class="n">coeff</span><span class="p">[</span><span class="n">i</span><span class="p">,</span> <span class="mi">1</span><span class="p">]</span><span class="o">*</span><span class="mi">4</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="s1">&#39;r&#39;</span><span class="p">,</span> <span class="n">alpha</span><span class="o">=</span><span class="mf">0.9</span><span class="p">,</span> <span class="n">head_width</span><span class="o">=</span><span class="mf">0.05</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">labels</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
            <span class="n">plt</span><span class="o">.</span><span class="n">text</span><span class="p">(</span><span class="n">coeff</span><span class="p">[</span><span class="n">i</span><span class="p">,</span> <span class="mi">0</span><span class="p">]</span> <span class="o">*</span> <span class="mf">4.2</span><span class="p">,</span> <span class="n">coeff</span><span class="p">[</span><span class="n">i</span><span class="p">,</span> <span class="mi">1</span><span class="p">]</span> <span class="o">*</span> <span class="mf">4.2</span><span class="p">,</span> <span class="s2">&quot;Var&quot;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">i</span> <span class="o">+</span> <span class="mi">1</span><span class="p">),</span> <span class="n">color</span><span class="o">=</span><span class="s1">&#39;black&#39;</span><span class="p">,</span> <span class="n">ha</span><span class="o">=</span><span class="s1">&#39;center&#39;</span><span class="p">,</span> <span class="n">va</span><span class="o">=</span><span class="s1">&#39;center&#39;</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">plt</span><span class="o">.</span><span class="n">text</span><span class="p">(</span><span class="n">coeff</span><span class="p">[</span><span class="n">i</span><span class="p">,</span> <span class="mi">0</span><span class="p">]</span> <span class="o">*</span> <span class="mf">4.2</span><span class="p">,</span> <span class="n">coeff</span><span class="p">[</span><span class="n">i</span><span class="p">,</span> <span class="mi">1</span><span class="p">]</span> <span class="o">*</span> <span class="mf">4.2</span><span class="p">,</span> <span class="n">labels</span><span class="p">[</span><span class="n">i</span><span class="p">],</span> <span class="n">color</span><span class="o">=</span><span class="s1">&#39;black&#39;</span><span class="p">,</span> <span class="n">ha</span><span class="o">=</span><span class="s1">&#39;center&#39;</span><span class="p">,</span> <span class="n">va</span><span class="o">=</span><span class="s1">&#39;center&#39;</span><span class="p">,</span> <span class="n">fontsize</span><span class="o">=</span><span class="mi">12</span><span class="p">)</span>
    
    <span class="n">plt</span><span class="o">.</span><span class="n">xlabel</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;Principal Component 1 (</span><span class="si">{</span><span class="n">explained_variance</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="si">:</span><span class="s1">.2%</span><span class="si">}</span><span class="s1">)&#39;</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">ylabel</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;Principal Component 2 (</span><span class="si">{</span><span class="n">explained_variance</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="si">:</span><span class="s1">.2%</span><span class="si">}</span><span class="s1">)&#39;</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">title</span><span class="p">(</span><span class="s1">&#39;Biplot of Iris Dataset&#39;</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">legend</span><span class="p">()</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">grid</span><span class="p">()</span>

<span class="c1"># Call the function with our data</span>
<span class="c1"># Note: we need to transpose pca.components_ to get the loadings in the right shape</span>
<span class="n">biplot</span><span class="p">(</span><span class="n">X_pca</span><span class="p">,</span> <span class="n">np</span><span class="o">.</span><span class="n">transpose</span><span class="p">(</span><span class="n">pca</span><span class="o">.</span><span class="n">components_</span><span class="p">),</span> <span class="n">labels</span><span class="o">=</span><span class="n">feature_names</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>
</pre></div>
</div>
</div>
</details>
<div class="cell_output docutils container">
<img alt="../../_images/c159be43cdaf0caabdbb9aa7e50842ad6ffc9671364b971a26b3903da249fc38.png" src="../../_images/c159be43cdaf0caabdbb9aa7e50842ad6ffc9671364b971a26b3903da249fc38.png" />
</div>
</div>
<p><strong>Interpretation of the Biplot:</strong></p>
<ul class="simple">
<li><p><strong>PC1 (the horizontal axis):</strong> All four variables have vectors pointing to the right, but <code class="docutils literal notranslate"><span class="pre">petal</span> <span class="pre">length</span></code>, <code class="docutils literal notranslate"><span class="pre">petal</span> <span class="pre">width</span></code>, and <code class="docutils literal notranslate"><span class="pre">sepal</span> <span class="pre">length</span></code> point most strongly in this direction. This suggests PC1 is a measure of <strong>overall flower size</strong>. Larger flowers (like Virginica) are on the right (high PC1 score), and smaller flowers (like Setosa) are on the left (low PC1 score).</p></li>
<li><p><strong>PC2 (the vertical axis):</strong> This axis shows an interesting contrast. <code class="docutils literal notranslate"><span class="pre">sepal</span> <span class="pre">width</span></code> points up, while <code class="docutils literal notranslate"><span class="pre">petal</span> <span class="pre">width</span></code> and <code class="docutils literal notranslate"><span class="pre">petal</span> <span class="pre">length</span></code> point down. This component separates flowers with wide sepals relative to their petal size from those with the opposite characteristics. This explains the separation between Versicolor and Virginica, which have similar overall sizes (PC1) but different shapes (PC2).</p></li>
</ul>
</section>
</section>
</section>
<section id="application-exercises">
<h2><span class="section-number">4.12. </span>Application Exercises<a class="headerlink" href="#application-exercises" title="Link to this heading">#</a></h2>
<p>Now it’s your turn! Apply the techniques learned above to the following datasets. For each exercise, you’ll need to:</p>
<ol class="arabic simple">
<li><p>Load the data using <code class="docutils literal notranslate"><span class="pre">pandas</span></code>.</p></li>
<li><p>Select the feature columns.</p></li>
<li><p>Standardize the features.</p></li>
<li><p>Perform PCA.</p></li>
<li><p>Create and interpret a scores plot and/or a biplot.</p></li>
</ol>
<p><em>(Note: You will need to have the <code class="docutils literal notranslate"><span class="pre">.csv</span></code> files in the same directory as this notebook for the code to work.)</em></p>
<section id="genomics-cancer-subtype-identification">
<h3><span class="section-number">4.12.1. </span>Genomics - Cancer Subtype Identification<a class="headerlink" href="#genomics-cancer-subtype-identification" title="Link to this heading">#</a></h3>
<p><strong>Dataset:</strong> <code class="docutils literal notranslate"><span class="pre">cancer_data.csv</span></code>
<strong>Task:</strong> Perform PCA on the gene expression data. Can you identify distinct clusters in a scores plot? What might they represent?</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="k">try</span><span class="p">:</span>
    <span class="n">cancer_df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="s1">&#39;cancer_data.csv&#39;</span><span class="p">)</span>
    <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Cancer data loaded successfully!&quot;</span><span class="p">)</span>
    <span class="c1"># Your code here</span>
    <span class="c1"># 1. Select all columns except &#39;Sample_ID&#39; and &#39;Subtype&#39; as your features (X)</span>
    <span class="c1"># X_cancer = ...</span>
    
    <span class="c1"># 2. Standardize X_cancer</span>
    <span class="c1"># X_cancer_scaled = ...</span>
    
    <span class="c1"># 3. Perform PCA (n_components=2)</span>
    <span class="c1"># pca_cancer = ...</span>
    <span class="c1"># X_cancer_pca = ...</span>
    
    <span class="c1"># 4. Create a scores plot. You can color by the &#39;Subtype&#39; column to see if the clusters match.</span>
    <span class="c1"># plt.figure(...)</span>
    
<span class="k">except</span> <span class="ne">FileNotFoundError</span><span class="p">:</span>
    <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;File &#39;cancer_data.csv&#39; not found. Please make sure it&#39;s in the same directory as the notebook.&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>File &#39;cancer_data.csv&#39; not found. Please make sure it&#39;s in the same directory as the notebook.
</pre></div>
</div>
</div>
</div>
</section>
<section id="chemistry-classifying-olive-oils">
<h3><span class="section-number">4.12.2. </span>Chemistry - Classifying Olive Oils<a class="headerlink" href="#chemistry-classifying-olive-oils" title="Link to this heading">#</a></h3>
<p><strong>Dataset:</strong> <code class="docutils literal notranslate"><span class="pre">olive_oil_spectra.csv</span></code>
<strong>Task:</strong> Use PCA on the spectral data. Can you distinguish oils from different regions? Which wavelengths (columns) are most important for this separation?</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="k">try</span><span class="p">:</span>
    <span class="n">oil_df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="s1">&#39;olive_oil_spectra.csv&#39;</span><span class="p">)</span>
    <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Olive oil data loaded successfully!&quot;</span><span class="p">)</span>
    <span class="c1"># Your code here</span>
    <span class="c1"># 1. Select the wavelength columns as features.</span>
    <span class="c1"># X_oil = ...</span>

    <span class="c1"># 2. Standardize and perform PCA.</span>
    <span class="c1"># ...</span>

    <span class="c1"># 3. Create a biplot. It might be too cluttered to label all the variables (wavelengths),</span>
    <span class="c1">#    so focus on the scores plot part and the general direction of the loadings cloud.</span>
    <span class="c1">#    Color the points by the &#39;Region&#39; column.</span>
    <span class="c1"># ...</span>
<span class="k">except</span> <span class="ne">FileNotFoundError</span><span class="p">:</span>
    <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;File &#39;olive_oil_spectra.csv&#39; not found. Please make sure it&#39;s in the same directory as the notebook.&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>File &#39;olive_oil_spectra.csv&#39; not found. Please make sure it&#39;s in the same directory as the notebook.
</pre></div>
</div>
</div>
</div>
</section>
<section id="environmental-science-air-pollution-sources">
<h3><span class="section-number">4.12.3. </span>Environmental Science - Air Pollution Sources<a class="headerlink" href="#environmental-science-air-pollution-sources" title="Link to this heading">#</a></h3>
<p><strong>Dataset:</strong> <code class="docutils literal notranslate"><span class="pre">air_pollution.csv</span></code>
<strong>Task:</strong> Use PCA to identify patterns in air quality data. Create a biplot and interpret the loadings. Can you hypothesize what physical processes PC1 and PC2 represent?</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="k">try</span><span class="p">:</span>
    <span class="n">air_df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="s1">&#39;air_pollution.csv&#39;</span><span class="p">)</span>
    <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Air pollution data loaded successfully!&quot;</span><span class="p">)</span>
    <span class="c1"># Your code here</span>
    <span class="c1"># 1. Select the pollutant and meteorological columns as features.</span>
    <span class="c1"># X_air = ...</span>

    <span class="c1"># 2. Standardize and perform PCA.</span>
    <span class="c1"># ...</span>

    <span class="c1"># 3. Create a biplot and interpret the loadings.</span>
    <span class="c1">#    Look at how variables like Ozone, NOx, and Temperature are related in the PCA space.</span>
    <span class="c1"># ...</span>
<span class="k">except</span> <span class="ne">FileNotFoundError</span><span class="p">:</span>
    <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;File &#39;air_pollution.csv&#39; not found. Please make sure it&#39;s in the same directory as the notebook.&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>File &#39;air_pollution.csv&#39; not found. Please make sure it&#39;s in the same directory as the notebook.
</pre></div>
</div>
</div>
</div>
</section>
<section id="agriculture-crop-yield-analysis">
<h3><span class="section-number">4.12.4. </span>Agriculture - Crop Yield Analysis<a class="headerlink" href="#agriculture-crop-yield-analysis" title="Link to this heading">#</a></h3>
<p><strong>Dataset:</strong> <code class="docutils literal notranslate"><span class="pre">crop_data.csv</span></code>
<strong>Task:</strong> Perform PCA on the input variables (everything except <code class="docutils literal notranslate"><span class="pre">Crop_Yield</span></code>). Then, plot the PC1 scores against <code class="docutils literal notranslate"><span class="pre">Crop_Yield</span></code>. Is there a relationship? This demonstrates using PCA for feature engineering.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="k">try</span><span class="p">:</span>
    <span class="n">crop_df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="s1">&#39;crop_data.csv&#39;</span><span class="p">)</span>
    <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Crop data loaded successfully!&quot;</span><span class="p">)</span>
    <span class="c1"># Your code here</span>
    <span class="c1"># 1. Select the input variables as features.</span>
    <span class="c1"># input_vars = [&#39;Rainfall&#39;, &#39;Sunlight_Hours&#39;, &#39;Fertilizer_Amount&#39;, &#39;Soil_pH&#39;]</span>
    <span class="c1"># X_crop = crop_df[input_vars]</span>
    <span class="c1"># y_crop = crop_df[&#39;Crop_Yield&#39;]</span>

    <span class="c1"># 2. Standardize and perform PCA.</span>
    <span class="c1"># ...</span>
    <span class="c1"># X_crop_pca = ...</span>

    <span class="c1"># 3. Create a scatter plot of the first principal component vs. Crop_Yield.</span>
    <span class="c1"># plt.figure(...)</span>
    <span class="c1"># plt.scatter(X_crop_pca[:, 0], y_crop)</span>
    <span class="c1"># plt.xlabel(&#39;Principal Component 1&#39;)</span>
    <span class="c1"># plt.ylabel(&#39;Crop Yield&#39;)</span>
    <span class="c1"># plt.title(&#39;PC1 vs. Crop Yield&#39;)</span>
    <span class="c1"># plt.show()</span>
    <span class="c1"># What does the relationship (or lack thereof) tell you?</span>

<span class="k">except</span> <span class="ne">FileNotFoundError</span><span class="p">:</span>
    <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;File &#39;crop_data.csv&#39; not found. Please make sure it&#39;s in the same directory as the notebook.&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>File &#39;crop_data.csv&#39; not found. Please make sure it&#39;s in the same directory as the notebook.
</pre></div>
</div>
</div>
</div>
</section>
<section id="pharmaceutical-analysis">
<h3><span class="section-number">4.12.5. </span>Pharmaceutical Analysis<a class="headerlink" href="#pharmaceutical-analysis" title="Link to this heading">#</a></h3>
<p><strong>Scenario</strong>: A pharmaceutical company measured 50 chemical properties of 200 drug candidates to predict effectiveness.</p>
<p><strong>Data Structure</strong>:</p>
<ul class="simple">
<li><p>Rows: 200 drug candidates</p></li>
<li><p>Columns: 50 chemical properties (molecular weight, logP, polar surface area, etc.)</p></li>
</ul>
<p><strong>Tasks</strong>:</p>
<ol class="arabic simple">
<li><p><strong>Preprocessing</strong>: Should you standardize the variables? Why?</p></li>
<li><p><strong>Analysis</strong>: Apply PCA and determine how many components to retain</p></li>
<li><p><strong>Interpretation</strong>: If PC1 loads heavily on molecular weight, logP, and size-related properties, what does this PC represent?</p></li>
<li><p><strong>Application</strong>: How would you use PCA scores to select promising drug candidates?</p></li>
</ol>
<p><strong>Expected Results</strong>:</p>
<ul class="simple">
<li><p>PC1 (30%): “Molecular size” - larger, more lipophilic molecules</p></li>
<li><p>PC2 (20%): “Polarity” - hydrophilic vs. hydrophobic character</p></li>
<li><p>PC3 (15%): “Complexity” - structural complexity measures</p></li>
</ul>
</section>
<section id="protein-structure-analysis">
<h3><span class="section-number">4.12.6. </span>Protein Structure Analysis<a class="headerlink" href="#protein-structure-analysis" title="Link to this heading">#</a></h3>
<p><strong>Scenario</strong>: You have 3D coordinates for all atoms in a protein from molecular dynamics simulations (1000 time points).</p>
<p><strong>Challenge</strong>: Identify main modes of protein flexibility.</p>
<p><strong>Tasks</strong>:</p>
<ol class="arabic simple">
<li><p><strong>Data Setup</strong>: How would you arrange the coordinate data for PCA?</p></li>
<li><p><strong>Preprocessing</strong>: What preprocessing steps are crucial?</p></li>
<li><p><strong>Interpretation</strong>: What do the principal components represent physically?</p></li>
<li><p><strong>Validation</strong>: How would you verify your results make biological sense?</p></li>
</ol>
<p><strong>Solution Approach</strong>:</p>
<ul class="simple">
<li><p><strong>Data matrix</strong>: time_points × (3 × number_of_atoms)</p></li>
<li><p><strong>Preprocessing</strong>: Center each structure, possibly align to remove rotation</p></li>
<li><p><strong>PC1</strong>: Often the “breathing” mode (overall expansion/contraction)</p></li>
<li><p><strong>PC2-PC3</strong>: Hinge motions, domain movements</p></li>
</ul>
</section>
<section id="astronomical-data">
<h3><span class="section-number">4.12.7. </span>Astronomical Data<a class="headerlink" href="#astronomical-data" title="Link to this heading">#</a></h3>
<p><strong>Scenario</strong>: Telescope survey measured brightness in 20 wavelength bands for 10,000 stars.</p>
<p><strong>Goal</strong>: Classify star types using PCA.</p>
<p><strong>Questions</strong>:</p>
<ol class="arabic simple">
<li><p>How many principal components might you expect to need?</p></li>
<li><p>What would the principal components represent physically?</p></li>
<li><p>How would you use PCA results for star classification?</p></li>
</ol>
<p><strong>Hints</strong>:</p>
<ul class="simple">
<li><p>Different star types have characteristic spectra</p></li>
<li><p>Temperature affects overall brightness curve shape</p></li>
<li><p>Chemical composition affects specific absorption lines</p></li>
<li><p>PC1 likely relates to stellar temperature</p></li>
<li><p>PC2-PC3 might capture metallicity, surface gravity</p></li>
</ul>
</section>
</section>
<section id="summary">
<h2><span class="section-number">4.13. </span>Summary<a class="headerlink" href="#summary" title="Link to this heading">#</a></h2>
<p><strong>Key Takeaways</strong>:</p>
<ol class="arabic simple">
<li><p><strong>PCA transforms data</strong> to new coordinates that maximize variance</p></li>
<li><p><strong>Principal components are linear combinations</strong> of original variables</p></li>
<li><p><strong>Interpretation requires examining loadings</strong> and variance explained</p></li>
<li><p><strong>Standardization is crucial</strong> when variables have different scales</p></li>
<li><p><strong>PCA is exploratory</strong> - use it to understand data structure</p></li>
<li><p><strong>Limitations exist</strong> - PCA assumes linear relationships</p></li>
</ol>
<p><strong>When to Use PCA</strong>:</p>
<ul class="simple">
<li><p>High-dimensional numerical data</p></li>
<li><p>Variables are correlated</p></li>
<li><p>Want to visualize or reduce dimensionality</p></li>
<li><p>Need to identify main patterns</p></li>
</ul>
<p><strong>When to Avoid PCA</strong>:</p>
<ul class="simple">
<li><p>Variables are already uncorrelated</p></li>
<li><p>All components are equally important</p></li>
<li><p>Nonlinear relationships dominate</p></li>
<li><p>Small sample size relative to variables</p></li>
</ul>
</section>
<hr class="docutils" />
<section id="further-reading">
<h2><span class="section-number">4.14. </span>Further Reading<a class="headerlink" href="#further-reading" title="Link to this heading">#</a></h2>
<p><strong>Books</strong>:</p>
<ul class="simple">
<li><p>“The Elements of Statistical Learning” - Hastie, Tibshirani, Friedman</p></li>
<li><p>“Pattern Recognition and Machine Learning” - Bishop</p></li>
<li><p>“Applied Multivariate Statistical Analysis” - Johnson &amp; Wichern</p></li>
</ul>
<p><strong>Online Resources</strong>:</p>
<ul class="simple">
<li><p>Scikit-learn PCA documentation</p></li>
<li><p>StatQuest PCA videos (Josh Starmer)</p></li>
<li><p>Andrew Ng’s Machine Learning Course (PCA section)</p></li>
</ul>
<p><strong>Research Papers</strong>:</p>
<ul class="simple">
<li><p>Jolliffe, I.T. “Principal Component Analysis” (comprehensive review)</p></li>
<li><p>Ringner, M. “What is principal component analysis?” (Nature Biotechnology)</p></li>
<li><p>Domain-specific PCA applications in your field of interest</p></li>
</ul>
</section>
</section>

    <script type="text/x-thebe-config">
    {
        requestKernel: true,
        binderOptions: {
            repo: "iluvatar1/ML4Sci-lectures",
            ref: "master",
        },
        codeMirrorConfig: {
            theme: "abcdef",
            mode: "python"
        },
        kernelOptions: {
            name: "python3",
            path: "./lectures/03-PCA"
        },
        predefinedOutput: true
    }
    </script>
    <script>kernelName = 'python3'</script>

                </article>
              

              
              
              
              
                <footer class="prev-next-footer d-print-none">
                  
<div class="prev-next-area">
    <a class="left-prev"
       href="../02-LimpiezayPrepDatos/DataCleaning.html"
       title="previous page">
      <i class="fa-solid fa-angle-left"></i>
      <div class="prev-next-info">
        <p class="prev-next-subtitle">previous</p>
        <p class="prev-next-title"><span class="section-number">3. </span>Pandas for data cleaning and analysis</p>
      </div>
    </a>
    <a class="right-next"
       href="../01-reviewPython/PythonReview.html"
       title="next page">
      <div class="prev-next-info">
        <p class="prev-next-subtitle">next</p>
        <p class="prev-next-title"><span class="section-number">5. </span>Python Review:</p>
      </div>
      <i class="fa-solid fa-angle-right"></i>
    </a>
</div>
                </footer>
              
            </div>
            
            
              
                <div class="bd-sidebar-secondary bd-toc"><div class="sidebar-secondary-items sidebar-secondary__inner">


  <div class="sidebar-secondary-item">
  <div class="page-toc tocsection onthispage">
    <i class="fa-solid fa-list"></i> Contents
  </div>
  <nav class="bd-toc-nav page-toc">
    <ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#dimensionality-reduction-methods">4.1. Dimensionality reduction methods</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#introduction-to-pca">4.2. Introduction to PCA</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#example-applications-to-basic-sciences">4.3. Example applications to basic sciences</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#biology-molecular-biology">4.3.1. Biology &amp; Molecular Biology</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#environmental-science-earth-science">4.3.2. Environmental Science &amp; Earth Science</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#chemistry">4.3.3. Chemistry</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#physics">4.3.4. Physics</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#mathematical-foundation">4.4. Mathematical Foundation</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#general-algorithm">4.5. General algorithm</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#step-1-data-centering">4.5.1. Step 1: Data Centering</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#step-2-covariance-matrix">4.5.2. Step 2: Covariance Matrix</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#step-3-eigendecomposition">4.5.3. Step 3: Eigendecomposition</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#step-4-variance-explained">4.5.4. Step 4: Variance Explained</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#pca-applied-to-the-previous-data-using-scikit-learn">4.6. PCA applied to the previous data using scikit learn</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#pca-applied-to-real-data">4.7. PCA applied to real data</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#interpretation-guidelines">4.8. Interpretation Guidelines</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#choosing-number-of-components">4.8.1. Choosing Number of Components</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#loading-interpretation">4.8.2. Loading Interpretation</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#common-pitfalls">4.9. Common Pitfalls</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#advanced-details">4.10. Advanced Details</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#pca-assumptions-and-limitations">4.10.1. PCA Assumptions and Limitations</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#robust-pca">4.10.2. Robust PCA</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#a-detailed-worked-example-the-iris-dataset">4.11. A detailed Worked Example - The Iris Dataset</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#setup-importing-libraries">4.11.1. Setup - Importing Libraries</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#loading-data">4.11.2. Loading data</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#step-1-standardize-the-data">4.11.3. Step 1: Standardize the Data</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#step-2-perform-pca">4.11.4. Step 2: Perform PCA</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#step-3-analyze-and-visualize-the-results">4.11.5. Step 3: Analyze and Visualize the Results</a><ul class="nav section-nav flex-column">
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#scree-plot">4.11.5.1. Scree Plot</a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#scores-plot">4.11.5.2. Scores Plot</a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#biplot">4.11.5.3. Biplot</a></li>
</ul>
</li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#application-exercises">4.12. Application Exercises</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#genomics-cancer-subtype-identification">4.12.1. Genomics - Cancer Subtype Identification</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#chemistry-classifying-olive-oils">4.12.2. Chemistry - Classifying Olive Oils</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#environmental-science-air-pollution-sources">4.12.3. Environmental Science - Air Pollution Sources</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#agriculture-crop-yield-analysis">4.12.4. Agriculture - Crop Yield Analysis</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#pharmaceutical-analysis">4.12.5. Pharmaceutical Analysis</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#protein-structure-analysis">4.12.6. Protein Structure Analysis</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#astronomical-data">4.12.7. Astronomical Data</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#summary">4.13. Summary</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#further-reading">4.14. Further Reading</a></li>
</ul>
  </nav></div>

</div></div>
              
            
          </div>
          <footer class="bd-footer-content">
            
<div class="bd-footer-content__inner container">
  
  <div class="footer-item">
    
<p class="component-author">
By Veronica Arias, Carlos Viviescas, William Oquendo
</p>

  </div>
  
  <div class="footer-item">
    

  <p class="copyright">
    
      © Copyright 2025.
      <br/>
    
  </p>

  </div>
  
  <div class="footer-item">
    
  </div>
  
  <div class="footer-item">
    
  </div>
  
</div>
          </footer>
        

      </main>
    </div>
  </div>
  
  <!-- Scripts loaded after <body> so the DOM is not blocked -->
  <script src="../../_static/scripts/bootstrap.js?digest=dfe6caa3a7d634c4db9b"></script>
<script src="../../_static/scripts/pydata-sphinx-theme.js?digest=dfe6caa3a7d634c4db9b"></script>

  <footer class="bd-footer">
  </footer>
  </body>
</html>